---
title: "Old Code Chunks"
author: "Myron K Gibert Jr"
date: "7/12/2020"
output: pdf_document
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Section 5: TUCRs are contained within mRNA or lncRNA transcripts

## Results 

Of the 481 TUCRs, 186 were previously characterized as intergenic, or not being contained within a known coding or non-coding transcript.  We chose to focus on these TUCRs, as they potentially represent a partial or full novel long non-coding RNA (lncRNA) transcripts.  To identify the RNA transcript(s) that contain these TUCRs, we performed a de novo transcript reassembly using GBM RNA-Seq data from TCGA.  Using this method, we identified 2636 transcripts that contain TUCRs.  We prioritized transcripts that did not overlap with known genes, as these are likely to represent novel genes.  Transcripts that were greater than 10,000 nucleotide (nt) were also eliminated.  Using these criteria, 8 transcripts were identified in GBM, LGG, or both disease types. (Figure 7A, Table 1)  These transcripts were submitted to the Coding Potential Assessing Tool (CPAT) to determine whether they express any coding potential.  Of the 8 transcripts, five represented no coding potential, while three contained coding potential.  In particular, one of these transcripts with coding potential (TUC110-) overlaps with the most highly upregulated TUCR in both LGG and GBM (uc.110, Figures 7B and 7C).

## Methodology

### Stringtie GBM

The hg19 genomic coordinates for each TUCR were acquired from _____________ and were lifted over to hg38 coordinates. The stringtie bioinformatics package was then used to perform a de novo transcript reassembly of all transcripts in 161 GBM, 400 LGG, and 5 normal brain TCGA Cancer RNA-Seq datasets.  First, transcripts present in each RNA-Seq BAM file were compiled into individual BED/GTF format genomic coordinate files using Stringtie.  These files were then merged into one master file containing all transcripts in all samples using Stringtie and an hg38 genome reference file.  The bedtools package was then used to find the genomic coordinates of all transcripts containing ultraconserved regions.  The ________________ group used stringtie to analyze the entirety of the GTex RNA-Seq normal tissue catalog, so bedtools was also used to determine the intersection of TUCRs with their results.  This methodology provides insight into TUCR transcript presence in normal tissues and in brain cancer.

```{bash stringtie, eval = FALSE}
wget ftp://ftp.ensembl.org/pub/release-99/gtf/homo_sapiens/Homo_sapiens.GRCh38.99.gtf.gz
mv Homo_sapiens.GRCh38.99.gtf.gz stringtie
gunzip stringtie/Homo_sapiens.GRCh38.99.gtf.gz

module load gcc/7.1.0
module load stringtie/2.0.6

for bam in $(find GBM-RNASEQ-RAW/* -name '*.bam')
do
name=$(echo $bam | awk -F ".bam" '{print $1}')
name2=$(basename "$name")
if [ ! -f stringtie/$name2.gtf ] 
then
    echo "$name2.gtf does not exist"
    stringtie $bam -o $name.gtf
    mv $name.gtf ./stringtie
else
    echo "$name2.gtf exists... skipping"
fi

if [ ! -f $name.gtf ]; then
  stringtie $bam -o $name.gtf
fi
mv $name.gtf ./stringtie
done

cd stringtie

stringtie --merge *rehead.gtf -G Homo_sapiens.GRCh38.99.gtf -o stringtie_merged.gtf

intersectBed -a hg38.ultraConserved.bed -b CHESSgenes.bed -wa -wb 
  > intersectchess_TUCRs.bed

intersectBed -a hg38.ultraConserved.bed -b stringtie_merged.gtf -wa -wb 
  > intersectstringtie_TUCRs.bed

cp stringtie_merged.gtf stringtie_merged.bed
```

### Stringtie LGG

```{bash stringtie lgg, eval = FALSE}
wget ftp://ftp.ensembl.org/pub/release-99/gtf/homo_sapiens/Homo_sapiens.GRCh38.99.gtf.gz
mv Homo_sapiens.GRCh38.99.gtf.gz stringtie
gunzip stringtie/Homo_sapiens.GRCh38.99.gtf.gz

module load gcc/7.1.0
module load stringtie/2.0.6

for bam in $(find LGG-RNASEQ-RAW/* -name '*.bam')
do
name=$(echo $bam | awk -F ".bam" '{print $1}')
name2=$(basename "$name")
if [ ! -f stringtie/$name2.gtf ] 
then
    echo "$name2.gtf does not exist"
    stringtie $bam -o $name.gtf
    mv $name.gtf ./stringtie
else
    echo "$name2.gtf exists... skipping"
fi

if [ ! -f $name.gtf ]; then
  stringtie $bam -o $name.gtf
fi
mv $name.gtf ./stringtie
done

cd stringtie

stringtie --merge *rehead.gtf -G Homo_sapiens.GRCh38.99.gtf -o stringtie_merged.gtf

intersectBed -a TUCR_intergenic.bed -b stringtie_merged.gtf -wa -wb > intersectintergenic_TUCRs.bed

intersectBed -a intersectintergenic_TUCRs.bed -b stringtie_merged.gtf -wa -wb 
  > intersectintergenic_TUCRs.bed

cp stringtie_merged.gtf stringtie_merged.bed
```

## Show that TUCRs are contained within novel genes

```{r novelgenes, eval = FALSE}
lgg_novelgenes <- read_delim("lgg_novelgenes.bed", col_names = FALSE, delim = "\t",skip_empty_rows = FALSE) %>%     dplyr::select(X5,X7,X9,X10,X11,X13) %>%
  filter(X9 == "transcript") %>%
  mutate(disease = "LGG",name = paste("TUC",str_remove(X5,"uc."),X13,sep = ""), length = X11-X10) %>%
  filter(length <= 30000) %>%
  group_by(name) %>%
  mutate(max = max(length,na.rm=TRUE)) %>%
  filter(length == max) %>%
  ungroup() %>%
  dplyr::select(name,disease,length,"chromosome" = X7,"start" = X10,"end" = X11, "strand" = X13)

gbm_novelgenes <- read_delim("gbm_novelgenes.bed", col_names = FALSE, delim = "\t",skip_empty_rows = FALSE) %>%     dplyr::select(X5,X7,X9,X10,X11,X13) %>%
  filter(X9 == "transcript") %>%
  mutate(disease = "GBM",name = paste("TUC",str_remove(X5,"uc."),X13,sep = ""), length = X11-X10) %>%
  filter(length <= 30000) %>%
  group_by(name) %>%
  mutate(max = max(length,na.rm=TRUE)) %>%
  filter(length == max) %>%
  ungroup() %>%
  dplyr::select(name,disease,length,"chromosome" = X7,"start" = X10,"end" = X11, "strand" = X13)

all_novelgenes <- rbind(gbm_novelgenes,lgg_novelgenes) 
all_novelgenes <- all_novelgenes %>%
  group_by(name) %>%
  mutate(n = n(), annot = ifelse(n==2, "GBM/LGG",disease)) %>%
  dplyr::select(-n)


  

write.csv(all_novelgenes,file="all_novelgenes.csv")

CPAT_bedfile <- all_novelgenes %>%
  dplyr::select(chromosome,start,end,strand,name)

write_delim(CPAT_bedfile,file="CPAT.bed",delim = "\t",col_name = FALSE)

```

### Get Fasta

```{bash novelgenes_getFASTA}

```

### Use CPAT Web Browser

### Integrate CPAT results

```{r}
all_novelgenes <- read.csv(file="all_novelgenes.csv",header=TRUE,row.names=1)

CPAT_results <- read_delim("result.txt", col_names = TRUE, delim = "\t",skip_empty_rows = FALSE) %>%
  dplyr::select("length" = `RNA size`, `ORF size`,`Ficket Score`,`Hexamer Score`,`Coding Probability`,`Coding Label`)

CPAT_novelgenes <- all_novelgenes %>%
  full_join(CPAT_results, by = "length")

write.csv(CPAT_novelgenes,file="CPAT_novelgenes.csv")

```


### Summarizing Novel Gene data

``` {r sum_novelgenes}

CPAT_novelgenes <- read.csv(file="CPAT_novelgenes.csv",header=TRUE,row.names=1)

CPAT_novelgenes_sum <- CPAT_novelgenes %>%
  group_by(annot,Coding.Label) %>% 
  mutate(group_id = cur_group_id(),count = n(),Group = 
    ifelse(group_id==1, "GBM_non_coding",
    ifelse(group_id==2, "GBM_coding",
    ifelse(group_id==3, "GBM/LGG_non_coding",
    ifelse(group_id==4, "GBM/LGG_coding",
    ifelse(group_id==5, "LGG_non_coding",
    ifelse(group_id==6, "LGG_coding","ERROR"))))))) %>%
  group_by(Group,group_id) %>%
  summarize(count=n())

p <- ggplot(CPAT_novelgenes_sum, aes(x=reorder(Group,group_id),y=count,fill=Group)) + geom_bar(stat="identity", width = 0.7, color="black") +
  geom_text(aes(label=round(count,2)), vjust=1.6, color="black", size=rel(4)) + ggtitle(paste("Some intergenic TUCRs are \n contained within Novel Genes")) + xlab("Disease") + ylab("Number of Genes") + theme(
plot.title = element_text(size=rel(1.5), face="bold",hjust = 0.5),
axis.title = element_text(size = rel(1.25), face="bold"),
axis.text.x = element_text(angle = -90, size = 10),
legend.title = element_blank(),
legend.position = "none",
panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black"))

ggsave(file=paste(outputdir,"/tucr_novelgenes.png",sep=""),plot = print(p), width = 5, height = 5, dpi = 300) 
  

```

```{bash ATAC, eval = FALSE}

wget ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3669nnn/GSM3669999/suppl/GSM3669999_GSC23_ATAC.highPValue.distal_intersect.H3K27Ac.bed.gz

gunzip GSM3669999_GSC23_ATAC.highPValue.distal_intersect.H3K27Ac.bed.gz

bedtools shuffle -i GSM3669999_GSC23_ATAC.highPValue.distal_intersect.H3K27Ac.bed -g hg38_genomeFile.txt > GSC23_ATAC.shuffle.bed

bedtools sort -i GSM3669999_GSC23_ATAC.highPValue.distal_intersect.H3K27Ac.bed > GSC23_ATAC.sort.bed

bedtools sort -i GSC23_ATAC.shuffle.bed > GSC23_ATAC.shuffle.sort.bed

bedtools fisher -a tucrs_pm10kb.bed -b GSC23_ATAC.shuffle.sort.bed -g hg38_genomeFile.txt > GSC23_ATAC_random_output.txt

bedtools fisher -a tucrs_pm10kb.bed -b GSC23_ATAC.sort.bed -g hg38_genomeFile.txt > GSC23_ATAC_tucrs_output.txt



wget ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3669nnn/GSM3669993/suppl/GSM3669993_U87_ATAC.highPValue.distal_intersect.H3K27Ac.bed.gz

gunzip GSM3669993_U87_ATAC.highPValue.distal_intersect.H3K27Ac.bed.gz

bedtools shuffle -i GSM3669993_U87_ATAC.highPValue.distal_intersect.H3K27Ac.bed -g hg38_genomeFile.txt > U87_ATAC.shuffle.bed

bedtools sort -i GSM3669993_U87_ATAC.highPValue.distal_intersect.H3K27Ac.bed > U87_ATAC.sort.bed

bedtools sort -i U87_ATAC.shuffle.bed > U87_ATAC.shuffle.sort.bed

bedtools fisher -a tucrs_pm10kb.bed -b U87_ATAC.shuffle.sort.bed -g hg38_genomeFile.txt > U87_ATAC_random_output.txt

bedtools fisher -a tucrs_pm10kb.bed -b U87_ATAC.sort.bed -g hg38_genomeFile.txt > U87_ATAC_tucrs_output.txt

```

```{r}
tidy_heatmap(heatmap.allgo2,
             rows = gene,
             columns = Term,
             values = direction,
             scale = "row",
             #annotation_col = c(sample_type, condition, group),
             #annotation_row = c(is_immune_gene, direction),
             #gaps_row = direction,
             #gaps_col = group
)

tidy_heatmap(df,
             rows = external_gene_name,
             columns = sample,
             values = expression,
             scale = "row",
             #annotation_col = c(sample_type, condition, group),
             #annotation_row = c(is_immune_gene, direction),
             #gaps_row = direction,
             #gaps_col = group
)

class1 <- lapply(data_exprs,class)
class2 <- lapply(heatmap.allgo2,class)
class2 <- lapply(df,class)

write.csv(data_exprs,"datatest1.csv")

df <- read.csv("datatest1.csv",header=TRUE)

write.csv(heatmap.allgo2,"datatest2.csv")
```


```{r}

heatmap.allgo.up <- heatmap.allgo2 %>%
  filter(sum >0) %>%
  group_by(Term,sum) %>%
  summarise()

colors <- colorRampPalette(c("blue","white","red"))(length(unique(heatmap.allgo2$direction)))
heatmap.allgo2$direction <- factor(heatmap.allgo2$direction)
N <- nlevels(heatmap.allgo2$direction)

p <- ggplot(data = heatmap.allgo2,mapping = aes(x = reorder(Term,sum),y = gene, fill = direction)) +
  geom_tile() +
  scale_fill_manual(values = colors,breaks=levels(heatmap.allgo2$direction)[seq(1, N, by=1)]) + 
  ggtitle("Title") + 
  xlab("X-axis") + 
  ylab("Y-axis") +
  theme(plot.title = element_text(hjust = 0.5),
        panel.grid = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank())

ggsave(p,file=paste(outputdir,"/GOTerms_heatmap.png",sep=""), width = 2.5, height = 7, dpi = 300)

```

## Acquiring SNPs

```{bash, eval = FALSE}
mkdir SNPs

cd SNPs

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_1.bed.gz
gunzip bed_chr_1.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_10.bed.gz
gunzip bed_chr_10.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_11.bed.gz
gunzip bed_chr_11.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_13.bed.gz
gunzip bed_chr_13.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_12.bed.gz
gunzip bed_chr_12.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_14.bed.gz
gunzip bed_chr_14.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_15.bed.gz
gunzip bed_chr_15.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_16.bed.gz
gunzip bed_chr_16.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_17.bed.gz
gunzip bed_chr_17.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_18.bed.gz
gunzip bed_chr_18.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_19.bed.gz
gunzip bed_chr_19.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_2.bed.gz
gunzip bed_chr_2.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_20.bed.gz
gunzip bed_chr_20.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_21.bed.gz
gunzip bed_chr_21.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_22.bed.gz
gunzip bed_chr_22.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_3.bed.gz
gunzip bed_chr_3.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_4.bed.gz
gunzip bed_chr_4.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_5.bed.gz
gunzip bed_chr_5.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_6.bed.gz
gunzip bed_chr_6.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_7.bed.gz
gunzip bed_chr_7.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_8.bed.gz
gunzip bed_chr_8.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_9.bed.gz
gunzip bed_chr_9.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_X.bed.gz
gunzip bed_chr_X.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_Y.bed.gz
gunzip bed_chr_Y.bed.gz

<<<<<<< HEAD

=======
```

### Remove first line

``` {r SNPgetter}
for bed in bed_chr_*
do

tail -n +2 $bed > $bed.stdout
mv $bed.stdout $bed

done
>>>>>>> 4e0fba6348ce12fb6f44cb5236d74db55822a941

module load gcc/7.1.0
module load bedtools/2.26.0

for bed in bed_chr_*
do
echo $bed
intersectBed -a hg38.ultraConserved.bed -b $bed -wa -wb > $bed.txt
done

cat *.txt > merged.txt
```

### 

``` {r SNPreader}

library("tidyverse")

SNPcounts <- read.csv("TUCR_SNPs.csv")

```

## Make figures for fold change

```{r overexp_TUCR}

overexp_TUCR_cp <- c("#C77CFF","#F8766D","#619CFF","#00BA38","#FFFF00")

filename <- "tucroverexpression_U87.csv"
x <- read.csv(filename)

p <- ggplot(x,aes(x = Target, y = FC, fill = Sample)) + 
  labs(title = "TUCR Overexpression in U87 cells") +
  ylab("Fold Change") +
  geom_bar(stat = "identity",position="dodge",colour="black") + 
  theme(axis.text.x = element_text(angle = -90, size = 10),
        axis.text.y = element_text(size = 10),
        axis.ticks.x = element_blank(),
        plot.title = element_text(size = 15, face = "bold",hjust = 0.5),
        axis.title.y = element_text("Fold Change",size = 10)) +
  facet_zoom(ylim = c(0, 2)) +
  scale_fill_manual(values=overexp_TUCR_cp)

ggsave(p,file=paste(outputdir,"/",filename,".png",sep=""), width = 7.5, height = 5, dpi = 300)


```

## Generate figures for cell fractionation

```{r cellfrac_TUCR}

if(!dir.exists(paste(outputdir,"/cellfrac",sep=""))){
  dir.create(paste(outputdir,"/cellfrac",sep=""))
}

CellFrackin <- function(filename){
all_content = readLines(filename)
skip_second = all_content[-c(1:19)]
x <- read.csv(textConnection(skip_second), header = TRUE, stringsAsFactors = FALSE)
samplenames <- t(as.data.frame(strsplit(as.character(x$Sample), " ")))
colnames(samplenames) <- c("Line","Fraction")
x <- cbind(x,samplenames)
x<- x %>%
    dplyr::select(Well,Target,Sample,Line,Fraction,Cq) %>%
    dplyr::group_by(Target,Line,Fraction) %>%
    dplyr::summarize(meanCq = mean(Cq)) %>%
    dplyr::mutate(dct = meanCq - min(meanCq)) %>%
    dplyr::mutate(FoldChange = 2^(-dct)) %>%
    dplyr::mutate(ID = paste(Target,Line,Fraction,sep=" "))

p <- ggplot(x,aes(x = Target, y = FoldChange, fill = Fraction)) + 
  labs(title = "Cellular localization of TUCRs in GBM cell lines") +
  ylab("Fold Change") +
  geom_bar(stat = "identity",position="dodge") + 
  theme(axis.text.x = element_text(angle = -90, size = 10),
        axis.text.y = element_text(size = 10),
        axis.ticks.x = element_blank(),
        plot.title = element_text(size = 15, face = "bold",hjust = 0.5),
        axis.title.y = element_text("Fold Change",size = 10),
panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  facet_wrap(~Line)

ggsave(p,file=paste(outputdir,"/",filename,".png",sep=""), width = 7.5, height = 5, dpi = 300)
x %>%
  select(-ID) %>%
  write_csv(paste(outputdir,"/",filename," (Output).csv",sep=""))
}

path <- list.files("cellfrac")
i = 1

for(i in 1:length(path)){
  print(i)
  CellFrackin(paste("cellfrac",path[i],sep="/"))
}

```

## Summarize GO Terms (old)

```{r summary GO Terms old}

allgo <- read.csv(paste(outputdir,"/GOterm_all_summary_TUCRs.csv",sep=""),header=TRUE)

allgo_bp <- read.csv(paste(outputdir,"/GOterm_bp_summary_TUCRs.csv",sep=""),header=TRUE)

allgo_mf <- read.csv(paste(outputdir,"/GOterm_mf_summary_TUCRs.csv",sep=""),header=TRUE)

allgo_cc <- read.csv(paste(outputdir,"/GOterm_cc_summary_TUCRs.csv",sep=""),header=TRUE)

allgo_sum <- allgo %>% 
  dplyr::group_by(Term) %>%
  dplyr::summarise(count = n()) %>%
  dplyr::arrange(desc(count)) %>%
  dplyr::mutate(Term=factor(Term, levels=Term))

allgo_sum <- allgo_sum[1:20,]

p <- ggplot(allgo_sum, aes(x=Term,y=count,fill=Term)) + geom_bar(stat="identity", width = 0.7) + geom_text(aes(label=round(count,2)), vjust=1.6, color="black", size=rel(2)) + ggtitle(paste("Most common TUCR GO Terms in ",disease,sep="")) +xlab("GO Term") + ylab("# of TUCRs") + 
  theme(
  plot.title = element_text(size=rel(1.5), face="bold",hjust = 0.5),
axis.title = element_text(size = rel(1.25), face="bold"),
legend.title = element_blank(),
legend.position = "none",
axis.text.x = element_text(size=rel(1.5)),
axis.text.y = element_text(size=rel(1.5)),
panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black")) + coord_flip()

ggsave(file=paste(outputdir,"/tucr_allgo.png",sep=""),plot = print(p), width = 15, height = 10, dpi = 300) 


allgo_bp_sum <- allgo_bp %>% 
  dplyr::group_by(Term) %>%
  dplyr::summarise(count = n()) %>%
  dplyr::arrange(desc(count)) %>%
  dplyr::mutate(Term=factor(Term, levels=Term))

allgo_bp_sum <- allgo_bp_sum[1:20,]

p1 <- ggplot(allgo_bp_sum, aes(x=Term,y=count,fill=Term)) + geom_bar(stat="identity", width = 0.7) + geom_text(aes(label=round(count,2)), vjust=1.6, color="black", size=rel(2)) + ggtitle(paste("Most common TUCR Biological Processes in ",disease,sep="")) + xlab("Biological Processes") + ylab("# of TUCRs") + theme(
  plot.title = element_text(size=rel(1.5), face="bold",hjust = 0.5),
axis.title = element_text(size = rel(1.25), face="bold"),
legend.title = element_blank(),
legend.position = "none",
axis.text.x = element_text(size=rel(1.5)),
axis.text.y = element_text(size=rel(1.5)),
panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black")) + coord_flip()

ggsave(file=paste(outputdir,"/tucr_allgo_bp.png",sep=""),plot = print(p1), width = 15, height = 10, dpi = 300) 

allgo_mf_sum <- allgo_mf %>% 
  dplyr::group_by(Term) %>%
  dplyr::summarise(count = n()) %>%
  dplyr::arrange(desc(count)) %>%
  dplyr::mutate(Term=factor(Term, levels=Term))

allgo_mf_sum <- allgo_mf_sum[1:20,]

p2 <- ggplot(allgo_mf_sum, aes(x=Term,y=count,fill=Term)) + geom_bar(stat="identity", width = 0.7) + geom_text(aes(label=round(count,2)), vjust=1.6, color="black", size=rel(2)) + ggtitle(paste("Most common TUCR Molecular Functions in ",disease,sep="")) + xlab("Molecular Function") + ylab("# of TUCRs") + theme(
  plot.title = element_text(size=rel(1.5), face="bold",hjust = 0.5),
axis.title = element_text(size = rel(1.25), face="bold"),
legend.title = element_blank(),
legend.position = "none",
axis.text.x = element_text(size=rel(1.5)),
axis.text.y = element_text(size=rel(1.5)),
panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black")) + coord_flip()

ggsave(file=paste(outputdir,"/tucr_allgo_mf.png",sep=""),plot = print(p2), width = 15, height = 10, dpi = 300) 

allgo_cc_sum <- allgo_cc %>% 
  dplyr::group_by(Term) %>%
  dplyr::summarise(count = n()) %>%
  dplyr::arrange(desc(count)) %>%
  dplyr::mutate(Term=factor(Term, levels=Term))

allgo_cc_sum <- allgo_cc_sum[1:20,]

p3 <- ggplot(allgo_cc_sum, aes(x=Term,y=count,fill=Term)) + geom_bar(stat="identity", width = 0.7) + geom_text(aes(label=round(count,2)), vjust=1.6, color="black", size=rel(2)) + ggtitle(paste("Most common TUCR Cellular Compartments in ",disease,sep="")) + xlab("Cellular Compartment") + ylab("# of TUCRs") + theme(
  plot.title = element_text(size=rel(1.5), face="bold",hjust = 0.5),
axis.title = element_text(size = rel(1.25), face="bold"),
legend.title = element_blank(),
legend.position = "none",
axis.text.x = element_text(size=rel(1.5)),
axis.text.y = element_text(size=rel(1.5)),
panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black")) + coord_flip()

ggsave(file=paste(outputdir,"/tucr_allgo_cc.png",sep=""),plot = print(p3), width = 15, height = 10, dpi = 300) 

allgo_bp_sum <- allgo_bp_sum %>%
  dplyr::mutate(Ontology = "Biological Process")

allgo_mf_sum <- allgo_mf_sum %>%
  dplyr::mutate(Ontology = "Molecular Function")

allgo_cc_sum <- allgo_cc_sum %>%
  dplyr::mutate(Ontology = "Cellular Compartment")

allgo_fig <- rbind(allgo_bp_sum,allgo_mf_sum)

p4 <- ggplot(allgo_fig, aes(x=Term,y=count,fill=Term)) + geom_bar(stat="identity",position="dodge", width = 0.7) + geom_text(aes(label=round(count,2)), vjust=1.6, color="black", size=rel(4)) + ggtitle(paste("Most common TUCR GO Terms in ",disease,sep="")) + xlab("GO Term") + ylab("# of TUCRs") + facet_grid(~ Ontology,scales="free",space="free") + theme(
  plot.title = element_text(size=rel(1.75), face="bold",hjust = 0.5),
axis.title = element_text(size = rel(1.25), face="bold"),
legend.title = element_blank(),
legend.position = "none",
axis.text.x = element_text(size=rel(1.5)),
axis.text.y = element_text(size=rel(1.5)),
strip.text = element_text(size=rel(1.5)),
panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black")) 

ggsave(file=paste(outputdir,"/tucr_allgo_fig.png",sep=""),plot = print(p4), width = 15, height = 15, dpi = 300)


```

## Generate heatmap for my TUCRs of interest

```{r heatmap_Myron}
heatmap.TUCRs <- rpkm.TUCRs2[,6:ncol(rpkm.TUCRs2)]
TUCRids <- rpkm.TUCRs2[,4]
means <- apply(heatmap.TUCRs,1,mean)

heatmap.TUCRs2 <- as.data.frame(apply(heatmap.TUCRs,1:2,function(x) {ifelse(x>=10,10,x)}))
heatmap.TUCRs2 <- cbind(TUCRids,means,heatmap.TUCRs2)

heatmap.TUCRs2 <- heatmap.TUCRs2 %>% 
  gather(key = "Sample", value = "RPKM",-TUCRids,-means) %>% 
  filter(TUCRids == "uc.62"|TUCRids == "uc.110"|TUCRids == "uc.193"|TUCRids == "uc.210"|TUCRids == "uc.427"|TUCRids == "uc.212"|TUCRids == "uc.341")

p <- ggplot(data = heatmap.TUCRs2,mapping = aes(x = Sample,y = reorder(TUCRids,means), fill = RPKM)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("blue", "white", "red"), values = c(0,0.1,1)) +
  theme(panel.grid = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks = element_blank())

ggsave(p,file=paste(outputdir,"/Myron_heatmap.png",sep=""))
```

## Identifying differentially expressed stringtie transcripts containing TUCRs with DESeq2

```{r DESeq2_stringtie, results="hide", message = FALSE, eval = FALSE}

stringtiecounts <- read.table("stringtie_TUCR_counts.txt",header = TRUE)
stringtiecounts <- distinct(stringtiecounts)
metadata <- read_csv(file = metadatafile)

stringtiecounts2 <- stringtiecounts[,5:ncol(stringtiecounts)]
rownames(stringtiecounts2) <- stringtiecounts[,4]
stringtiecounts.info <- stringtiecounts[,1:4]
stringtiecounts.info$length <- (stringtiecounts$end - stringtiecounts$start)/1000
stringtiecounts <- cbind(stringtiecounts.info,stringtiecounts2)

dds_stringtie <- DESeqDataSetFromMatrix(countData = stringtiecounts2,
                              colData = metadata,
                              design = ~dex)
dds_stringtie <- DESeq(dds_stringtie)
res_stringtie <- results(dds_stringtie, tidy=TRUE)
res_stringtie <- as_tibble(res_stringtie)

write.csv(res_stringtie, file = paste(outputdir,"/results_stringtie.csv",sep=""))

```

## Completing survival analysis for stringtie transcripts containing TUCRs

```{r surcor_stringtie, message = FALSE,eval = FALSE}
if(!file.exists("GBM.clin.merged.txt")){unzip("GBM.clin.merged.zip")}

countdata <- read.table("stringtie_TUCR_counts.txt",header = TRUE)
metadata <- read_csv(file = metadatafile)

posdata <- countdata[,1:4]
countdata <- countdata[,5:length(colnames(countdata))]

countdata <- as.matrix(countdata)


n_index <- c(seq(36,39),73)
t_index <- c(seq(1,35),seq(40,72),seq(74,ncol(countdata)))

vm <- function(x){
  cond <- factor(ifelse(seq(1,dim(x)[2],1) %in% t_index, 1,  0))
  d <- model.matrix(~1+cond)
  x <- t(apply(x,1,as.numeric))
  ex <- voom(x,d,plot=F)
  return(ex$E)
}

count_vm <- vm(countdata)
colnames(count_vm) <- metadata$barcode


scal <- function(x,y){
  mean_n <- rowMeans(y)  # mean of normal
  sd_n <- apply(y,1,sd)  # SD of normal
  # z score as (value - mean normal)/SD normal
  res <- matrix(nrow=nrow(x), ncol=ncol(x))
  colnames(res) <- colnames(x)
  rownames(res) <- rownames(x)
  for(i in 1:dim(x)[1]){
    for(j in 1:dim(x)[2]){
      res[i,j] <- (x[i,j]-mean_n[i])/sd_n[i]
    }
  }
  return(res)
}

z_rna <- scal(count_vm[,t_index],count_vm[,n_index])
rownames(z_rna) <- posdata[,4]

clinical <- read.table("GBM.clin.merged.txt",header = TRUE)
clinical$time <- as.numeric(clinical$time)

sum(clinical$patient %in% colnames(z_rna))

ind_tum <- which(unique(colnames(z_rna)) %in% clinical$patient)
ind_clin <- which(clinical$patient %in% colnames(z_rna))

out.tab <- c()
for(x in 1:nrow(count_vm)){
  ind_gene <- x
  s <- Surv(clinical$time[ind_clin],clinical$status[ind_clin])
  cx <- coxph(formula = s ~ count_vm[ind_gene,ind_tum])
  cx <- tidy(cx)
  out.tab <- rbind(out.tab,cx)
}

surv_stringtie <- cbind(posdata,out.tab)
surv_stringtie <- surv_stringtie %>% select(-term)

write_csv(surv_stringtie,paste(outputdir,"/survival_TUCRs_stringtie.csv",sep=""))
```

## Generating volcano plot to visualize DE stringtie transcripts

```{r DESeq2plots_stringtie, message = FALSE, eval = FALSE}

### Volcano plot

volcanoplot <- function (res,title = "Deregulated Genes", output = "results_volcanoplot.png"){
  #res <- res_TUCR
  #genes = c("uc.110","uc.62")
  #title = "Deregulated Genes"
  vres <- res
  vres <- vres %>% mutate(filter1 = abs(log2FoldChange)>=1,filter2 = padj <=0.05,filter3="")#,gene="")
  for (i in 1:length(vres$row)){
    if(is.na(vres$padj[i])){vres$filter3[i] <- "black"}else{
    if(vres$filter1[i] == TRUE && vres$filter2[i] == TRUE){
    vres$filter3[i] <- "red"
  }else{
      if(vres$filter1[i] == TRUE){
    vres$filter3[i] <- "blue"
      }else{vres$filter3[i] <- "black"}
  }}
    #ifelse(genes!="all",ifelse(!is.na(match(vres$row[i],genes)),vres$gene[i] <- vres$row[i], ""),vres$gene[i] <- vres$row[i])
    }
  
  
  
  ## plot
  p <- ggplot(vres) +
        geom_point(aes(x=log2FoldChange, y=-log10(padj),col=filter3)) +
        scale_color_manual(values = c("black","blue","red")) +
        ggtitle(title) +
        xlab("log2 fold change") + 
        ylab("-log10 adjusted p-value") +
        theme(Legend.position = "none",
              plot.title = element_text(size = rel(1.5), hjust = 0.5),
              axis.title = element_text(size = rel(1.25)))
  
  ggsave(output)
}

volcanoplot(res_stringtie,title = "Deregulation of stringtie transcripts containing TUCRs in GBM", output = paste(outputdir,"/stringtie_results_volcanoplot.png",sep=""))

```

## Generate stringtie RPKMs for filtering

```{r RPKM_stringtie, results="hide", eval = FALSE}

stringtie.rpkmcounts <- stringtiecounts[,6:ncol(stringtiecounts)]
stringtie.rpkmcounts.info <- stringtiecounts[,1:5]
rownames(stringtie.rpkmcounts) <- stringtiecounts$id

genelength <- stringtiecounts$length

seqdepth <- read.csv(file = "seqdepth_counts.csv",row.names=1)

seqdepth <- as.vector(seqdepth$counts)

rpkm <- function(counts,len,dep){
  x <- counts/len
  return(t(t(x)/dep))
}

rpkm.stringtie <- rpkm(stringtie.rpkmcounts,genelength,seqdepth)
rpkm.stringtie2 <- cbind(stringtie.rpkmcounts.info,rpkm.stringtie)

write.csv(rpkm.stringtie2,paste(outputdir,"/rpkm.stringtie.csv",sep=""))
```

## Generate heatmap for all stringtie TUCR transcripts

```{r heatmap_stringtie, message=FALSE, eval = FALSE}
heatmap.stringtie <- rpkm.stringtie2[,6:ncol(rpkm.stringtie2)]
stringtieids <- rpkm.stringtie2[,5]
means <- apply(heatmap.stringtie,1,mean)

heatmap.stringtie2 <- as.data.frame(apply(heatmap.stringtie,1:2,function(x) {ifelse(x>=10,10,x)}))
heatmap.stringtie2 <- cbind(stringtieids,means,heatmap.stringtie2)

heatmap.stringtie2 <- heatmap.stringtie2 %>% 
  gather(key = "Sample", value = "RPKM",-stringtieids,-means)

p <- ggplot(data = heatmap.stringtie2,mapping = aes(x = Sample,y = reorder(stringtieids,means), fill = RPKM)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("blue", "white", "red"), values = c(0,0.1,1)) +
  theme(panel.grid = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank())

ggsave(p,file=paste(outputdir,"/stringtie_heatmap.png",sep=""))
```

## Identifying differentially expressed CHESS genes containing TUCRs with DESeq2

```{r DESeq2_CHESS, results="hide", message = FALSE, eval = FALSE}

CHESScounts <- read.table("intersectchess_TUCR_counts.txt",header = TRUE)
CHESScounts <- distinct(CHESScounts)
metadata <- read_csv(file = metadatafile)

CHESScounts2 <- CHESScounts[,5:ncol(CHESScounts)]
rownames(CHESScounts2) <- CHESScounts[,4]
CHESScounts.info <- CHESScounts[,1:4]
CHESScounts.info$length <- (CHESScounts$end - CHESScounts$start)/1000
CHESScounts <- cbind(CHESScounts.info,CHESScounts2)

dds_CHESS <- DESeqDataSetFromMatrix(countData = CHESScounts2,
                              colData = metadata,
                              design = ~dex)
dds_CHESS <- DESeq(dds_CHESS)
res_CHESS <- results(dds_CHESS, tidy=TRUE)
res_CHESS <- as_tibble(res_CHESS)

write.csv(res_CHESS, file = paste(outputdir,"/results_CHESS.csv",sep=""))

```

## Generating volcano plot to visualize DE CHESS genes containing TUCRs

```{r DESeq2plots_CHESS, message = FALSE, eval = FALSE}

### Volcano plot

volcanoplot <- function (res,title = "Deregulated Genes", output = "results_volcanoplot.png"){
  #res <- res_TUCR
  #genes = c("uc.110","uc.62")
  #title = "Deregulated Genes"
  vres <- res
  vres <- vres %>% mutate(filter1 = abs(log2FoldChange)>=1,filter2 = padj <=0.05,filter3="")#,gene="")
  for (i in 1:length(vres$row)){
    if(is.na(vres$padj[i])){vres$filter3[i] <- "black"}else{
    if(vres$filter1[i] == TRUE && vres$filter2[i] == TRUE){
    vres$filter3[i] <- "red"
  }else{
      if(vres$filter1[i] == TRUE){
    vres$filter3[i] <- "blue"
      }else{vres$filter3[i] <- "black"}
  }}
    #ifelse(genes!="all",ifelse(!is.na(match(vres$row[i],genes)),vres$gene[i] <- vres$row[i], ""),vres$gene[i] <- vres$row[i])
    }
  
  
  
  ## plot
  p <- ggplot(vres) +
        geom_point(aes(x=log2FoldChange, y=-log10(padj),col=filter3)) +
        scale_color_manual(values = c("black","blue","red")) +
        ggtitle(title) +
        xlab("log2 fold change") + 
        ylab("-log10 adjusted p-value") +
        theme(Legend.position = "none",
              plot.title = element_text(size = rel(1.5), hjust = 0.5),
              axis.title = element_text(size = rel(1.25)))
  
  ggsave(output)
}

volcanoplot(res_CHESS,title = "Deregulation of CHESS genes containing TUCRs in GBM", output = paste(outputdir,"/CHESS_results_volcanoplot.png",sep=""))

```

## Completing survival analysis for stringtie transcripts containing TUCRs

```{r surcor_CHESS, message = FALSE, eval = FALSE}
if(!file.exists("GBM.clin.merged.txt")){unzip("GBM.clin.merged.zip")}

countdata <- read.table("intersectchess_TUCR_counts.txt",header = TRUE)
metadata <- read_csv(file = metadatafile)

posdata <- countdata[,1:4]
countdata <- countdata[,5:length(colnames(countdata))]

countdata <- as.matrix(countdata)

n_index <- c(seq(36,39),73)
t_index <- c(seq(1,35),seq(40,72),seq(74,ncol(countdata)))

vm <- function(x){
  cond <- factor(ifelse(seq(1,dim(x)[2],1) %in% t_index, 1,  0))
  d <- model.matrix(~1+cond)
  x <- t(apply(x,1,as.numeric))
  ex <- voom(x,d,plot=F)
  return(ex$E)
}

count_vm <- vm(countdata)
colnames(count_vm) <- metadata$barcode


scal <- function(x,y){
  mean_n <- rowMeans(y)  # mean of normal
  sd_n <- apply(y,1,sd)  # SD of normal
  # z score as (value - mean normal)/SD normal
  res <- matrix(nrow=nrow(x), ncol=ncol(x))
  colnames(res) <- colnames(x)
  rownames(res) <- rownames(x)
  for(i in 1:dim(x)[1]){
    for(j in 1:dim(x)[2]){
      res[i,j] <- (x[i,j]-mean_n[i])/sd_n[i]
    }
  }
  return(res)
}

z_rna <- scal(count_vm[,t_index],count_vm[,n_index])
rownames(z_rna) <- posdata[,4]

clinical <- read.table("GBM.clin.merged.txt",header = TRUE)
clinical$time <- as.numeric(clinical$time)

sum(clinical$patient %in% colnames(z_rna))

ind_tum <- which(unique(colnames(z_rna)) %in% clinical$patient)
ind_clin <- which(clinical$patient %in% colnames(z_rna))

out.tab <- c()
for(x in 1:nrow(count_vm)){
  ind_gene <- x
  s <- Surv(clinical$time[ind_clin],clinical$status[ind_clin])
  cx <- coxph(formula = s ~ count_vm[ind_gene,ind_tum])
  cx <- tidy(cx)
  out.tab <- rbind(out.tab,cx)
}

surv_CHESS <- cbind(posdata,out.tab)
surv_CHESS <- surv_CHESS %>% select(-term)

write_csv(surv_CHESS,paste(outputdir,"/survival_TUCRs_CHESS.csv",sep=""))

```

## Generate CHESS gene RPKMs for filtering

```{r RPKM_CHESS, results="hide", eval = FALSE}

CHESS.rpkmcounts <- CHESScounts[,6:ncol(CHESScounts)]
CHESS.rpkmcounts.info <- CHESScounts[,1:5]
rownames(CHESS.rpkmcounts) <- CHESScounts$id

genelength <- CHESScounts$length

seqdepth <- read.csv(file = "seqdepth_counts.csv",row.names=1)

seqdepth <- as.vector(seqdepth$counts)

rpkm <- function(counts,len,dep){
  x <- counts/len
  return(t(t(x)/dep))
}

rpkm.CHESS <- rpkm(CHESS.rpkmcounts,genelength,seqdepth)
rpkm.CHESS2 <- cbind(CHESS.rpkmcounts.info,rpkm.CHESS)

write.csv(rpkm.CHESS2,paste(outputdir,"/rpkm.CHESS.csv",sep=""))
```

## Generate heatmap for all CHESS genes containing TUCRs

```{r heatmap_CHESS, message=FALSE, eval = FALSE}
heatmap.CHESS <- rpkm.CHESS2[,6:ncol(rpkm.CHESS2)]
CHESSids <- rpkm.CHESS2[,5]
means <- apply(heatmap.CHESS,1,mean)

heatmap.CHESS2 <- as.data.frame(apply(heatmap.CHESS,1:2,function(x) {ifelse(x>=10,10,x)}))
heatmap.CHESS2 <- cbind(CHESSids,means,heatmap.CHESS2)

heatmap.CHESS2 <- heatmap.CHESS2 %>% 
  gather(key = "Sample", value = "RPKM",-CHESSids,-means)

p <- ggplot(data = heatmap.CHESS2,mapping = aes(x = Sample,y = reorder(CHESSids,means), fill = RPKM)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("blue", "white", "red"), values = c(0,0.1,1)) +
  theme(panel.grid = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank())

ggsave(p,file=paste(outputdir,"/CHESS_heatmap.png",sep=""))
```

## Visualize Stephen and Alex's FPKMs to compare to my own RPKMs

```{r sdt_heatmap, message=FALSE}
fpkm.sdt <- read_csv("aboundaer_tucr_gbm-survival_fpkms.csv")
heatmap.sdt <- fpkm.sdt[,2:ncol(fpkm.sdt)]
sdtids <- fpkm.sdt[,1]
means <- apply(heatmap.sdt,1,mean)

heatmap.sdt2 <- as.data.frame(apply(heatmap.sdt,1:2,function(x) {ifelse(x>=10,10,x)}))
heatmap.sdt2 <- cbind(sdtids,means,heatmap.sdt2)

heatmap.sdt2 <- heatmap.sdt2 %>% 
  gather(key = "Sample", value = "FPKM",-id,-means)

p <- ggplot(data = heatmap.sdt2,mapping = aes(x = Sample,y = reorder(id,means), fill = FPKM)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("blue", "white", "red"), values = c(0,0.1,1)) +
  theme(panel.grid = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank())

ggsave(p,file=paste(outputdir,"/sdt_heatmap.png",sep=""))
```

```{r TUCRintrons}
library(tidyverse)

tucr.introns <- read.table("TUCRintrons.bed",header=FALSE)

i = 1

df <- as.data.frame(matrix(nrow=0,ncol=4))

for(i in 1:length(tucr.introns$V1)){
  print(i)
  data <- tucr.introns[i,]
  chrom <- data$V1
  start <- data$V2
  end <- data$V3
  TUCR <- data$V4
  d <- seq(start,end)
  breaks <- as.matrix(split(d, ceiling(seq_along(d)/50)))
  breaks <- as.data.frame(matrix(unlist(breaks), nrow=length(breaks), byrow=TRUE))
  colnames(breaks)[ncol(breaks)] <- "last"
  breaks$row_num <- seq.int(nrow(breaks))
  breaks$TUCR <- TUCR
  breaks$chrom  <- chrom
  breaks_rbind <- breaks %>%
    mutate(TUCR_num = paste(TUCR,row_num,sep="_")) %>%
    dplyr::select(chrom,"start"=V1,"end"=last,TUCR_num)
  df <- rbind(df,breaks_rbind) %>%
    mutate(diff = end - start) %>%
    filter(diff >0) %>%
    select(-diff)
}

write_delim(df,file="TUCRintervals.bed",delim="\t")

```

<<<<<<< HEAD
# Section 5: TUCRs are contained within mRNA or lncRNA transcripts

## Results 

Of the 481 TUCRs, 186 were previously characterized as intergenic, or not being contained within a known coding or non-coding transcript.  We chose to focus on these TUCRs, as they potentially represent a partial or full novel long non-coding RNA (lncRNA) transcripts.  To identify the RNA transcript(s) that contain these TUCRs, we performed a de novo transcript reassembly using GBM RNA-Seq data from TCGA.  Using this method, we identified 2636 transcripts that contain TUCRs.  We prioritized transcripts that did not overlap with known genes, as these are likely to represent novel genes.  Transcripts that were greater than 10,000 nucleotide (nt) were also eliminated.  Using these criteria, 8 transcripts were identified in GBM, LGG, or both disease types. (Figure 7A, Table 1)  These transcripts were submitted to the Coding Potential Assessing Tool (CPAT) to determine whether they express any coding potential.  Of the 8 transcripts, five represented no coding potential, while three contained coding potential.  In particular, one of these transcripts with coding potential (TUC110-) overlaps with the most highly upregulated TUCR in both LGG and GBM (uc.110, Figures 7B and 7C).

## Methodology

### Stringtie GBM

The hg19 genomic coordinates for each TUCR were acquired from _____________ and were lifted over to hg38 coordinates. The stringtie bioinformatics package was then used to perform a de novo transcript reassembly of all transcripts in 161 GBM, 400 LGG, and 5 normal brain TCGA Cancer RNA-Seq datasets.  First, transcripts present in each RNA-Seq BAM file were compiled into individual BED/GTF format genomic coordinate files using Stringtie.  These files were then merged into one master file containing all transcripts in all samples using Stringtie and an hg38 genome reference file.  The bedtools package was then used to find the genomic coordinates of all transcripts containing ultraconserved regions.  The ________________ group used stringtie to analyze the entirety of the GTex RNA-Seq normal tissue catalog, so bedtools was also used to determine the intersection of TUCRs with their results.  This methodology provides insight into TUCR transcript presence in normal tissues and in brain cancer.

```{bash stringtie, eval = FALSE}
wget ftp://ftp.ensembl.org/pub/release-99/gtf/homo_sapiens/Homo_sapiens.GRCh38.99.gtf.gz
mv Homo_sapiens.GRCh38.99.gtf.gz stringtie
gunzip stringtie/Homo_sapiens.GRCh38.99.gtf.gz

module load gcc/7.1.0
module load stringtie/2.0.6

for bam in $(find GBM-RNASEQ-RAW/* -name '*.bam')
do
name=$(echo $bam | awk -F ".bam" '{print $1}')
name2=$(basename "$name")
if [ ! -f stringtie/$name2.gtf ] 
then
    echo "$name2.gtf does not exist"
    stringtie $bam -o $name.gtf
    mv $name.gtf ./stringtie
else
    echo "$name2.gtf exists... skipping"
fi

if [ ! -f $name.gtf ]; then
  stringtie $bam -o $name.gtf
fi
mv $name.gtf ./stringtie
done

cd stringtie

stringtie --merge *rehead.gtf -G Homo_sapiens.GRCh38.99.gtf -o stringtie_merged.gtf

intersectBed -a hg38.ultraConserved.bed -b CHESSgenes.bed -wa -wb 
  > intersectchess_TUCRs.bed

intersectBed -a hg38.ultraConserved.bed -b stringtie_merged.gtf -wa -wb 
  > intersectstringtie_TUCRs.bed

cp stringtie_merged.gtf stringtie_merged.bed
```

### Stringtie LGG

```{bash stringtie lgg, eval = FALSE}
wget ftp://ftp.ensembl.org/pub/release-99/gtf/homo_sapiens/Homo_sapiens.GRCh38.99.gtf.gz
mv Homo_sapiens.GRCh38.99.gtf.gz stringtie
gunzip stringtie/Homo_sapiens.GRCh38.99.gtf.gz

module load gcc/7.1.0
module load stringtie/2.0.6

for bam in $(find LGG-RNASEQ-RAW/* -name '*.bam')
do
name=$(echo $bam | awk -F ".bam" '{print $1}')
name2=$(basename "$name")
if [ ! -f stringtie/$name2.gtf ] 
then
    echo "$name2.gtf does not exist"
    stringtie $bam -o $name.gtf
    mv $name.gtf ./stringtie
else
    echo "$name2.gtf exists... skipping"
fi

if [ ! -f $name.gtf ]; then
  stringtie $bam -o $name.gtf
fi
mv $name.gtf ./stringtie
done

cd stringtie

stringtie --merge *rehead.gtf -G Homo_sapiens.GRCh38.99.gtf -o stringtie_merged.gtf

intersectBed -a TUCR_intergenic.bed -b stringtie_merged.gtf -wa -wb > intersectintergenic_TUCRs.bed

intersectBed -a intersectintergenic_TUCRs.bed -b stringtie_merged.gtf -wa -wb 
  > intersectintergenic_TUCRs.bed

cp stringtie_merged.gtf stringtie_merged.bed
```

## Show that TUCRs are contained within novel genes

```{r novelgenes, eval = FALSE}
lgg_novelgenes <- read_delim("lgg_novelgenes.bed", col_names = FALSE, delim = "\t",skip_empty_rows = FALSE) %>%     dplyr::select(X5,X7,X9,X10,X11,X13) %>%
  filter(X9 == "transcript") %>%
  mutate(disease = "LGG",name = paste("TUC",str_remove(X5,"uc."),X13,sep = ""), length = X11-X10) %>%
  filter(length <= 30000) %>%
  group_by(name) %>%
  mutate(max = max(length,na.rm=TRUE)) %>%
  filter(length == max) %>%
  ungroup() %>%
  dplyr::select(name,disease,length,"chromosome" = X7,"start" = X10,"end" = X11, "strand" = X13)

gbm_novelgenes <- read_delim("gbm_novelgenes.bed", col_names = FALSE, delim = "\t",skip_empty_rows = FALSE) %>%     dplyr::select(X5,X7,X9,X10,X11,X13) %>%
  filter(X9 == "transcript") %>%
  mutate(disease = "GBM",name = paste("TUC",str_remove(X5,"uc."),X13,sep = ""), length = X11-X10) %>%
  filter(length <= 30000) %>%
  group_by(name) %>%
  mutate(max = max(length,na.rm=TRUE)) %>%
  filter(length == max) %>%
  ungroup() %>%
  dplyr::select(name,disease,length,"chromosome" = X7,"start" = X10,"end" = X11, "strand" = X13)

all_novelgenes <- rbind(gbm_novelgenes,lgg_novelgenes) 
all_novelgenes <- all_novelgenes %>%
  group_by(name) %>%
  mutate(n = n(), annot = ifelse(n==2, "GBM/LGG",disease)) %>%
  dplyr::select(-n)


  

write.csv(all_novelgenes,file="all_novelgenes.csv")

CPAT_bedfile <- all_novelgenes %>%
  dplyr::select(chromosome,start,end,strand,name)

write_delim(CPAT_bedfile,file="CPAT.bed",delim = "\t",col_name = FALSE)

```

### Get Fasta

```{bash novelgenes_getFASTA}

```

### Use CPAT Web Browser

### Integrate CPAT results

```{r}
all_novelgenes <- read.csv(file="all_novelgenes.csv",header=TRUE,row.names=1)

CPAT_results <- read_delim("result.txt", col_names = TRUE, delim = "\t",skip_empty_rows = FALSE) %>%
  dplyr::select("length" = `RNA size`, `ORF size`,`Ficket Score`,`Hexamer Score`,`Coding Probability`,`Coding Label`)

CPAT_novelgenes <- all_novelgenes %>%
  full_join(CPAT_results, by = "length")

write.csv(CPAT_novelgenes,file="CPAT_novelgenes.csv")

```


### Summarizing Novel Gene data

``` {r sum_novelgenes}

CPAT_novelgenes <- read.csv(file="CPAT_novelgenes.csv",header=TRUE,row.names=1)

CPAT_novelgenes_sum <- CPAT_novelgenes %>%
  group_by(annot,Coding.Label) %>% 
  mutate(group_id = cur_group_id(),count = n(),Group = 
    ifelse(group_id==1, "GBM_non_coding",
    ifelse(group_id==2, "GBM_coding",
    ifelse(group_id==3, "GBM/LGG_non_coding",
    ifelse(group_id==4, "GBM/LGG_coding",
    ifelse(group_id==5, "LGG_non_coding",
    ifelse(group_id==6, "LGG_coding","ERROR"))))))) %>%
  group_by(Group,group_id) %>%
  summarize(count=n())

p <- ggplot(CPAT_novelgenes_sum, aes(x=reorder(Group,group_id),y=count,fill=Group)) + geom_bar(stat="identity", width = 0.7, color="black") +
  geom_text(aes(label=round(count,2)), vjust=1.6, color="black", size=rel(4)) + ggtitle(paste("Some intergenic TUCRs are \n contained within Novel Genes")) + xlab("Disease") + ylab("Number of Genes") + theme(
plot.title = element_text(size=rel(1.5), face="bold",hjust = 0.5),
axis.title = element_text(size = rel(1.25), face="bold"),
axis.text.x = element_text(angle = -90, size = 10),
legend.title = element_blank(),
legend.position = "none",
panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black"))

ggsave(file=paste(outputdir,"/tucr_novelgenes.png",sep=""),plot = print(p), width = 5, height = 5, dpi = 300) 
  

```

## Checking for miRNA binding to TUCRs

```{r}

if(!dir.exists(paste(outputdir,"/miRNA/",sep=""))){
  dir.create(paste(outputdir,"/miRNA/",sep=""))
}

tucr_sequence <- read.csv("tucr_sequence.csv",header=TRUE)

mirna_seeds <- read.csv("mirna-seeds.csv",header=TRUE)

tucrannot <- read.csv("tucr_annot.csv",header=TRUE) %>%
  dplyr::select("id" = row, annot)

uc.110 <- tucr_sequence %>%
  mutate(i=1) %>%
  full_join(mutate(mirna_seeds, i=1), by="i") %>%
  dplyr::select(-i) %>%
  filter(stringr::str_detect(sequence, as.character(revcomp))) %>%
  left_join(tucrannot,by="id") %>%
  distinct() %>%
  filter(id == "uc.110-" | id == "uc.110+") %>%
  write_csv(paste(outputdir,"/miRNA/uc110_mirna.csv",sep=""))

uc.454 <- tucr_sequence %>%
  mutate(i=1) %>%
  full_join(mutate(mirna_seeds, i=1), by="i") %>%
  dplyr::select(-i) %>%
  filter(stringr::str_detect(sequence, as.character(revcomp))) %>%
  left_join(tucrannot,by="id") %>%
  distinct() %>%
  filter(id == "uc.454-" | id == "uc.454+") %>%
  write_csv(paste(outputdir,"/miRNA/uc454_mirna.csv",sep=""))

tucr_sequence_plus <- tucr_sequence[1:481,]

tucr_mirna_plus <- tucr_sequence_plus %>%
mutate(i=1) %>%
mutate(id = gsub("\\+","",id)) %>%
full_join(mutate(mirna_seeds, i=1), by="i") %>%
dplyr::select(-i) %>%
filter(stringr::str_detect(sequence, as.character(revcomp))) %>%
left_join(tucrannot,by="id") %>%
distinct()

miRNA_summary_plus <- tucr_mirna_plus %>%
  group_by(mirna) %>%
  summarise(n = n()) %>%
  arrange(desc(n)) %>%
  write_csv(paste(outputdir,"/miRNA/mirna_summary_plus.csv",sep=""))

TUCR_summary_plus <- tucr_mirna_plus %>%
  group_by(id) %>%
  summarise(n = n()) %>%
  mutate(id = gsub("\\+","",id)) %>%
  left_join(tucrannot,by="id") %>%
  arrange(desc(n)) %>%
  write_csv(paste(outputdir,"/miRNA/tucr_summary_plus.csv",sep=""))

annot_summary_plus <- tucr_mirna_plus %>%
  dplyr::select(annot,mirna) %>%
  group_by(annot) %>%
  distinct() %>%
  summarise(n = n())

p <- ggplot(annot_summary_plus, aes(x=annot,y=n,fill=annot))          + 
  geom_bar(
    stat="identity",colour="black", 
    width = 1)                                                          +
  geom_text(
    aes(label = n), 
    color = "black")                                                    + 
  scale_fill_manual(values=c("#F8766D","#00BFC4","#7CAE00","#C77CFF"))  + 
  ggtitle(paste("Plus strand TUCRs are targeted by miRNAs"))            + 
  theme(
    plot.title = element_text(size=rel(1.0),
    face="bold",hjust = 0.5),
    axis.line=element_blank(),
    panel.grid.major = element_blank(), 
    panel.grid.minor = element_blank(),
    axis.ticks=element_blank(),
    panel.background = element_blank())

p

ggsave(file=paste(outputdir,"/miRNA/tucr_annotation_mirna_plus.png",sep=""),
       plot = print(p), 
       width = 5, 
       height = 5, 
       dpi = 300)

tucr_sequence_minus <- tucr_sequence[482:962,]

tucr_mirna_minus <- tucr_sequence_minus %>%
mutate(i=1) %>%
mutate(id = gsub("\\-","",id)) %>%
full_join(mutate(mirna_seeds, i=1), by="i") %>%
dplyr::select(-i) %>%
filter(stringr::str_detect(sequence, as.character(revcomp))) %>%
left_join(tucrannot,by="id") %>%
distinct()

miRNA_summary_minus <- tucr_mirna_minus %>%
  group_by(mirna) %>%
  summarise(n = n()) %>%
  arrange(desc(n)) %>%
  write_csv(paste(outputdir,"/miRNA/mirna_summary_minus.csv",sep=""))

TUCR_summary_minus <- tucr_mirna_minus %>%
  group_by(id) %>%
  summarise(n = n()) %>%
  mutate(id = gsub("\\+","",id)) %>%
  left_join(tucrannot,by="id") %>%
  arrange(desc(n)) %>%
  write_csv(paste(outputdir,"/miRNA/tucr_summary_minus.csv",sep=""))

annot_summary_minus <- tucr_mirna_minus %>%
  dplyr::select(annot,mirna) %>%
  group_by(annot) %>%
  distinct() %>%
  summarise(n = n())

p <- ggplot(annot_summary_minus, aes(x=annot,y=n,fill=annot))          + 
  geom_bar(
    stat="identity",colour="black", 
    width = 1)                                                          +
  geom_text(
    aes(label = n), 
    color = "black")                                                    + 
  scale_fill_manual(values=c("#F8766D","#00BFC4","#7CAE00","#C77CFF"))  + 
  ggtitle(paste("Minus strand TUCRs are targeted by miRNAs"))            + 
  theme(
    plot.title = element_text(size=rel(1.0),
    face="bold",hjust = 0.5),
    axis.line=element_blank(),
    panel.grid.major = element_blank(), 
    panel.grid.minor = element_blank(),
    axis.ticks=element_blank(),
    panel.background = element_blank())

p

ggsave(file=paste(outputdir,"/miRNA/tucr_annotation_mirna_minus.png",sep=""),
       plot = print(p), 
       width = 5, 
       height = 5, 
       dpi = 300)

```

# Making Figures for cell counting assay

```{r cellcounting}

if(!dir.exists(paste(outputdir,"/cellcountingdata/",sep=""))){
  dir.create(paste(outputdir,"/cellcountingdata/",sep=""))
}

cellcountdata <- read.csv("cellcountingdata.csv",header=TRUE)

cellcountdata$mean <- rowMeans(cellcountdata[,5:12],na.rm = TRUE)

cellcountdata <- cellcountdata %>%
  gather(key = "Replicate", value = "FoldChange",-Cell.Line,-Experiment,-Sample,-Day,-mean)

experimentlabels <- unique(as.character(cellcountdata$Experiment))

i = 1

for(i in 1:length(cellcountdata$Experiment)){
  print(i)
  experimentlabel <- experimentlabels[i]
  print(experimentlabel)

cellcountdatafiltered <- cellcountdata %>%
  filter(Experiment == experimentlabel)

cellcountdatafiltered <- cellcountdatafiltered[complete.cases(cellcountdatafiltered),]

p<- ggplot(cellcountdatafiltered, aes(x=Day, y=FoldChange,col=Sample)) +
  #geom_line(aes(x=Day,y=mean,col=Sample)) +
  #geom_point(aes(x=Day,y=mean,col=Sample),size=2) +
  geom_bar(stat="identity",aes(x=Day,y=mean,fill=Sample),position="dodge") +
  geom_jitter(binaxis='y', stackdir='center', stackratio=0.90, dotsize=0.30) +
  scale_fill_brewer(palette="Dark2") +
  scale_color_brewer(palette="Dark2") +
  scale_x_continuous(breaks = c(0,1,3,5,7))+ 
  ggtitle(paste(experimentlabel," of uc.110",sep="")) +
  theme(plot.title = element_text(hjust = 0.5),
  panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
  panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  labs(y="Fold Change", x = "Days after T-zero") + 
  facet_wrap(~Cell.Line)

ggsave(p,file=paste(outputdir,"/cellcountingdata/",experimentlabel,".png",sep = ""),width=12,height=8,dpi=300)}

```

