---
title: "Old Code Chunks"
author: "Myron K Gibert Jr"
date: "7/12/2020"
output: pdf_document
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{bash ATAC, eval = FALSE}

wget ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3669nnn/GSM3669999/suppl/GSM3669999_GSC23_ATAC.highPValue.distal_intersect.H3K27Ac.bed.gz

gunzip GSM3669999_GSC23_ATAC.highPValue.distal_intersect.H3K27Ac.bed.gz

bedtools shuffle -i GSM3669999_GSC23_ATAC.highPValue.distal_intersect.H3K27Ac.bed -g hg38_genomeFile.txt > GSC23_ATAC.shuffle.bed

bedtools sort -i GSM3669999_GSC23_ATAC.highPValue.distal_intersect.H3K27Ac.bed > GSC23_ATAC.sort.bed

bedtools sort -i GSC23_ATAC.shuffle.bed > GSC23_ATAC.shuffle.sort.bed

bedtools fisher -a tucrs_pm10kb.bed -b GSC23_ATAC.shuffle.sort.bed -g hg38_genomeFile.txt > GSC23_ATAC_random_output.txt

bedtools fisher -a tucrs_pm10kb.bed -b GSC23_ATAC.sort.bed -g hg38_genomeFile.txt > GSC23_ATAC_tucrs_output.txt



wget ftp://ftp.ncbi.nlm.nih.gov/geo/samples/GSM3669nnn/GSM3669993/suppl/GSM3669993_U87_ATAC.highPValue.distal_intersect.H3K27Ac.bed.gz

gunzip GSM3669993_U87_ATAC.highPValue.distal_intersect.H3K27Ac.bed.gz

bedtools shuffle -i GSM3669993_U87_ATAC.highPValue.distal_intersect.H3K27Ac.bed -g hg38_genomeFile.txt > U87_ATAC.shuffle.bed

bedtools sort -i GSM3669993_U87_ATAC.highPValue.distal_intersect.H3K27Ac.bed > U87_ATAC.sort.bed

bedtools sort -i U87_ATAC.shuffle.bed > U87_ATAC.shuffle.sort.bed

bedtools fisher -a tucrs_pm10kb.bed -b U87_ATAC.shuffle.sort.bed -g hg38_genomeFile.txt > U87_ATAC_random_output.txt

bedtools fisher -a tucrs_pm10kb.bed -b U87_ATAC.sort.bed -g hg38_genomeFile.txt > U87_ATAC_tucrs_output.txt

```

```{r}
tidy_heatmap(heatmap.allgo2,
             rows = gene,
             columns = Term,
             values = direction,
             scale = "row",
             #annotation_col = c(sample_type, condition, group),
             #annotation_row = c(is_immune_gene, direction),
             #gaps_row = direction,
             #gaps_col = group
)

tidy_heatmap(df,
             rows = external_gene_name,
             columns = sample,
             values = expression,
             scale = "row",
             #annotation_col = c(sample_type, condition, group),
             #annotation_row = c(is_immune_gene, direction),
             #gaps_row = direction,
             #gaps_col = group
)

class1 <- lapply(data_exprs,class)
class2 <- lapply(heatmap.allgo2,class)
class2 <- lapply(df,class)

write.csv(data_exprs,"datatest1.csv")

df <- read.csv("datatest1.csv",header=TRUE)

write.csv(heatmap.allgo2,"datatest2.csv")
```


```{r}

heatmap.allgo.up <- heatmap.allgo2 %>%
  filter(sum >0) %>%
  group_by(Term,sum) %>%
  summarise()

colors <- colorRampPalette(c("blue","white","red"))(length(unique(heatmap.allgo2$direction)))
heatmap.allgo2$direction <- factor(heatmap.allgo2$direction)
N <- nlevels(heatmap.allgo2$direction)

p <- ggplot(data = heatmap.allgo2,mapping = aes(x = reorder(Term,sum),y = gene, fill = direction)) +
  geom_tile() +
  scale_fill_manual(values = colors,breaks=levels(heatmap.allgo2$direction)[seq(1, N, by=1)]) + 
  ggtitle("Title") + 
  xlab("X-axis") + 
  ylab("Y-axis") +
  theme(plot.title = element_text(hjust = 0.5),
        panel.grid = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank())

ggsave(p,file=paste(outputdir,"/GOTerms_heatmap.png",sep=""), width = 2.5, height = 7, dpi = 300)

```

## Acquiring SNPs

```{bash, eval = FALSE}
mkdir SNPs

cd SNPs

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_1.bed.gz
gunzip bed_chr_1.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_10.bed.gz
gunzip bed_chr_10.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_11.bed.gz
gunzip bed_chr_11.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_13.bed.gz
gunzip bed_chr_13.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_12.bed.gz
gunzip bed_chr_12.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_14.bed.gz
gunzip bed_chr_14.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_15.bed.gz
gunzip bed_chr_15.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_16.bed.gz
gunzip bed_chr_16.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_17.bed.gz
gunzip bed_chr_17.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_18.bed.gz
gunzip bed_chr_18.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_19.bed.gz
gunzip bed_chr_19.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_2.bed.gz
gunzip bed_chr_2.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_20.bed.gz
gunzip bed_chr_20.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_21.bed.gz
gunzip bed_chr_21.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_22.bed.gz
gunzip bed_chr_22.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_3.bed.gz
gunzip bed_chr_3.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_4.bed.gz
gunzip bed_chr_4.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_5.bed.gz
gunzip bed_chr_5.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_6.bed.gz
gunzip bed_chr_6.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_7.bed.gz
gunzip bed_chr_7.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_8.bed.gz
gunzip bed_chr_8.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_9.bed.gz
gunzip bed_chr_9.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_X.bed.gz
gunzip bed_chr_X.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_Y.bed.gz
gunzip bed_chr_Y.bed.gz

```

### Remove first line

``` {r SNPgetter}
for bed in bed_chr_*
do

tail -n +2 $bed > $bed.stdout
mv $bed.stdout $bed

done

module load gcc/7.1.0
module load bedtools/2.26.0

for bed in bed_chr_*
do
echo $bed
intersectBed -a hg38.ultraConserved.bed -b $bed -wa -wb > $bed.txt
done

cat *.txt > merged.txt
```

### 

``` {r SNPreader}

library("tidyverse")

SNPcounts <- read.csv("TUCR_SNPs.csv")

```

## Make figures for fold change

```{r overexp_TUCR}

overexp_TUCR_cp <- c("#C77CFF","#F8766D","#619CFF","#00BA38","#FFFF00")

filename <- "tucroverexpression_U87.csv"
x <- read.csv(filename)

p <- ggplot(x,aes(x = Target, y = FC, fill = Sample)) + 
  labs(title = "TUCR Overexpression in U87 cells") +
  ylab("Fold Change") +
  geom_bar(stat = "identity",position="dodge",colour="black") + 
  theme(axis.text.x = element_text(angle = -90, size = 10),
        axis.text.y = element_text(size = 10),
        axis.ticks.x = element_blank(),
        plot.title = element_text(size = 15, face = "bold",hjust = 0.5),
        axis.title.y = element_text("Fold Change",size = 10)) +
  facet_zoom(ylim = c(0, 2)) +
  scale_fill_manual(values=overexp_TUCR_cp)

ggsave(p,file=paste(outputdir,"/",filename,".png",sep=""), width = 7.5, height = 5, dpi = 300)


```

## Generate figures for cell fractionation

```{r cellfrac_TUCR}

if(!dir.exists(paste(outputdir,"/cellfrac",sep=""))){
  dir.create(paste(outputdir,"/cellfrac",sep=""))
}

CellFrackin <- function(filename){
all_content = readLines(filename)
skip_second = all_content[-c(1:19)]
x <- read.csv(textConnection(skip_second), header = TRUE, stringsAsFactors = FALSE)
samplenames <- t(as.data.frame(strsplit(as.character(x$Sample), " ")))
colnames(samplenames) <- c("Line","Fraction")
x <- cbind(x,samplenames)
x<- x %>%
    dplyr::select(Well,Target,Sample,Line,Fraction,Cq) %>%
    dplyr::group_by(Target,Line,Fraction) %>%
    dplyr::summarize(meanCq = mean(Cq)) %>%
    dplyr::mutate(dct = meanCq - min(meanCq)) %>%
    dplyr::mutate(FoldChange = 2^(-dct)) %>%
    dplyr::mutate(ID = paste(Target,Line,Fraction,sep=" "))

p <- ggplot(x,aes(x = Target, y = FoldChange, fill = Fraction)) + 
  labs(title = "Cellular localization of TUCRs in GBM cell lines") +
  ylab("Fold Change") +
  geom_bar(stat = "identity",position="dodge") + 
  theme(axis.text.x = element_text(angle = -90, size = 10),
        axis.text.y = element_text(size = 10),
        axis.ticks.x = element_blank(),
        plot.title = element_text(size = 15, face = "bold",hjust = 0.5),
        axis.title.y = element_text("Fold Change",size = 10),
panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  facet_wrap(~Line)

ggsave(p,file=paste(outputdir,"/",filename,".png",sep=""), width = 7.5, height = 5, dpi = 300)
x %>%
  select(-ID) %>%
  write_csv(paste(outputdir,"/",filename," (Output).csv",sep=""))
}

path <- list.files("cellfrac")
i = 1

for(i in 1:length(path)){
  print(i)
  CellFrackin(paste("cellfrac",path[i],sep="/"))
}

```

## Summarize GO Terms (old)

```{r summary GO Terms old}

allgo <- read.csv(paste(outputdir,"/GOterm_all_summary_TUCRs.csv",sep=""),header=TRUE)

allgo_bp <- read.csv(paste(outputdir,"/GOterm_bp_summary_TUCRs.csv",sep=""),header=TRUE)

allgo_mf <- read.csv(paste(outputdir,"/GOterm_mf_summary_TUCRs.csv",sep=""),header=TRUE)

allgo_cc <- read.csv(paste(outputdir,"/GOterm_cc_summary_TUCRs.csv",sep=""),header=TRUE)

allgo_sum <- allgo %>% 
  dplyr::group_by(Term) %>%
  dplyr::summarise(count = n()) %>%
  dplyr::arrange(desc(count)) %>%
  dplyr::mutate(Term=factor(Term, levels=Term))

allgo_sum <- allgo_sum[1:20,]

p <- ggplot(allgo_sum, aes(x=Term,y=count,fill=Term)) + geom_bar(stat="identity", width = 0.7) + geom_text(aes(label=round(count,2)), vjust=1.6, color="black", size=rel(2)) + ggtitle(paste("Most common TUCR GO Terms in ",disease,sep="")) +xlab("GO Term") + ylab("# of TUCRs") + 
  theme(
  plot.title = element_text(size=rel(1.5), face="bold",hjust = 0.5),
axis.title = element_text(size = rel(1.25), face="bold"),
legend.title = element_blank(),
legend.position = "none",
axis.text.x = element_text(size=rel(1.5)),
axis.text.y = element_text(size=rel(1.5)),
panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black")) + coord_flip()

ggsave(file=paste(outputdir,"/tucr_allgo.png",sep=""),plot = print(p), width = 15, height = 10, dpi = 300) 


allgo_bp_sum <- allgo_bp %>% 
  dplyr::group_by(Term) %>%
  dplyr::summarise(count = n()) %>%
  dplyr::arrange(desc(count)) %>%
  dplyr::mutate(Term=factor(Term, levels=Term))

allgo_bp_sum <- allgo_bp_sum[1:20,]

p1 <- ggplot(allgo_bp_sum, aes(x=Term,y=count,fill=Term)) + geom_bar(stat="identity", width = 0.7) + geom_text(aes(label=round(count,2)), vjust=1.6, color="black", size=rel(2)) + ggtitle(paste("Most common TUCR Biological Processes in ",disease,sep="")) + xlab("Biological Processes") + ylab("# of TUCRs") + theme(
  plot.title = element_text(size=rel(1.5), face="bold",hjust = 0.5),
axis.title = element_text(size = rel(1.25), face="bold"),
legend.title = element_blank(),
legend.position = "none",
axis.text.x = element_text(size=rel(1.5)),
axis.text.y = element_text(size=rel(1.5)),
panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black")) + coord_flip()

ggsave(file=paste(outputdir,"/tucr_allgo_bp.png",sep=""),plot = print(p1), width = 15, height = 10, dpi = 300) 

allgo_mf_sum <- allgo_mf %>% 
  dplyr::group_by(Term) %>%
  dplyr::summarise(count = n()) %>%
  dplyr::arrange(desc(count)) %>%
  dplyr::mutate(Term=factor(Term, levels=Term))

allgo_mf_sum <- allgo_mf_sum[1:20,]

p2 <- ggplot(allgo_mf_sum, aes(x=Term,y=count,fill=Term)) + geom_bar(stat="identity", width = 0.7) + geom_text(aes(label=round(count,2)), vjust=1.6, color="black", size=rel(2)) + ggtitle(paste("Most common TUCR Molecular Functions in ",disease,sep="")) + xlab("Molecular Function") + ylab("# of TUCRs") + theme(
  plot.title = element_text(size=rel(1.5), face="bold",hjust = 0.5),
axis.title = element_text(size = rel(1.25), face="bold"),
legend.title = element_blank(),
legend.position = "none",
axis.text.x = element_text(size=rel(1.5)),
axis.text.y = element_text(size=rel(1.5)),
panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black")) + coord_flip()

ggsave(file=paste(outputdir,"/tucr_allgo_mf.png",sep=""),plot = print(p2), width = 15, height = 10, dpi = 300) 

allgo_cc_sum <- allgo_cc %>% 
  dplyr::group_by(Term) %>%
  dplyr::summarise(count = n()) %>%
  dplyr::arrange(desc(count)) %>%
  dplyr::mutate(Term=factor(Term, levels=Term))

allgo_cc_sum <- allgo_cc_sum[1:20,]

p3 <- ggplot(allgo_cc_sum, aes(x=Term,y=count,fill=Term)) + geom_bar(stat="identity", width = 0.7) + geom_text(aes(label=round(count,2)), vjust=1.6, color="black", size=rel(2)) + ggtitle(paste("Most common TUCR Cellular Compartments in ",disease,sep="")) + xlab("Cellular Compartment") + ylab("# of TUCRs") + theme(
  plot.title = element_text(size=rel(1.5), face="bold",hjust = 0.5),
axis.title = element_text(size = rel(1.25), face="bold"),
legend.title = element_blank(),
legend.position = "none",
axis.text.x = element_text(size=rel(1.5)),
axis.text.y = element_text(size=rel(1.5)),
panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black")) + coord_flip()

ggsave(file=paste(outputdir,"/tucr_allgo_cc.png",sep=""),plot = print(p3), width = 15, height = 10, dpi = 300) 

allgo_bp_sum <- allgo_bp_sum %>%
  dplyr::mutate(Ontology = "Biological Process")

allgo_mf_sum <- allgo_mf_sum %>%
  dplyr::mutate(Ontology = "Molecular Function")

allgo_cc_sum <- allgo_cc_sum %>%
  dplyr::mutate(Ontology = "Cellular Compartment")

allgo_fig <- rbind(allgo_bp_sum,allgo_mf_sum)

p4 <- ggplot(allgo_fig, aes(x=Term,y=count,fill=Term)) + geom_bar(stat="identity",position="dodge", width = 0.7) + geom_text(aes(label=round(count,2)), vjust=1.6, color="black", size=rel(4)) + ggtitle(paste("Most common TUCR GO Terms in ",disease,sep="")) + xlab("GO Term") + ylab("# of TUCRs") + facet_grid(~ Ontology,scales="free",space="free") + theme(
  plot.title = element_text(size=rel(1.75), face="bold",hjust = 0.5),
axis.title = element_text(size = rel(1.25), face="bold"),
legend.title = element_blank(),
legend.position = "none",
axis.text.x = element_text(size=rel(1.5)),
axis.text.y = element_text(size=rel(1.5)),
strip.text = element_text(size=rel(1.5)),
panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black")) 

ggsave(file=paste(outputdir,"/tucr_allgo_fig.png",sep=""),plot = print(p4), width = 15, height = 15, dpi = 300)


```

## Generate heatmap for my TUCRs of interest

```{r heatmap_Myron}
heatmap.TUCRs <- rpkm.TUCRs2[,6:ncol(rpkm.TUCRs2)]
TUCRids <- rpkm.TUCRs2[,4]
means <- apply(heatmap.TUCRs,1,mean)

heatmap.TUCRs2 <- as.data.frame(apply(heatmap.TUCRs,1:2,function(x) {ifelse(x>=10,10,x)}))
heatmap.TUCRs2 <- cbind(TUCRids,means,heatmap.TUCRs2)

heatmap.TUCRs2 <- heatmap.TUCRs2 %>% 
  gather(key = "Sample", value = "RPKM",-TUCRids,-means) %>% 
  filter(TUCRids == "uc.62"|TUCRids == "uc.110"|TUCRids == "uc.193"|TUCRids == "uc.210"|TUCRids == "uc.427"|TUCRids == "uc.212"|TUCRids == "uc.341")

p <- ggplot(data = heatmap.TUCRs2,mapping = aes(x = Sample,y = reorder(TUCRids,means), fill = RPKM)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("blue", "white", "red"), values = c(0,0.1,1)) +
  theme(panel.grid = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks = element_blank())

ggsave(p,file=paste(outputdir,"/Myron_heatmap.png",sep=""))
```

## Identifying differentially expressed stringtie transcripts containing TUCRs with DESeq2

```{r DESeq2_stringtie, results="hide", message = FALSE, eval = FALSE}

stringtiecounts <- read.table("stringtie_TUCR_counts.txt",header = TRUE)
stringtiecounts <- distinct(stringtiecounts)
metadata <- read_csv(file = metadatafile)

stringtiecounts2 <- stringtiecounts[,5:ncol(stringtiecounts)]
rownames(stringtiecounts2) <- stringtiecounts[,4]
stringtiecounts.info <- stringtiecounts[,1:4]
stringtiecounts.info$length <- (stringtiecounts$end - stringtiecounts$start)/1000
stringtiecounts <- cbind(stringtiecounts.info,stringtiecounts2)

dds_stringtie <- DESeqDataSetFromMatrix(countData = stringtiecounts2,
                              colData = metadata,
                              design = ~dex)
dds_stringtie <- DESeq(dds_stringtie)
res_stringtie <- results(dds_stringtie, tidy=TRUE)
res_stringtie <- as_tibble(res_stringtie)

write.csv(res_stringtie, file = paste(outputdir,"/results_stringtie.csv",sep=""))

```

## Completing survival analysis for stringtie transcripts containing TUCRs

```{r surcor_stringtie, message = FALSE,eval = FALSE}
if(!file.exists("GBM.clin.merged.txt")){unzip("GBM.clin.merged.zip")}

countdata <- read.table("stringtie_TUCR_counts.txt",header = TRUE)
metadata <- read_csv(file = metadatafile)

posdata <- countdata[,1:4]
countdata <- countdata[,5:length(colnames(countdata))]

countdata <- as.matrix(countdata)


n_index <- c(seq(36,39),73)
t_index <- c(seq(1,35),seq(40,72),seq(74,ncol(countdata)))

vm <- function(x){
  cond <- factor(ifelse(seq(1,dim(x)[2],1) %in% t_index, 1,  0))
  d <- model.matrix(~1+cond)
  x <- t(apply(x,1,as.numeric))
  ex <- voom(x,d,plot=F)
  return(ex$E)
}

count_vm <- vm(countdata)
colnames(count_vm) <- metadata$barcode


scal <- function(x,y){
  mean_n <- rowMeans(y)  # mean of normal
  sd_n <- apply(y,1,sd)  # SD of normal
  # z score as (value - mean normal)/SD normal
  res <- matrix(nrow=nrow(x), ncol=ncol(x))
  colnames(res) <- colnames(x)
  rownames(res) <- rownames(x)
  for(i in 1:dim(x)[1]){
    for(j in 1:dim(x)[2]){
      res[i,j] <- (x[i,j]-mean_n[i])/sd_n[i]
    }
  }
  return(res)
}

z_rna <- scal(count_vm[,t_index],count_vm[,n_index])
rownames(z_rna) <- posdata[,4]

clinical <- read.table("GBM.clin.merged.txt",header = TRUE)
clinical$time <- as.numeric(clinical$time)

sum(clinical$patient %in% colnames(z_rna))

ind_tum <- which(unique(colnames(z_rna)) %in% clinical$patient)
ind_clin <- which(clinical$patient %in% colnames(z_rna))

out.tab <- c()
for(x in 1:nrow(count_vm)){
  ind_gene <- x
  s <- Surv(clinical$time[ind_clin],clinical$status[ind_clin])
  cx <- coxph(formula = s ~ count_vm[ind_gene,ind_tum])
  cx <- tidy(cx)
  out.tab <- rbind(out.tab,cx)
}

surv_stringtie <- cbind(posdata,out.tab)
surv_stringtie <- surv_stringtie %>% select(-term)

write_csv(surv_stringtie,paste(outputdir,"/survival_TUCRs_stringtie.csv",sep=""))
```

## Generating volcano plot to visualize DE stringtie transcripts

```{r DESeq2plots_stringtie, message = FALSE, eval = FALSE}

### Volcano plot

volcanoplot <- function (res,title = "Deregulated Genes", output = "results_volcanoplot.png"){
  #res <- res_TUCR
  #genes = c("uc.110","uc.62")
  #title = "Deregulated Genes"
  vres <- res
  vres <- vres %>% mutate(filter1 = abs(log2FoldChange)>=1,filter2 = padj <=0.05,filter3="")#,gene="")
  for (i in 1:length(vres$row)){
    if(is.na(vres$padj[i])){vres$filter3[i] <- "black"}else{
    if(vres$filter1[i] == TRUE && vres$filter2[i] == TRUE){
    vres$filter3[i] <- "red"
  }else{
      if(vres$filter1[i] == TRUE){
    vres$filter3[i] <- "blue"
      }else{vres$filter3[i] <- "black"}
  }}
    #ifelse(genes!="all",ifelse(!is.na(match(vres$row[i],genes)),vres$gene[i] <- vres$row[i], ""),vres$gene[i] <- vres$row[i])
    }
  
  
  
  ## plot
  p <- ggplot(vres) +
        geom_point(aes(x=log2FoldChange, y=-log10(padj),col=filter3)) +
        scale_color_manual(values = c("black","blue","red")) +
        ggtitle(title) +
        xlab("log2 fold change") + 
        ylab("-log10 adjusted p-value") +
        theme(Legend.position = "none",
              plot.title = element_text(size = rel(1.5), hjust = 0.5),
              axis.title = element_text(size = rel(1.25)))
  
  ggsave(output)
}

volcanoplot(res_stringtie,title = "Deregulation of stringtie transcripts containing TUCRs in GBM", output = paste(outputdir,"/stringtie_results_volcanoplot.png",sep=""))

```

## Generate stringtie RPKMs for filtering

```{r RPKM_stringtie, results="hide", eval = FALSE}

stringtie.rpkmcounts <- stringtiecounts[,6:ncol(stringtiecounts)]
stringtie.rpkmcounts.info <- stringtiecounts[,1:5]
rownames(stringtie.rpkmcounts) <- stringtiecounts$id

genelength <- stringtiecounts$length

seqdepth <- read.csv(file = "seqdepth_counts.csv",row.names=1)

seqdepth <- as.vector(seqdepth$counts)

rpkm <- function(counts,len,dep){
  x <- counts/len
  return(t(t(x)/dep))
}

rpkm.stringtie <- rpkm(stringtie.rpkmcounts,genelength,seqdepth)
rpkm.stringtie2 <- cbind(stringtie.rpkmcounts.info,rpkm.stringtie)

write.csv(rpkm.stringtie2,paste(outputdir,"/rpkm.stringtie.csv",sep=""))
```

## Generate heatmap for all stringtie TUCR transcripts

```{r heatmap_stringtie, message=FALSE, eval = FALSE}
heatmap.stringtie <- rpkm.stringtie2[,6:ncol(rpkm.stringtie2)]
stringtieids <- rpkm.stringtie2[,5]
means <- apply(heatmap.stringtie,1,mean)

heatmap.stringtie2 <- as.data.frame(apply(heatmap.stringtie,1:2,function(x) {ifelse(x>=10,10,x)}))
heatmap.stringtie2 <- cbind(stringtieids,means,heatmap.stringtie2)

heatmap.stringtie2 <- heatmap.stringtie2 %>% 
  gather(key = "Sample", value = "RPKM",-stringtieids,-means)

p <- ggplot(data = heatmap.stringtie2,mapping = aes(x = Sample,y = reorder(stringtieids,means), fill = RPKM)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("blue", "white", "red"), values = c(0,0.1,1)) +
  theme(panel.grid = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank())

ggsave(p,file=paste(outputdir,"/stringtie_heatmap.png",sep=""))
```

## Identifying differentially expressed CHESS genes containing TUCRs with DESeq2

```{r DESeq2_CHESS, results="hide", message = FALSE, eval = FALSE}

CHESScounts <- read.table("intersectchess_TUCR_counts.txt",header = TRUE)
CHESScounts <- distinct(CHESScounts)
metadata <- read_csv(file = metadatafile)

CHESScounts2 <- CHESScounts[,5:ncol(CHESScounts)]
rownames(CHESScounts2) <- CHESScounts[,4]
CHESScounts.info <- CHESScounts[,1:4]
CHESScounts.info$length <- (CHESScounts$end - CHESScounts$start)/1000
CHESScounts <- cbind(CHESScounts.info,CHESScounts2)

dds_CHESS <- DESeqDataSetFromMatrix(countData = CHESScounts2,
                              colData = metadata,
                              design = ~dex)
dds_CHESS <- DESeq(dds_CHESS)
res_CHESS <- results(dds_CHESS, tidy=TRUE)
res_CHESS <- as_tibble(res_CHESS)

write.csv(res_CHESS, file = paste(outputdir,"/results_CHESS.csv",sep=""))

```

## Generating volcano plot to visualize DE CHESS genes containing TUCRs

```{r DESeq2plots_CHESS, message = FALSE, eval = FALSE}

### Volcano plot

volcanoplot <- function (res,title = "Deregulated Genes", output = "results_volcanoplot.png"){
  #res <- res_TUCR
  #genes = c("uc.110","uc.62")
  #title = "Deregulated Genes"
  vres <- res
  vres <- vres %>% mutate(filter1 = abs(log2FoldChange)>=1,filter2 = padj <=0.05,filter3="")#,gene="")
  for (i in 1:length(vres$row)){
    if(is.na(vres$padj[i])){vres$filter3[i] <- "black"}else{
    if(vres$filter1[i] == TRUE && vres$filter2[i] == TRUE){
    vres$filter3[i] <- "red"
  }else{
      if(vres$filter1[i] == TRUE){
    vres$filter3[i] <- "blue"
      }else{vres$filter3[i] <- "black"}
  }}
    #ifelse(genes!="all",ifelse(!is.na(match(vres$row[i],genes)),vres$gene[i] <- vres$row[i], ""),vres$gene[i] <- vres$row[i])
    }
  
  
  
  ## plot
  p <- ggplot(vres) +
        geom_point(aes(x=log2FoldChange, y=-log10(padj),col=filter3)) +
        scale_color_manual(values = c("black","blue","red")) +
        ggtitle(title) +
        xlab("log2 fold change") + 
        ylab("-log10 adjusted p-value") +
        theme(Legend.position = "none",
              plot.title = element_text(size = rel(1.5), hjust = 0.5),
              axis.title = element_text(size = rel(1.25)))
  
  ggsave(output)
}

volcanoplot(res_CHESS,title = "Deregulation of CHESS genes containing TUCRs in GBM", output = paste(outputdir,"/CHESS_results_volcanoplot.png",sep=""))

```

## Completing survival analysis for stringtie transcripts containing TUCRs

```{r surcor_CHESS, message = FALSE, eval = FALSE}
if(!file.exists("GBM.clin.merged.txt")){unzip("GBM.clin.merged.zip")}

countdata <- read.table("intersectchess_TUCR_counts.txt",header = TRUE)
metadata <- read_csv(file = metadatafile)

posdata <- countdata[,1:4]
countdata <- countdata[,5:length(colnames(countdata))]

countdata <- as.matrix(countdata)

n_index <- c(seq(36,39),73)
t_index <- c(seq(1,35),seq(40,72),seq(74,ncol(countdata)))

vm <- function(x){
  cond <- factor(ifelse(seq(1,dim(x)[2],1) %in% t_index, 1,  0))
  d <- model.matrix(~1+cond)
  x <- t(apply(x,1,as.numeric))
  ex <- voom(x,d,plot=F)
  return(ex$E)
}

count_vm <- vm(countdata)
colnames(count_vm) <- metadata$barcode


scal <- function(x,y){
  mean_n <- rowMeans(y)  # mean of normal
  sd_n <- apply(y,1,sd)  # SD of normal
  # z score as (value - mean normal)/SD normal
  res <- matrix(nrow=nrow(x), ncol=ncol(x))
  colnames(res) <- colnames(x)
  rownames(res) <- rownames(x)
  for(i in 1:dim(x)[1]){
    for(j in 1:dim(x)[2]){
      res[i,j] <- (x[i,j]-mean_n[i])/sd_n[i]
    }
  }
  return(res)
}

z_rna <- scal(count_vm[,t_index],count_vm[,n_index])
rownames(z_rna) <- posdata[,4]

clinical <- read.table("GBM.clin.merged.txt",header = TRUE)
clinical$time <- as.numeric(clinical$time)

sum(clinical$patient %in% colnames(z_rna))

ind_tum <- which(unique(colnames(z_rna)) %in% clinical$patient)
ind_clin <- which(clinical$patient %in% colnames(z_rna))

out.tab <- c()
for(x in 1:nrow(count_vm)){
  ind_gene <- x
  s <- Surv(clinical$time[ind_clin],clinical$status[ind_clin])
  cx <- coxph(formula = s ~ count_vm[ind_gene,ind_tum])
  cx <- tidy(cx)
  out.tab <- rbind(out.tab,cx)
}

surv_CHESS <- cbind(posdata,out.tab)
surv_CHESS <- surv_CHESS %>% select(-term)

write_csv(surv_CHESS,paste(outputdir,"/survival_TUCRs_CHESS.csv",sep=""))

```

## Generate CHESS gene RPKMs for filtering

```{r RPKM_CHESS, results="hide", eval = FALSE}

CHESS.rpkmcounts <- CHESScounts[,6:ncol(CHESScounts)]
CHESS.rpkmcounts.info <- CHESScounts[,1:5]
rownames(CHESS.rpkmcounts) <- CHESScounts$id

genelength <- CHESScounts$length

seqdepth <- read.csv(file = "seqdepth_counts.csv",row.names=1)

seqdepth <- as.vector(seqdepth$counts)

rpkm <- function(counts,len,dep){
  x <- counts/len
  return(t(t(x)/dep))
}

rpkm.CHESS <- rpkm(CHESS.rpkmcounts,genelength,seqdepth)
rpkm.CHESS2 <- cbind(CHESS.rpkmcounts.info,rpkm.CHESS)

write.csv(rpkm.CHESS2,paste(outputdir,"/rpkm.CHESS.csv",sep=""))
```

## Generate heatmap for all CHESS genes containing TUCRs

```{r heatmap_CHESS, message=FALSE, eval = FALSE}
heatmap.CHESS <- rpkm.CHESS2[,6:ncol(rpkm.CHESS2)]
CHESSids <- rpkm.CHESS2[,5]
means <- apply(heatmap.CHESS,1,mean)

heatmap.CHESS2 <- as.data.frame(apply(heatmap.CHESS,1:2,function(x) {ifelse(x>=10,10,x)}))
heatmap.CHESS2 <- cbind(CHESSids,means,heatmap.CHESS2)

heatmap.CHESS2 <- heatmap.CHESS2 %>% 
  gather(key = "Sample", value = "RPKM",-CHESSids,-means)

p <- ggplot(data = heatmap.CHESS2,mapping = aes(x = Sample,y = reorder(CHESSids,means), fill = RPKM)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("blue", "white", "red"), values = c(0,0.1,1)) +
  theme(panel.grid = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank())

ggsave(p,file=paste(outputdir,"/CHESS_heatmap.png",sep=""))
```

## Visualize Stephen and Alex's FPKMs to compare to my own RPKMs

```{r sdt_heatmap, message=FALSE}
fpkm.sdt <- read_csv("aboundaer_tucr_gbm-survival_fpkms.csv")
heatmap.sdt <- fpkm.sdt[,2:ncol(fpkm.sdt)]
sdtids <- fpkm.sdt[,1]
means <- apply(heatmap.sdt,1,mean)

heatmap.sdt2 <- as.data.frame(apply(heatmap.sdt,1:2,function(x) {ifelse(x>=10,10,x)}))
heatmap.sdt2 <- cbind(sdtids,means,heatmap.sdt2)

heatmap.sdt2 <- heatmap.sdt2 %>% 
  gather(key = "Sample", value = "FPKM",-id,-means)

p <- ggplot(data = heatmap.sdt2,mapping = aes(x = Sample,y = reorder(id,means), fill = FPKM)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("blue", "white", "red"), values = c(0,0.1,1)) +
  theme(panel.grid = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank())

ggsave(p,file=paste(outputdir,"/sdt_heatmap.png",sep=""))
```