---
title: "Old Code Chunks"
author: "Myron K Gibert Jr"
date: "7/12/2020"
output: pdf_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Generate heatmap for my TUCRs of interest

```{r heatmap_Myron}
heatmap.TUCRs <- rpkm.TUCRs2[,6:ncol(rpkm.TUCRs2)]
TUCRids <- rpkm.TUCRs2[,4]
means <- apply(heatmap.TUCRs,1,mean)

heatmap.TUCRs2 <- as.data.frame(apply(heatmap.TUCRs,1:2,function(x) {ifelse(x>=10,10,x)}))
heatmap.TUCRs2 <- cbind(TUCRids,means,heatmap.TUCRs2)

heatmap.TUCRs2 <- heatmap.TUCRs2 %>% 
  gather(key = "Sample", value = "RPKM",-TUCRids,-means) %>% 
  filter(TUCRids == "uc.62"|TUCRids == "uc.110"|TUCRids == "uc.193"|TUCRids == "uc.210"|TUCRids == "uc.427"|TUCRids == "uc.212"|TUCRids == "uc.341")

p <- ggplot(data = heatmap.TUCRs2,mapping = aes(x = Sample,y = reorder(TUCRids,means), fill = RPKM)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("blue", "white", "red"), values = c(0,0.1,1)) +
  theme(panel.grid = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks = element_blank())

ggsave(p,file=paste(outputdir,"/Myron_heatmap.png",sep=""))
```

## Identifying differentially expressed stringtie transcripts containing TUCRs with DESeq2

```{r DESeq2_stringtie, results="hide", message = FALSE, eval = FALSE}

stringtiecounts <- read.table("stringtie_TUCR_counts.txt",header = TRUE)
stringtiecounts <- distinct(stringtiecounts)
metadata <- read_csv(file = metadatafile)

stringtiecounts2 <- stringtiecounts[,5:ncol(stringtiecounts)]
rownames(stringtiecounts2) <- stringtiecounts[,4]
stringtiecounts.info <- stringtiecounts[,1:4]
stringtiecounts.info$length <- (stringtiecounts$end - stringtiecounts$start)/1000
stringtiecounts <- cbind(stringtiecounts.info,stringtiecounts2)

dds_stringtie <- DESeqDataSetFromMatrix(countData = stringtiecounts2,
                              colData = metadata,
                              design = ~dex)
dds_stringtie <- DESeq(dds_stringtie)
res_stringtie <- results(dds_stringtie, tidy=TRUE)
res_stringtie <- as_tibble(res_stringtie)

write.csv(res_stringtie, file = paste(outputdir,"/results_stringtie.csv",sep=""))

```

## Completing survival analysis for stringtie transcripts containing TUCRs

```{r surcor_stringtie, message = FALSE,eval = FALSE}
if(!file.exists("GBM.clin.merged.txt")){unzip("GBM.clin.merged.zip")}

countdata <- read.table("stringtie_TUCR_counts.txt",header = TRUE)
metadata <- read_csv(file = metadatafile)

posdata <- countdata[,1:4]
countdata <- countdata[,5:length(colnames(countdata))]

countdata <- as.matrix(countdata)


n_index <- c(seq(36,39),73)
t_index <- c(seq(1,35),seq(40,72),seq(74,ncol(countdata)))

vm <- function(x){
  cond <- factor(ifelse(seq(1,dim(x)[2],1) %in% t_index, 1,  0))
  d <- model.matrix(~1+cond)
  x <- t(apply(x,1,as.numeric))
  ex <- voom(x,d,plot=F)
  return(ex$E)
}

count_vm <- vm(countdata)
colnames(count_vm) <- metadata$barcode


scal <- function(x,y){
  mean_n <- rowMeans(y)  # mean of normal
  sd_n <- apply(y,1,sd)  # SD of normal
  # z score as (value - mean normal)/SD normal
  res <- matrix(nrow=nrow(x), ncol=ncol(x))
  colnames(res) <- colnames(x)
  rownames(res) <- rownames(x)
  for(i in 1:dim(x)[1]){
    for(j in 1:dim(x)[2]){
      res[i,j] <- (x[i,j]-mean_n[i])/sd_n[i]
    }
  }
  return(res)
}

z_rna <- scal(count_vm[,t_index],count_vm[,n_index])
rownames(z_rna) <- posdata[,4]

clinical <- read.table("GBM.clin.merged.txt",header = TRUE)
clinical$time <- as.numeric(clinical$time)

sum(clinical$patient %in% colnames(z_rna))

ind_tum <- which(unique(colnames(z_rna)) %in% clinical$patient)
ind_clin <- which(clinical$patient %in% colnames(z_rna))

out.tab <- c()
for(x in 1:nrow(count_vm)){
  ind_gene <- x
  s <- Surv(clinical$time[ind_clin],clinical$status[ind_clin])
  cx <- coxph(formula = s ~ count_vm[ind_gene,ind_tum])
  cx <- tidy(cx)
  out.tab <- rbind(out.tab,cx)
}

surv_stringtie <- cbind(posdata,out.tab)
surv_stringtie <- surv_stringtie %>% select(-term)

write_csv(surv_stringtie,paste(outputdir,"/survival_TUCRs_stringtie.csv",sep=""))
```

## Generating volcano plot to visualize DE stringtie transcripts

```{r DESeq2plots_stringtie, message = FALSE, eval = FALSE}

### Volcano plot

volcanoplot <- function (res,title = "Deregulated Genes", output = "results_volcanoplot.png"){
  #res <- res_TUCR
  #genes = c("uc.110","uc.62")
  #title = "Deregulated Genes"
  vres <- res
  vres <- vres %>% mutate(filter1 = abs(log2FoldChange)>=1,filter2 = padj <=0.05,filter3="")#,gene="")
  for (i in 1:length(vres$row)){
    if(is.na(vres$padj[i])){vres$filter3[i] <- "black"}else{
    if(vres$filter1[i] == TRUE && vres$filter2[i] == TRUE){
    vres$filter3[i] <- "red"
  }else{
      if(vres$filter1[i] == TRUE){
    vres$filter3[i] <- "blue"
      }else{vres$filter3[i] <- "black"}
  }}
    #ifelse(genes!="all",ifelse(!is.na(match(vres$row[i],genes)),vres$gene[i] <- vres$row[i], ""),vres$gene[i] <- vres$row[i])
    }
  
  
  
  ## plot
  p <- ggplot(vres) +
        geom_point(aes(x=log2FoldChange, y=-log10(padj),col=filter3)) +
        scale_color_manual(values = c("black","blue","red")) +
        ggtitle(title) +
        xlab("log2 fold change") + 
        ylab("-log10 adjusted p-value") +
        theme(Legend.position = "none",
              plot.title = element_text(size = rel(1.5), hjust = 0.5),
              axis.title = element_text(size = rel(1.25)))
  
  ggsave(output)
}

volcanoplot(res_stringtie,title = "Deregulation of stringtie transcripts containing TUCRs in GBM", output = paste(outputdir,"/stringtie_results_volcanoplot.png",sep=""))

```

## Generate stringtie RPKMs for filtering

```{r RPKM_stringtie, results="hide", eval = FALSE}

stringtie.rpkmcounts <- stringtiecounts[,6:ncol(stringtiecounts)]
stringtie.rpkmcounts.info <- stringtiecounts[,1:5]
rownames(stringtie.rpkmcounts) <- stringtiecounts$id

genelength <- stringtiecounts$length

seqdepth <- read.csv(file = "seqdepth_counts.csv",row.names=1)

seqdepth <- as.vector(seqdepth$counts)

rpkm <- function(counts,len,dep){
  x <- counts/len
  return(t(t(x)/dep))
}

rpkm.stringtie <- rpkm(stringtie.rpkmcounts,genelength,seqdepth)
rpkm.stringtie2 <- cbind(stringtie.rpkmcounts.info,rpkm.stringtie)

write.csv(rpkm.stringtie2,paste(outputdir,"/rpkm.stringtie.csv",sep=""))
```

## Generate heatmap for all stringtie TUCR transcripts

```{r heatmap_stringtie, message=FALSE, eval = FALSE}
heatmap.stringtie <- rpkm.stringtie2[,6:ncol(rpkm.stringtie2)]
stringtieids <- rpkm.stringtie2[,5]
means <- apply(heatmap.stringtie,1,mean)

heatmap.stringtie2 <- as.data.frame(apply(heatmap.stringtie,1:2,function(x) {ifelse(x>=10,10,x)}))
heatmap.stringtie2 <- cbind(stringtieids,means,heatmap.stringtie2)

heatmap.stringtie2 <- heatmap.stringtie2 %>% 
  gather(key = "Sample", value = "RPKM",-stringtieids,-means)

p <- ggplot(data = heatmap.stringtie2,mapping = aes(x = Sample,y = reorder(stringtieids,means), fill = RPKM)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("blue", "white", "red"), values = c(0,0.1,1)) +
  theme(panel.grid = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank())

ggsave(p,file=paste(outputdir,"/stringtie_heatmap.png",sep=""))
```

## Identifying differentially expressed CHESS genes containing TUCRs with DESeq2

```{r DESeq2_CHESS, results="hide", message = FALSE, eval = FALSE}

CHESScounts <- read.table("intersectchess_TUCR_counts.txt",header = TRUE)
CHESScounts <- distinct(CHESScounts)
metadata <- read_csv(file = metadatafile)

CHESScounts2 <- CHESScounts[,5:ncol(CHESScounts)]
rownames(CHESScounts2) <- CHESScounts[,4]
CHESScounts.info <- CHESScounts[,1:4]
CHESScounts.info$length <- (CHESScounts$end - CHESScounts$start)/1000
CHESScounts <- cbind(CHESScounts.info,CHESScounts2)

dds_CHESS <- DESeqDataSetFromMatrix(countData = CHESScounts2,
                              colData = metadata,
                              design = ~dex)
dds_CHESS <- DESeq(dds_CHESS)
res_CHESS <- results(dds_CHESS, tidy=TRUE)
res_CHESS <- as_tibble(res_CHESS)

write.csv(res_CHESS, file = paste(outputdir,"/results_CHESS.csv",sep=""))

```

## Generating volcano plot to visualize DE CHESS genes containing TUCRs

```{r DESeq2plots_CHESS, message = FALSE, eval = FALSE}

### Volcano plot

volcanoplot <- function (res,title = "Deregulated Genes", output = "results_volcanoplot.png"){
  #res <- res_TUCR
  #genes = c("uc.110","uc.62")
  #title = "Deregulated Genes"
  vres <- res
  vres <- vres %>% mutate(filter1 = abs(log2FoldChange)>=1,filter2 = padj <=0.05,filter3="")#,gene="")
  for (i in 1:length(vres$row)){
    if(is.na(vres$padj[i])){vres$filter3[i] <- "black"}else{
    if(vres$filter1[i] == TRUE && vres$filter2[i] == TRUE){
    vres$filter3[i] <- "red"
  }else{
      if(vres$filter1[i] == TRUE){
    vres$filter3[i] <- "blue"
      }else{vres$filter3[i] <- "black"}
  }}
    #ifelse(genes!="all",ifelse(!is.na(match(vres$row[i],genes)),vres$gene[i] <- vres$row[i], ""),vres$gene[i] <- vres$row[i])
    }
  
  
  
  ## plot
  p <- ggplot(vres) +
        geom_point(aes(x=log2FoldChange, y=-log10(padj),col=filter3)) +
        scale_color_manual(values = c("black","blue","red")) +
        ggtitle(title) +
        xlab("log2 fold change") + 
        ylab("-log10 adjusted p-value") +
        theme(Legend.position = "none",
              plot.title = element_text(size = rel(1.5), hjust = 0.5),
              axis.title = element_text(size = rel(1.25)))
  
  ggsave(output)
}

volcanoplot(res_CHESS,title = "Deregulation of CHESS genes containing TUCRs in GBM", output = paste(outputdir,"/CHESS_results_volcanoplot.png",sep=""))

```

## Completing survival analysis for stringtie transcripts containing TUCRs

```{r surcor_CHESS, message = FALSE, eval = FALSE}
if(!file.exists("GBM.clin.merged.txt")){unzip("GBM.clin.merged.zip")}

countdata <- read.table("intersectchess_TUCR_counts.txt",header = TRUE)
metadata <- read_csv(file = metadatafile)

posdata <- countdata[,1:4]
countdata <- countdata[,5:length(colnames(countdata))]

countdata <- as.matrix(countdata)

n_index <- c(seq(36,39),73)
t_index <- c(seq(1,35),seq(40,72),seq(74,ncol(countdata)))

vm <- function(x){
  cond <- factor(ifelse(seq(1,dim(x)[2],1) %in% t_index, 1,  0))
  d <- model.matrix(~1+cond)
  x <- t(apply(x,1,as.numeric))
  ex <- voom(x,d,plot=F)
  return(ex$E)
}

count_vm <- vm(countdata)
colnames(count_vm) <- metadata$barcode


scal <- function(x,y){
  mean_n <- rowMeans(y)  # mean of normal
  sd_n <- apply(y,1,sd)  # SD of normal
  # z score as (value - mean normal)/SD normal
  res <- matrix(nrow=nrow(x), ncol=ncol(x))
  colnames(res) <- colnames(x)
  rownames(res) <- rownames(x)
  for(i in 1:dim(x)[1]){
    for(j in 1:dim(x)[2]){
      res[i,j] <- (x[i,j]-mean_n[i])/sd_n[i]
    }
  }
  return(res)
}

z_rna <- scal(count_vm[,t_index],count_vm[,n_index])
rownames(z_rna) <- posdata[,4]

clinical <- read.table("GBM.clin.merged.txt",header = TRUE)
clinical$time <- as.numeric(clinical$time)

sum(clinical$patient %in% colnames(z_rna))

ind_tum <- which(unique(colnames(z_rna)) %in% clinical$patient)
ind_clin <- which(clinical$patient %in% colnames(z_rna))

out.tab <- c()
for(x in 1:nrow(count_vm)){
  ind_gene <- x
  s <- Surv(clinical$time[ind_clin],clinical$status[ind_clin])
  cx <- coxph(formula = s ~ count_vm[ind_gene,ind_tum])
  cx <- tidy(cx)
  out.tab <- rbind(out.tab,cx)
}

surv_CHESS <- cbind(posdata,out.tab)
surv_CHESS <- surv_CHESS %>% select(-term)

write_csv(surv_CHESS,paste(outputdir,"/survival_TUCRs_CHESS.csv",sep=""))

```

## Generate CHESS gene RPKMs for filtering

```{r RPKM_CHESS, results="hide", eval = FALSE}

CHESS.rpkmcounts <- CHESScounts[,6:ncol(CHESScounts)]
CHESS.rpkmcounts.info <- CHESScounts[,1:5]
rownames(CHESS.rpkmcounts) <- CHESScounts$id

genelength <- CHESScounts$length

seqdepth <- read.csv(file = "seqdepth_counts.csv",row.names=1)

seqdepth <- as.vector(seqdepth$counts)

rpkm <- function(counts,len,dep){
  x <- counts/len
  return(t(t(x)/dep))
}

rpkm.CHESS <- rpkm(CHESS.rpkmcounts,genelength,seqdepth)
rpkm.CHESS2 <- cbind(CHESS.rpkmcounts.info,rpkm.CHESS)

write.csv(rpkm.CHESS2,paste(outputdir,"/rpkm.CHESS.csv",sep=""))
```

## Generate heatmap for all CHESS genes containing TUCRs

```{r heatmap_CHESS, message=FALSE, eval = FALSE}
heatmap.CHESS <- rpkm.CHESS2[,6:ncol(rpkm.CHESS2)]
CHESSids <- rpkm.CHESS2[,5]
means <- apply(heatmap.CHESS,1,mean)

heatmap.CHESS2 <- as.data.frame(apply(heatmap.CHESS,1:2,function(x) {ifelse(x>=10,10,x)}))
heatmap.CHESS2 <- cbind(CHESSids,means,heatmap.CHESS2)

heatmap.CHESS2 <- heatmap.CHESS2 %>% 
  gather(key = "Sample", value = "RPKM",-CHESSids,-means)

p <- ggplot(data = heatmap.CHESS2,mapping = aes(x = Sample,y = reorder(CHESSids,means), fill = RPKM)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("blue", "white", "red"), values = c(0,0.1,1)) +
  theme(panel.grid = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank())

ggsave(p,file=paste(outputdir,"/CHESS_heatmap.png",sep=""))
```

## Visualize Stephen and Alex's FPKMs to compare to my own RPKMs

```{r sdt_heatmap, message=FALSE}
fpkm.sdt <- read_csv("aboundaer_tucr_gbm-survival_fpkms.csv")
heatmap.sdt <- fpkm.sdt[,2:ncol(fpkm.sdt)]
sdtids <- fpkm.sdt[,1]
means <- apply(heatmap.sdt,1,mean)

heatmap.sdt2 <- as.data.frame(apply(heatmap.sdt,1:2,function(x) {ifelse(x>=10,10,x)}))
heatmap.sdt2 <- cbind(sdtids,means,heatmap.sdt2)

heatmap.sdt2 <- heatmap.sdt2 %>% 
  gather(key = "Sample", value = "FPKM",-id,-means)

p <- ggplot(data = heatmap.sdt2,mapping = aes(x = Sample,y = reorder(id,means), fill = FPKM)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("blue", "white", "red"), values = c(0,0.1,1)) +
  theme(panel.grid = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank())

ggsave(p,file=paste(outputdir,"/sdt_heatmap.png",sep=""))
```