---
title: 'T-type calcium channel expression in neuron connect versus neuron disconnected glioma'
author: "Myron Keith Gibert Jr"
date: "February 26, 2020"
output: pdf_document
toc: true
---

## Introduction

## Set parameters

```{r parameters}

countfilename <- "mousegene_counts.txt"

outputdir <- "Outputs"

```

## Install required programs

```{r setup, message = FALSE}
knitr::opts_chunk$set(echo = TRUE)

if (!require("tidyverse")) install.packages("tidyverse")
library("tidyverse")

if (!require("ggplot2")) install.packages("ggplot2")
library("ggplot2")

if (!require("ggforce")) install.packages("ggforce")
library("ggforce")

if (!require("ggrepel")) install.packages("ggrepel")
library("ggrepel")

if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
if (!requireNamespace("DESeq2", quietly = TRUE))BiocManager::install("DESeq2")
library("DESeq2")

```

```{r outputs}
if(!dir.exists(outputdir)){
  dir.create(outputdir)
}
```


```{r tinytex, include=FALSE}
## INSTALLING LATEX FOR RMARKDOWN

#RMarkdown requires LaTex to create pdf documents as an output. More information can be found [here](https://bookdown.org/yihui/rmarkdown/pdf-document.html). Alternatively, output can be set to "html_document" or "word_document". End users may install LaTex by setting the "wanttinytex" variable to FALSE and running the following script:

#Install tinytex to let RMarkdown create a pdf document? Default: wanttinytex <- FALSE
wanttinytex <- TRUE

if(tinytex:::is_tinytex()==FALSE && wanttinytex == TRUE){
if (!require("tinytex")) install.packages("tinytex")
tinytex::install_tinytex()
}
```

## Aquire the data using the SRA toolkit

```{bash setup, eval = FALSE}

#using SRA Toolkit version 2.10.5

#RNA-Seq connected Glioma 2

prefetch SRR444442

#RNA-Seq connected Glioma 1

prefetch SRR444478

```

## Generate fastq files

```{bash fastq, eval = FALSE}

for sra in SRR*
do
echo $sra
fastq-dump $sra
done

```

## Align mm10 genome

```{bash mm10, eval = FALSE}

wget https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz
gunzip hg38.fa.gz

#Rivanna only.  Skip otherwise.
module load gcc/7.1.0
module load bowtie2/2.2.9

bowtie2-build hg38.fa hg38

```

## Align fastq files to reference genome

```{bash sam, eval = FALSE}

for fq in *.fastq
do
name=$(echo $fq | awk -F".fastq" '{print $1}')
echo $name
bowtie2 -x hg38 -U $fq -S $name.sam
done

```

## Convert sam to bam

```{bash bam, eval = FALSE}

for sam in *.sam
do
name=$(echo $sam | awk -F".sam" '{print $1}')
echo $name
samtools view -b $sam | samtools sort -o $name.bam
done
rm *sam

```

## Index bam files

```{bash index, eval = FALSE}

for bam in *bam
do
name=$(echo $bam | awk -F ".bam" '{print $1}')
echo $name
samtools index $bam
done

```

## Download mouse genes and convert to bed

I got this methodology from [here](https://github.com/stevekm/reference-annotations/blob/master/Makefile) since my other methods were not working.

```{bash refgenes, eval = FALSE}

macs2 callpeak -t SRR444442.bam -n SRR444442 -g hs
macs2 callpeak -t SRR444478.bam -n SRR444478 -g hs

wc -l *Peak

cat SRR444442_peaks.narrowPeak | cut -f 1-4 | grep -v chrMT | grep -v chrGL | grep -v chrKI > h3k4me3.bed
cat SRR444478_peaks.narrowPeak | cut -f 1-4 | grep -v chrMT | grep -v chrGL | grep -v chrKI > pol2.bed

```

## Generate counts for mouse genes

```{bash counts, eval = FALSE}

wget ftp://ftp.ensembl.org/pub/release-90/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz

gunzip Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz

samtools faidx Homo_sapiens.GRCh38.dna.primary_assembly.fa

awk -v OFS='\t' {'print $1,$2'} Homo_sapiens.GRCh38.dna.primary_assembly.fa.fai > hg38_genomeFile.txt

cat hg38_genomeFile.txt | grep -v KI | grep -v MT | grep -v GL | sed -e "s/^/chr/g" > temp

mv temp hg38_genomeFile.txt

```

## Identifying differentially expressed TUCRs with DESeq2

```{bash slop, results="hide", eval = FALSE}

bedtools slop -b 10000 -i hg38.ultraConserved.bed -g hg38_genomeFile.txt > tucrs_pm10kb.bed

bedtools slop -b 10000 -i TUCR_intergenic.bed -g hg38_genomeFile.txt > intergenic_tucrs_pm10kb.bed

bedtools slop -b 10000 -i CHESScoding.bed -g hg38_genomeFile.txt > coding_pm10kb.bed

bedtools slop -b 10000 -i CHESSlncRNA.bed -g hg38_genomeFile.txt > lncRNAs_pm10kb.bed

bedtools slop -b 10000 -i CHESSantisense.bed -g hg38_genomeFile.txt > antisense_pm10kb.bed

bedtools slop -b 10000 -i CHESSmisc.bed -g hg38_genomeFile.txt > misc_pm10kb.bed

```

## Fisher

```{bash fisher, results="hide", eval = FALSE}

bedtools fisher -a tucrs_pm10kb.bed -b pol2.bed -g hg38_genomeFile.txt > 
closest_pol2_tucrs_output.txt

bedtools fisher -a intergenic_tucrs_pm10kb.bed -b pol2.bed -g hg38_genomeFile.txt > pol2_intergenic_tucrs_output.txt

bedtools fisher -a CHESScoding.bed -b pol2.bed -g hg38_genomeFile.txt > 
pol2_coding_output.txt

bedtools fisher -a CHESSlncRNA.bed -b pol2.bed -g hg38_genomeFile.txt > 
pol2_lncRNA_output.txt

bedtools fisher -a CHESSantisense.bed -b pol2.bed -g hg38_genomeFile.txt > 
pol2_antisense_output.txt

bedtools fisher -a CHESSmisc.bed -b pol2.bed -g hg38_genomeFile.txt > 
pol2_misc_output.txt

bedtools fisher -a tucrs_pm10kb.bed -b h3k4me3.bed -g hg38_genomeFile.txt > 
h3k4me3_tucrs_output.txt

bedtools fisher -a intergenic_tucrs_pm10kb.bed -b h3k4me3.bed -g hg38_genomeFile.txt >
h3k4me3_intergenic_tucrs_output.txt

bedtools fisher -a CHESScoding.bed -b h3k4me3.bed -g hg38_genomeFile.txt > 
h3k4me3_coding_output.txt

bedtools fisher -a CHESSlncRNA.bed -b h3k4me3.bed -g hg38_genomeFile.txt > 
h3k4me3_lncRNA_output.txt

bedtools fisher -a CHESSantisense.bed -b h3k4me3.bed -g hg38_genomeFile.txt > h3k4me3_antisense_output.txt

bedtools fisher -a CHESSmisc.bed -b h3k4me3.bed -g hg38_genomeFile.txt > 
h3k4me3_misc_output.txt

```

## Fisher

```{bash fisher, results="hide", eval = FALSE}

bedtools random -l 800 -n 481 -g hg38_genomeFile.txt > random_intervals.bed

bedtools sort -i random_intervals.bed > random_intervals.bed

bedtools closest -a random_intervals.bed -b pol2.bed -d > 
closest_pol2_random_intervals_output.bed

bedtools sort -i hg38.ultraConserved.bed > hg38.ultraConserved.bed

bedtools closest -a hg38.ultraConserved.bed -b pol2.bed -d > 
closest_pol2_tucrs_output.bed

bedtools sort -i TUCR_intergenic.bed > TUCR_intergenic.bed

bedtools closest -a TUCR_intergenic.bed -b pol2.bed -g hg38_genomeFile.txt > closest_pol2_intergenic_tucrs_output.bed

bedtools sort -i CHESScoding.bed > CHESScoding.bed

bedtools closest -a CHESScoding.bed -b pol2.bed -g hg38_genomeFile.txt > 
closest_pol2_coding_output.bed

bedtools sort -i CHESSlncRNA.bed > CHESSlncRNA.bed

bedtools closest -a CHESSlncRNA.bed -b pol2.bed -g hg38_genomeFile.txt > 
closest_pol2_lncRNA_output.bed

bedtools sort -i CHESSantisense.bed > CHESSantisense.bed

bedtools closest -a CHESSantisense.bed -b pol2.bed -g hg38_genomeFile.txt > closest_pol2_antisense_output.bed

bedtools sort -i CHESSmisc.bed > CHESSmisc.bed

bedtools closest -a CHESSmisc.bed -b pol2.bed -g hg38_genomeFile.txt > 
closest_pol2_misc_output.bed

bedtools closest -a hg38.ultraConserved.bed -b h3k4me3.bed -g hg38_genomeFile.txt > closest_h3k4me3_tucrs_output.bed

bedtools closest -a TUCR_intergenic.bed -b h3k4me3.bed -g hg38_genomeFile.txt > closest_h3k4me3_intergenic_tucrs_output.bed

bedtools closest -a CHESScoding.bed -b h3k4me3.bed -g hg38_genomeFile.txt > closest_h3k4me3_coding_output.bed

bedtools closest -a CHESSlncRNA.bed -b h3k4me3.bed -g hg38_genomeFile.txt > closest_h3k4me3_lncRNA_output.bed

bedtools closest -a CHESSantisense.bed -b h3k4me3.bed -g hg38_genomeFile.txt > closest_h3k4me3_antisense_output.bed

bedtools closest -a CHESSmisc.bed -b h3k4me3.bed -g hg38_genomeFile.txt > closest_h3k4me3_misc_output.bed

```

```{r CHIPgraphs-h3k4me3}

filenames.H3K4me3 <- list.files("./closest_bed/H3K4me3/")

mastertable.H3k4me3 <- data.frame(matrix(ncol=10,nrow=0, dimnames=list(NULL,                  c("chromosome.gene","start.gene","end.gene","gene","annot",
"chromosome.H3K4me3","start.H3K4me3","end.H3K4me3","H3K4me3","dist"))))

i = 1
for(i in 1:length(filenames.H3K4me3)){
  print(i)
  df <- read.table(paste("./closest_bed/H3K4me3/",filenames.H3K4me3[i],sep=""))
  colnames(df) <- colnames(mastertable.H3k4me3)
  mastertable.H3k4me3 <- rbind(mastertable.H3k4me3,df)
}

mastertable.H3k4me3.sum <- mastertable.H3k4me3 %>%
  mutate(interval = ifelse(dist<500, 500,
                          ifelse(dist<1000, 1000,
                          ifelse(dist<1500, 1500,
                          ifelse(dist<2000, 2000,
                          ifelse(dist<2500, 2500,
                          ifelse(dist<3000, 3000,
                          ifelse(dist<3500, 3500,
                          ifelse(dist<4000, 4000,
                          ifelse(dist<4500, 4500,
                          ifelse(dist<5000, 5000,
                          ifelse(dist<5500, 5500,
                          ifelse(dist<6000, 6000,
                          ifelse(dist<6500, 6500,
                          ifelse(dist<7000, 7000,
                          ifelse(dist<7500, 7500,
                          ifelse(dist<8000, 8000,
                          ifelse(dist<8500, 8500,
                          ifelse(dist<9000, 9000,
                          ifelse(dist<9500, 9500,
                          ifelse(dist<10000, 10000,10500))))))))))))))))))))) %>%
  filter(interval != 10500) %>%
  mutate(order = ifelse(annot=="protein_coding", 1,
                          ifelse(annot=="lncRNA", 2,
                          ifelse(annot=="TUCR", 3,
                          ifelse(annot=="intergenic_TUCR", 4,
                          ifelse(annot=="antisense_RNA", 5,
                          ifelse(annot=="misc_RNA", 6,
                          ifelse(annot=="random", 7,0
                          )))))))) %>%
  mutate(annot = fct_reorder(annot,order)) %>%
  group_by(annot,interval,order) %>%
  summarise(total = n()) %>%
  ungroup() %>%
  group_by(annot) %>%
  mutate(div = sum(total)) %>%
  mutate(prop = total/div) %>%
  mutate(log10 = log10(total)) %>%
  mutate(log2 = log2(total)) %>%
  arrange(order)

p <- ggplot(mastertable.H3k4me3.sum, aes(x=interval,y=log2,fill=annot,order=order)) + 
  geom_bar(stat="identity")                                                      + 
  scale_fill_manual(values = c("red3","green4","blue3","cyan","deeppink","yellow2","orange2")) +
  xlab("distance (nt)")                                           + 
  ylab("log(density)")                                           + 
  theme(
    plot.title = element_blank(),
    axis.title = element_text(
      size = rel(1.25), 
      face="bold"),
    axis.text.x = element_text(
      angle = -90, 
      size = 10),
    panel.grid.major = element_blank(), 
    panel.grid.minor = element_blank(),
    panel.background = element_blank(), 
    axis.line = element_line(
      colour = "black"))

ggsave(
    file=paste(outputdir,"/tucr_dist_h3k4me3.png",sep=""),
    plot = print(p),
    width = 10, 
    height = 5, 
    dpi = 300) 

```

```{r CHIPgraphs-pol2}

filenames.pol2 <- list.files("./closest_bed/pol2/")

mastertable.pol2 <- data.frame(matrix(ncol=10,nrow=0, dimnames=list(NULL,                  c("chromosome.gene","start.gene","end.gene","gene","annot",
"chromosome.pol2","start.pol2","end.pol2","pol2","dist"))))

i = 1
for(i in 1:length(filenames.pol2)){
  print(i)
  df <- read.table(paste("./closest_bed/pol2/",filenames.pol2[i],sep=""))
  colnames(df) <- colnames(mastertable.pol2)
  mastertable.pol2 <- rbind(mastertable.pol2,df)
}

mastertable.pol2.sum <- mastertable.pol2 %>%
  mutate(interval = ifelse(dist<500, 500,
                          ifelse(dist<1000, 1000,
                          ifelse(dist<1500, 1500,
                          ifelse(dist<2000, 2000,
                          ifelse(dist<2500, 2500,
                          ifelse(dist<3000, 3000,
                          ifelse(dist<3500, 3500,
                          ifelse(dist<4000, 4000,
                          ifelse(dist<4500, 4500,
                          ifelse(dist<5000, 5000,
                          ifelse(dist<5500, 5500,
                          ifelse(dist<6000, 6000,
                          ifelse(dist<6500, 6500,
                          ifelse(dist<7000, 7000,
                          ifelse(dist<7500, 7500,
                          ifelse(dist<8000, 8000,
                          ifelse(dist<8500, 8500,
                          ifelse(dist<9000, 9000,
                          ifelse(dist<9500, 9500,
                          ifelse(dist<10000, 10000,10500))))))))))))))))))))) %>%
  filter(interval != 10500) %>%
  mutate(order = ifelse(annot=="protein_coding", 1,
                          ifelse(annot=="lncRNA", 2,
                          ifelse(annot=="TUCR", 3,
                          ifelse(annot=="intergenic_TUCR", 4,
                          ifelse(annot=="antisense_RNA", 5,
                          ifelse(annot=="misc_RNA", 6,
                          ifelse(annot=="random", 7,0
                          )))))))) %>%
  mutate(annot = fct_reorder(annot,order)) %>%
  group_by(annot,interval,order) %>%
  summarise(total = n()) %>%
  ungroup() %>%
  group_by(annot) %>%
  mutate(div = sum(total)) %>%
  mutate(prop = total/div) %>%
  mutate(log10 = log10(total)) %>%
  mutate(log2 = log2(total)) %>%
  arrange(order)

p <- ggplot(mastertable.pol2.sum, aes(x=interval,y=log2,fill=annot,order=order)) + 
  geom_bar(stat="identity")                                                      + 
  scale_fill_manual(values = c("red3","green4","blue3","cyan","deeppink","yellow2","orange2")) +
  xlab("distance (nt)")                                           + 
  ylab("log(density)")                                           + 
  theme(
    plot.title = element_blank(),
    axis.title = element_text(
      size = rel(1.25), 
      face="bold"),
    axis.text.x = element_text(
      angle = -90, 
      size = 10),
    panel.grid.major = element_blank(), 
    panel.grid.minor = element_blank(),
    panel.background = element_blank(), 
    axis.line = element_line(
      colour = "black"))

ggsave(
    file=paste(outputdir,"/tucr_dist_pol2.png",sep=""),
    plot = print(p),
    width = 10, 
    height = 5, 
    dpi = 300) 

```