---
title: 'Uncovering a role for transcribed ultra-conserved regions as long non-coding RNAs in glioblastoma'
author: "Myron Keith Gibert Jr"
date: "January 27, 2020"
output: pdf_document
toc: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r tinytex, include=FALSE}
## INSTALLING LATEX FOR RMARKDOWN

#RMarkdown requires LaTex to create pdf documents as an output. More information can be found [here](https://bookdown.org/yihui/rmarkdown/pdf-document.html). Alternatively, output can be set to "html_document" or "word_document". End users may install LaTex by setting the "wanttinytex" variable to FALSE and running the following script:

#Install tinytex to let RMarkdown create a pdf document? Default: wanttinytex <- FALSE
wanttinytex <- TRUE

if(tinytex:::is_tinytex()==FALSE && wanttinytex == TRUE){
if (!require("tinytex")) install.packages("tinytex")
tinytex::install_tinytex()
}
```

## Introduction
For this first programming assignment I wrote three functions that are meant to interact with dataset that accompanies this assignment. The dataset is contained in a zip file specdata.zip that is included in the GitHub repository.

## Data

```{bash}
$wget ftp://ftp.ensembl.org/pub/release-84/gff3/homo_sapiens/Homo_sapiens.GRCh38.84.gff3.gz

$wget ftp://ftp.ensembl.org/pub/release-99/gtf/homo_sapiens/Homo_sapiens.GRCh38.99.gtf.gz

$wget ftp://ftp.ensembl.org/pub/release-99/gff3/homo_sapiens/Homo_sapiens.GRCh38.99.gff3.gz

$stringtie input.bam -o output.gtf
$stringtie --merge -p 8 -G Homo_sapiens.GRCh38.84.gff3 -o stringtie_merged.gtf mergelist.txt

intersectBed -a hg38.ultraConserved.bed -b stringtie_merged.gtf -wa -wb > intersect_TUCRs.txt
intersectBed -a stringtie_merged.gtf -b hg38.ultraConserved.bed > intersect_Transcripts.txt
cp stringtie_merged.gtf > stringtie_merged.txt
intersectBed -a genes.bed -b stringtie_merged.gtf -wa -wb > intersect_ptns.txt
intersectBed -a genes.bed -b hg38.ultraConserved.bed -wa -wb > intersect_ptns_TUCR.txt
```


