---
title: 'Uncovering a role for transcribed ultra-conserved regions as long non-coding RNAs in glioblastoma'
author: "Myron Keith Gibert Jr"
date: "January 27, 2020"
output: pdf_document
toc: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r tinytex, include=FALSE}
## INSTALLING LATEX FOR RMARKDOWN

#RMarkdown requires LaTex to create pdf documents as an output. More information can be found [here](https://bookdown.org/yihui/rmarkdown/pdf-document.html). Alternatively, output can be set to "html_document" or "word_document". End users may install LaTex by setting the "wanttinytex" variable to FALSE and running the following script:

#Install tinytex to let RMarkdown create a pdf document? Default: wanttinytex <- FALSE
wanttinytex <- TRUE

if(tinytex:::is_tinytex()==FALSE && wanttinytex == TRUE){
if (!require("tinytex")) install.packages("tinytex")
tinytex::install_tinytex()
}
```

## Introduction
For this first programming assignment I wrote three functions that are meant to interact with dataset that accompanies this assignment. The dataset is contained in a zip file specdata.zip that is included in the GitHub repository.

## Data

```{bash stringtie, eval = FALSE}
wget ftp://ftp.ensembl.org/pub/release-99/gtf/homo_sapiens/Homo_sapiens.GRCh38.99.gtf.gz
mv Homo_sapiens.GRCh38.99.gtf.gz stringtie
gunzip stringtie/Homo_sapiens.GRCh38.99.gtf.gz

module load gcc/7.1.0
module load stringtie/2.0.6

for bam in $(find GBM-RNASEQ-RAW/* -name '*.bam')
do
name=$(echo $bam | awk -F ".bam" '{print $1}')
name2=$(basename "$name")
if [ ! -f stringtie/$name2.gtf ] 
then
    echo "$name2.gtf does not exist"
    stringtie $bam -o $name.gtf
    mv $name.gtf ./stringtie
else
    echo "$name2.gtf exists... skipping"
fi
=======
if [ ! -f $name.gtf ]; then
  stringtie $bam -o $name.gtf
fi
mv $name.gtf ./stringtie
done

cd stringtie

stringtie --merge *rehead.gtf -G Homo_sapiens.GRCh38.99.gtf -o stringtie_merged.gtf

intersectBed -a hg38.ultraConserved.bed -b CHESSgenes.bed -wa -wb > intersectchess_TUCRs.bed

intersectBed -a hg38.ultraConserved.bed -b stringtie_merged.gtf -wa -wb > intersectstringtie_TUCRs.bed

cp stringtie_merged.gtf stringtie_merged.bed
```

```{bash tcga, eval = FALSE}
## Create directory for all outputs from TCGA analysis data
mkdir tcga_analysis/
```

```{bash tcgaheader, eval = FALSE}
## Extract header information from BAM files
# Create directory for extracted headerfiles
mkdir tcga_analysis/head/

# Create extracted header files
for bam in $(find GBM-RNASEQ-RAW/* -name '*.bam')
do
name=$(echo $bam | awk -F ".bam" '{print $1}')
echo $name
samtools view -H $bam > $name.headerids.txt
mv $name.headerids.txt ./tcga_analysis/head/
done
```

```{bash tcgaids, eval = FALSE}
## Get new file names for BAM files using TCGA IDs from extracted header
mkdir tcga_analysis/ids/

# Initialize summary file with TCGA IDs
echo "RAWID,TCGAID" > tcga_analysis/ids/tcgaIDs.csv 

# Extract IDs from header files
for header in $(find tcga_analysis/head/* -name '*headerids.txt')
do
name=$(echo $header | awk -F ".headerids.txt" '{print $1}')
name2=$(echo $name | cut -c20-)
ids1=$(awk '$1=="@RG"{print $3}' $header)
ids2=$(echo $ids1 | cut -c4-)
echo "$name2.bam,$ids2.bam"
echo "$name2.bam,$ids2.bam" > $name.tcgaid.csv
echo "$name2.bam,$ids2.bam" >> tcga_analysis/ids/tcgaIDs.csv 
mv $name.tcgaid.csv ./tcga_analysis/ids/
done
```

```{bash tcgaindex, eval = FALSE}

##Rename BAM files using TCGA ids
# Create folder for new files
mkdir tcga_analysis/reindex/
# Copy BAM files into new directory with updated names

while IFS=, read orig target; do
orig2=$(find GBM-RNASEQ-RAW/* -name $orig)
echo $orig2
cp $orig2 $target
mv $target tcga_analysis/reindex/
done < tcga_analysis/ids/tcgaIDs.csv

# Reindex the renamed files using samtools.
for bam in $(find tcga_analysis/reindex/ -name '*.bam')
do
name=$(echo $bam | awk -F ".bam" '{print $1}')
echo $name
samtools index $bam
done
```

```{bash tcgacounts, eval = FALSE}
mkdir tcga_analysis/foldchange

#Write header line
echo -e chrom"\t"start"\t"end"\t"id tcga_analysis/reindex/*bam | sed -e "s/ /\t/g" > tcga_analysis/foldchange/TUCR_counts.txt

#Counts
multiBamCov -bams tcga_analysis/reindex/*bam -bed tucr_project/hg38.ultraConserved.bed -q 10 >> tcga_analysis/foldchange/TUCR_counts.txt

echo tcga_analysis/reindex/*bam > tcga_analysis/foldchange/TUCR_bamIDs.txt

#Write header line
echo -e chrom"\t"start"\t"end"\t"id *bam | sed -e "s/ /\t/g" > gene_counts.txt 
#Counts
multiBamCov -bams *bam -bed genes.bed -q 10 >> gene_counts.txt

echo tcga_analysis/reindex/*bam > tcga_analysis/foldchange/gene_bamIDs.txt

#Write header line
echo -e chrom"\t"start"\t"end"\t"id *bam | sed -e "s/ /\t/g" > transcript_counts.txt 
#Counts
multiBamCov -bams *bam -bed INPUT.bed -q 10 >> transcript_counts.txt

echo tcga_analysis/reindex/*bam > tcga_analysis/foldchange/transcript_bamIDs.txt
```


