---
title: 'Uncovering a role for transcribed ultra-conserved regions as long non-coding RNAs in gliomas'
author: "Myron Gibert Jr"
date: "January 27, 2020"
output: pdf_document
toc: true
---

```{r tinytex, include=FALSE}
## INSTALLING LATEX FOR RMARKDOWN

#RMarkdown requires LaTex to create pdf documents as an output. More information can be found [here](https://bookdown.org/yihui/rmarkdown/pdf-document.html). Alternatively, output can be set to "html_document" or "word_document". End users may install LaTex by setting the "wanttinytex" variable to FALSE and running the following script:

#Install tinytex to let RMarkdown create a pdf document? Default: wanttinytex <- FALSE
wanttinytex <- TRUE

if(tinytex:::is_tinytex()==FALSE && wanttinytex == TRUE){
if (!require("tinytex")) install.packages("tinytex")
tinytex::install_tinytex()
}
```

## Abstract

Glioblastoma (GBM) is the most common and most deadly malignant brain tumor. Most GBM research has focused on protein-coding genes and much less on the non-coding transcripts that make up 98% of cellular RNA. Transcribed Ultra-Conserved Regions (TUCRs) represent an understudied class of long non-coding RNAs (lncRNAs) that are found conserved across multiple species. These non-coding transcripts are highly resistant to variation and are commonly deregulated in cancer, suggesting regulatory and functional importance. We performed a first time analysis of TCGA RNA-Seq data and identified 197 TUCRs that are differentially expressed relative to normal brain. This project aims to identify and valid these deregulated TUCRs, identify and validate lncRNA transcripts containing these TUCRs, identify genes that are co-regulated with these TUCRs, and predict TUCR function using the functions that are governed by co-regulated genes.

## Introduction

Gliomas account for approximately 80% of all brain tumors.  In particular, glioblastoma (GBM) accounts for about half of all primary brain and central nervous system cancers.  It is also one of the most deadly cancers. One-, five-, and ten-year survival rates for patients with GBM are 37 percent, 5%, and 3%, respectively.  Although there has been extensive research on protein coding genes and pathways in GBM and low grade gliomas (LGGs), the current standard of care involves surgical resection and radiation, with limited targeted therapies.

While many targeted therapies in the context of GBM, LGGs, and other malignancies focus on protein-coding genes, these genes make up only ~1.5% of the ~90% of the genome that is transcribed. The remainder of the transcribed genome is represented by non-coding elements that can serve key regulatory roles. Of these elements, long non-coding RNAs (lncRNAs) have recently been determined to serve as important regulators of malignancy and potential therapeutic targets in cancer. 28-33 These lncRNAs are transcribed RNA molecules that contain greater than 200 nucleotides but typically lack significant protein coding capabilities. 

A subset of these lncRNAs is transcribed from ultra-conserved regions (TUCR) of the genome.  These TUCRs are 481 unique regions of the genome that are greater than 200 base pairs in length and are highly conserved across multiple species, including human, mouse, rat, dog, and chicken genomes. This is significant, as lncRNAs are typically poorly conserved as a class of molecules.  TUCRs are highly resistant to variation, with the 106,767 conserved bases overlapping with as few as six canonical single nucleotide polymorphisms (SNPs).  Their expression is commonly deregulated in cancer.

It is believed that these TUCRs may serve crucial regulatory roles as lncRNAs in cancer. The non-coding RNA field is a growing one.  While only a single lncRNA paper was published in 2007, that number has exploded to 1342 publications in the year 2016.  Despite this explosion of interest in lncRNAs, nothing is known on TUCR lncRNAs in GBM or LGGs as the literature elucidating their expression, functions and mechanisms of action is nonexistent. Studying them therefore represents an untouched avenue for understanding novel oncogenic mechanisms and discovering new biomarkers and therapeutic targets for these diseases.

Here, we use TCGA RNA-Seq GBM and LGG datasets to identify TUCR expression change effects on various parameters.  We identified and validated TUCRs that are deregulated in GBM tissues.  We also identified TUCRs that are correlated with GBM and LGG patient survival, TUCRs that are coregulated with protein coding genes, predicted novel lncRNAs containing TUCRs, TUCR lncRNA secondary structure, and predicted TUCR functions using gene ontology and pathway analyses.  


## Set parameters

This initial section is used to set the dependent variables that are used for this analysis. 

The following variables can be set:
* countfilename is the name of the RNA-Seq feature count file that is generated in the "Generating count tables for survival and differential expression" section.  This can be any tab-delimited text file where the row names are the genes, the column names are the samples, and the cells are the generated counts.  This table is used for both survival and differential expression analyses.
* makekpmplots is a logical vector.  When set to TRUE, the script will generate Kaplan-Meier survival plots for each gene in the counts file.
* makerpkmoxplots is also a logical vector.  When set to TRUE, the script will generate reads per kilobase million boxplots for each gene, showing median RPKM, mean RPKM (red dot), and additional summary information. This is used to determine to what extent a gene is expressed across all samples
* repel is also a logical vector.  When set to true, the volcano summary plot for RNA-Seq expression data will contain labels for a given set of genes specified in the "Generating volcano plot to visualize DE genes"
* outputdir is a character variable containing the "output directory", or the location where all output files will go. This can be any characters enclosed in quotations marks (ex. "Outputs"). Set to "Outputs" to use the default outputs folder
* useintermediate is a logical vector.  When set to TRUE, the analysis will use a pregenerated gene correlation matrix for the guilt by association analysis.  Unless there is a specific reason to generate a new matrix (VERY memory intensive), this variable should be set to TRUE.
* deleteoutputs is a logical vector that should only be set to TURE or FALSE.  When set to true, the output directory will be repopulated after every program run.  If set to false, the program will only run if the output directory does not already exist in the working directory, to prevent the data from being overwritten.

```{r parameters}
disease <- "GBM"

countfilename <- "TUCR_counts.txt"

metadatafile <- "tcga_metadata.csv" 

makekpmplots <- TRUE 

makeRPKMboxplots <- TRUE

repel <- TRUE

outputdir <- "Outputs_GBM"

useintermediate <- TRUE

mergedcounts <- "mergedcounts.txt"

intermediatefile <- "matrixintermediate.Rdata"

```

## Install required programs

```{r setup, message = FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = TRUE,eval=FALSE, warning=FALSE, message=FALSE)

if (!require("tidyverse")) install.packages("tidyverse")
library("tidyverse")

if (!require("ggplot2")) install.packages("ggplot2")
library("ggplot2")

if (!require("ggforce")) install.packages("ggforce")
library("ggforce")

if (!require("ggrepel")) install.packages("ggrepel")
library("ggrepel")

if (!require("ggfortify")) install.packages("ggfortify")
library("ggfortify")

if (!require("ggdendro")) install.packages("ggdendro")
library("ggdendro")

if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
if (!requireNamespace("DESeq2", quietly = TRUE))BiocManager::install("DESeq2")
library("DESeq2")

if (!require("survival")) install.packages("survival")
library("survival")

if (!require("broom")) install.packages("broom")
library("broom")

if (!requireNamespace("limma", quietly = TRUE))BiocManager::install("limma")
library("limma")

if (!require("Tmisc")) install.packages("Tmisc")
library("Tmisc")

if (!require("survminer")) install.packages("survminer")
library("survminer")

if (!require("corrr")) install.packages("corrr")
library("corrr")

if (!require("here")) install.packages("here")
library("here")

if (!require("lessR")) install.packages("lessR")
library("lessR")

if (!require("splitstackshape")) install.packages("splitstackshape")
library(splitstackshape)

if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")

if (!require('org.Hs.eg.db')) BiocManager::install('org.Hs.eg.db')
library(org.Hs.eg.db)

if (!require('GO.db')) BiocManager::install('GO.db')
library(GO.db)

if (!require('limma')) BiocManager::install('limma')
library(limma)

if (!require('factoextra')) BiocManager::install('factoextra')
library(factoextra)

```

## Create output directory

This script checks to see if the defined output directory already exists.  If it doesn't exist, it will create the directory.

```{r outputs}
if(!dir.exists(outputdir)){
  dir.create(outputdir)
}
```

## Stringtie

The hg19 genomic coordinates for each TUCR were acquired from _____________ and were lifted over to hg38 coordinates. The stringtie bioinformatics package was then used to perform a de novo transcript reassembly of all transcripts in 161 GBM, 400 LGG, and 5 normal brain TCGA Cancer RNA-Seq datasets.  First, transcripts present in each RNA-Seq BAM file were compiled into individual BED/GTF format genomic coordinate files using Stringtie.  These files were then merged into one master file containing all transcripts in all samples using Stringtie and an hg38 genome reference file.  The bedtools package was then used to find the genomic coordinates of all transcripts containing ultraconserved regions.  The ________________ group used stringtie to analyze the entirety of the GTex RNA-Seq normal tissue catalog, so bedtools was also used to determine the intersection of TUCRs with their results.  This methodology provides insight into TUCR transcript presence in normal tissues and in brain cancer.

```{bash stringtie, eval = FALSE}
wget ftp://ftp.ensembl.org/pub/release-99/gtf/homo_sapiens/Homo_sapiens.GRCh38.99.gtf.gz
mv Homo_sapiens.GRCh38.99.gtf.gz stringtie
gunzip stringtie/Homo_sapiens.GRCh38.99.gtf.gz

module load gcc/7.1.0
module load stringtie/2.0.6

for bam in $(find GBM-RNASEQ-RAW/* -name '*.bam')
do
name=$(echo $bam | awk -F ".bam" '{print $1}')
name2=$(basename "$name")
if [ ! -f stringtie/$name2.gtf ] 
then
    echo "$name2.gtf does not exist"
    stringtie $bam -o $name.gtf
    mv $name.gtf ./stringtie
else
    echo "$name2.gtf exists... skipping"
fi

if [ ! -f $name.gtf ]; then
  stringtie $bam -o $name.gtf
fi
mv $name.gtf ./stringtie
done

cd stringtie

stringtie --merge *rehead.gtf -G Homo_sapiens.GRCh38.99.gtf -o stringtie_merged.gtf

intersectBed -a hg38.ultraConserved.bed -b CHESSgenes.bed -wa -wb 
  > intersectchess_TUCRs.bed

intersectBed -a hg38.ultraConserved.bed -b stringtie_merged.gtf -wa -wb 
  > intersectstringtie_TUCRs.bed

cp stringtie_merged.gtf stringtie_merged.bed
```

## Setup for TCGA RNA-Seq analysis

The next part of this analysis involves using bash and R scripts to identify TUCRs that are deregulated and/or correlated with survival in GBM and LGG. The GBM analyses were stored in a directory named tcga_analysis, while the LGG analyses were stored in the lgg_tcga_analysis.  The following script generates these two directories.

```{bash tcga, eval = FALSE}
## Create directory for all outputs from TCGA analysis data
mkdir tcga_analysis/

mkdir lgg_tcga_analysis/
```

## Extract headers from BAM files

The raw BAM files for GBM and LGG were stored in GBM-RNASEQ-RAW and LGG-RNASEQ-RAW, respectively.  Since these files are labeled with unintuitive and encrypted names, I will generate clones of each BAM file using the TCGA ID as the name instead.  This will make it easier to see which files are normal brain and which are tumors at a glance.  TCGA files are not named this way by default because there is patient information contained within the BAM files that can be identified using the TCGA ID.  Therefore, renaming the file in secure storage is preferrable.  The first step to doing this is extracting the TCGA ID from the header present in each BAM file. 

In the following series of scripts, a bash script was used to extract the header from each encrypted BAM file and generate a text file containing the full header. 

```{bash tcgaheader, eval = FALSE}
## Extract header information from GBM BAM files
# Create directory for extracted headerfiles
mkdir tcga_analysis/head/

# Use a for loop to cycle through BAM files and extract header before creating a 
#text file containing each BAM file's header.
for bam in $(find GBM-RNASEQ-RAW/* -name '*.bam')
do
name=$(echo $bam | awk -F ".bam" '{print $1}')
echo $name
samtools view -H $bam > $name.headerids.txt
mv $name.headerids.txt ./tcga_analysis/head/
done

## Do the same thing for LGG

mkdir lgg_tcga_analysis/head/

for bam in $(find LGG-RNASEQ-RAW/* -name '*.bam')
do
name=$(echo $bam | awk -F ".bam" '{print $1}')
echo $name
samtools view -H $bam > $name.headerids.txt
mv $name.headerids.txt ./lgg_tcga_analysis/head/
done
```

## Extract TCGA patient barcodes from BAM headers

Once the headers have been extracted from each BAM file, the TCGA ID is then extracted from each header.  A master table is generated that contains the encrypted BAM file name and the new TCGA ID based named, which will be used to generate the renamed clone files in the next step.

```{bash tcgaids, eval = FALSE}
## Get new file names for BAM files using TCGA IDs from extracted header
mkdir tcga_analysis/ids/

mkdir lgg_tcga_analysis/ids/

# Initialize summary file with TCGA IDs
echo "RAWID,TCGAID" > tcga_analysis/ids/tcgaIDs.csv

echo "RAWID,TCGAID" > lgg_tcga_analysis/ids/tcgaIDs.csv 

# Extract IDs from header files
for header in $(find tcga_analysis/head/* -name '*headerids.txt')
do
name=$(echo $header | awk -F ".headerids.txt" '{print $1}')
name2=$(echo $name | cut -c20-)
ids1=$(awk '$1=="@RG"{print $3}' $header)
ids2=$(echo $ids1 | cut -c4-)
echo "$name2.bam,$ids2.bam"
echo "$name2.bam,$ids2.bam" > $name.tcgaid.csv
echo "$name2.bam,$ids2.bam" >> tcga_analysis/ids/tcgaIDs.csv
mv $name.tcgaid.csv ./tcga_analysis/ids/
done

for header in $(find lgg_tcga_analysis/head/* -name '*headerids.txt')
do
name=$(echo $header | awk -F ".headerids.txt" '{print $1}')
name2=$(echo $name | cut -c24-)
ids1=$(awk '$1=="@RG"{print $3}' $header)
ids2=$(echo $ids1 | cut -c4-)
echo "$name2.bam,$ids2.bam"
echo "$name2.bam,$ids2.bam" > $name.tcgaid.csv
echo "$name2.bam,$ids2.bam" >> lgg_tcga_analysis/ids/tcgaIDs.csv 
mv $name.tcgaid.csv ./lgg_tcga_analysis/ids/
done
```

## Reindex TCGA BAM files under new names

```{bash tcgaindex, eval = FALSE}

##Rename BAM files using TCGA ids
# Create folder for new files
mkdir tcga_analysis/reindex/

# Copy BAM files into new directory with updated names

while IFS=, read orig target; do
orig2=$(find GBM-RNASEQ-RAW/* -name $orig)
echo $orig2
cp $orig2 $target
mv $target tcga_analysis/reindex/
done < tcga_analysis/ids/tcgaIDs.csv

while IFS=, read orig target; do
orig2=$(find LGG-RNASEQ-RAW/* -name $orig)
echo $orig2
cp $orig2 $target
mv $target lgg_tcga_analysis/reindex/
done < lgg_tcga_analysis/ids/tcgaIDs.csv

# Reindex the renamed files using samtools.
for bam in $(find tcga_analysis/reindex/ -name '*.bam')
do
name=$(echo $bam | awk -F ".bam" '{print $1}')
echo $name
samtools index $bam
done

# Reindex the renamed files using samtools.
for bam in $(find lgg_tcga_analysis/reindex/ -name '*.bam')
do
samtools view -H TCGA-CS-4938-01B-11R-1896-07.bam > header
name=$(echo $bam | awk -F ".bam" '{print $1}')
echo $name
samtools reheader header $bam > $name.rehead.bam
samtools index $name.rehead.bam
done
```

## Generating count tables for survival and differential expression.

```{bash tcgacounts, eval = FALSE}
mkdir tcga_analysis/foldchange

#Write header line
echo -e chrom"\t"start"\t"end"\t"id tcga_analysis/reindex/*bam | sed -e "s/ /\t/g" 
  > tcga_analysis/foldchange/TUCR_counts_80.txt

#Counts
multiBamCov -bams tcga_analysis/reindex/*bam -bed hg38.ultraConserved.bed -q 10 -f 0.80 
  >> tcga_analysis/foldchange/TUCR_counts_80.txt

#Write header line
echo -e chrom"\t"start"\t"end"\t"id *bam | sed -e "s/ /\t/g" 
  > gene_counts.txt

#Counts
multiBamCov -bams lgg_tcga_analysis/reindex/*bam -bed CHESSgenes.bed -q 10 
  >> gene_counts.txt
```

## Acquiring total reads from RNA-Seq BAM files

```{bash, eval = FALSE}

mkdir tcga_analysis/readcounts/

module load samtools/1.9

for bam in $(find tcga_analysis/reindex/*.bam -name '*.bam')
do
name=$(echo $bam | awk -F ".bam" '{print $1}')
echo $name
samtools view -c -F 260 $bam > $name.readcounts.txt
mv $name.readcounts.txt ./tcga_analysis/readcounts/
done

echo "id,counts" > tcga_analysis/readcounts/seqdepth_counts.csv 

for txt in $(find tcga_analysis/readcounts/* -name '*.txt')
do
name=$(echo $txt | awk -F ".readcounts.txt" '{print $1}')
echo $name
reads=$(head $txt)
echo "$name,$reads" >> tcga_analysis/readcounts/seqdepth_counts.csv 
done

```

## Acquiring clinical survival data

```{bash, eval = FALSE}

wget http://gdac.broadinstitute.org/runs/stddata__2016_01_28/data/GBM/20160128/gdac.broadi
nstitute.org_GBM.Merge_Clinical.Level_1.2016012800.0.0.tar.gz

tar -xvzf gdac.broadinstitute.org_GBM.Merge_Clinical.Level_1.2016012800.0.0.tar.gz

mv gdac.broadinstitute.org_GBM.Merge_Clinical.Level_1.2016012800.0.0 GBM.Merge_Clinical

```

## Acquiring SNPs

```{bash, eval = FALSE}
mkdir SNPs

cd SNPs

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_1.bed.gz
gunzip bed_chr_1.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_10.bed.gz
gunzip bed_chr_10.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_11.bed.gz
gunzip bed_chr_11.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_13.bed.gz
gunzip bed_chr_13.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_12.bed.gz
gunzip bed_chr_12.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_14.bed.gz
gunzip bed_chr_14.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_15.bed.gz
gunzip bed_chr_15.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_16.bed.gz
gunzip bed_chr_16.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_17.bed.gz
gunzip bed_chr_17.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_18.bed.gz
gunzip bed_chr_18.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_19.bed.gz
gunzip bed_chr_19.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_2.bed.gz
gunzip bed_chr_2.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_20.bed.gz
gunzip bed_chr_20.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_21.bed.gz
gunzip bed_chr_21.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_22.bed.gz
gunzip bed_chr_22.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_3.bed.gz
gunzip bed_chr_3.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_4.bed.gz
gunzip bed_chr_4.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_5.bed.gz
gunzip bed_chr_5.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_6.bed.gz
gunzip bed_chr_6.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_7.bed.gz
gunzip bed_chr_7.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_8.bed.gz
gunzip bed_chr_8.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_9.bed.gz
gunzip bed_chr_9.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_X.bed.gz
gunzip bed_chr_X.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_Y.bed.gz
gunzip bed_chr_Y.bed.gz

module load gcc/7.1.0
module load bedtools/2.26.0

for bed in *.bed
do
echo $bed
intersectBed -a hg38.ultraConserved.bed -b $bed -wa -wb > $bed.txt
done

cat *.txt > merged.txt
```

## Generate figure for TUCR annotation

```{r annotation}

tucr_annot <- read.csv("tucr_annot.csv")

plot_annotation <- tucr_annot %>% 
  group_by(annot) %>%
  summarise(count = n()) %>%
  mutate(percent = count/sum(count)) %>%
  arrange(desc(count)) %>%
  mutate(lab.ypos = cumsum(count) - 0.5*count)

names(plot_annotation)[names(plot_annotation) == 'annot'] <- 
  'Annotation'
  
p <- ggplot(plot_annotation, aes(x=2,y=count,fill=Annotation))          + 
  geom_bar(
    stat="identity", 
    width = 1)                                                          +
  geom_text(
    aes(y = lab.ypos, label = count), 
    color = "black")                                                    + 
  coord_polar("y", start=0)                                             + 
  scale_fill_manual(values=c("#F8766D","#00BFC4","#7CAE00","#C77CFF"))  + 
  ggtitle(paste("TUCR Genomic Location"))                               + 
  theme(
    plot.title = element_text(size=rel(1.5),
    face="bold",hjust = 0.5,vjust=-2),
    legend.title.align=0.5, 
    axis.title = element_blank(),
    axis.text.x=element_blank(),
    axis.text.y=element_blank(),
    axis.line=element_blank(),
    panel.grid.major = element_blank(), 
    panel.grid.minor = element_blank(),
    axis.ticks=element_blank(),
    panel.background = element_blank())                                 + 
  xlim(0.5, 2.5)

p

ggsave(file=paste(outputdir,"/tucr_annotation.png",sep=""),
       plot = print(p), 
       width = 5, 
       height = 5, 
       dpi = 300)
```

## Show that TUCRs are enriched for hallmarks of lncRNAs

```{r lncRNAs}
tucr_lncRNAs <- read.csv("tucr_lncRNAs.csv")
  
p <- ggplot(tucr_lncRNAs, aes(x=row,y=value,fill=row))                + 
  geom_bar(
    stat="identity", 
    width = 0.7)                                                      + 
  geom_text(
    aes(label=round(value,2)), 
    vjust=1.6, 
    color="black", 
    size=rel(4))                                                      + 
  ggtitle(paste("TUCRs are enriched for \n RNA Pol.II and H3K4me3"))  + 
  xlab("Interval and Mark")                                           + 
  ylab("-log10 of p-value")                                           + 
  theme(
    plot.title = element_text(
      size=rel(1.5), 
      face="bold",hjust = 0.5),
    axis.title = element_text(
      size = rel(1.25), 
      face="bold"),
    axis.text.x = element_text(
      angle = -90, 
      size = 10),
    legend.title = element_blank(),
    legend.position = "none",
    panel.grid.major = element_blank(), 
    panel.grid.minor = element_blank(),
    panel.background = element_blank(), 
    axis.line = element_line(
      colour = "black"))

ggsave(
    file=paste(outputdir,"/tucr_lncRNAs.png",sep=""),
    plot = print(p),
    width = 5, 
    height = 5, 
    dpi = 300) 

```

## Identifying differentially expressed TUCRs with DESeq2

```{r DESeq2_TUCRs}

tucrcounts <- 
  read.table(countfilename,header = TRUE)

metadata <- 
  read_csv(file = metadatafile)

tucrcounts2 <- 
  tucrcounts[,5:ncol(tucrcounts)]

rownames(tucrcounts2) <- 
  tucrcounts[,4]

tucrcounts.info <- 
  tucrcounts[,1:4]

tucrcounts.info$length <- 
  (tucrcounts$end - tucrcounts$start)/1000

tucrcounts <- 
  cbind(tucrcounts.info,tucrcounts2)


dds_TUCR <- 
  DESeqDataSetFromMatrix(countData = tucrcounts2,
                              colData = metadata,
                              design = ~dex)
dds_TUCR <- 
  DESeq(dds_TUCR)

res_TUCR <- 
  results(dds_TUCR, tidy=TRUE)

res_TUCR <- 
  as_tibble(res_TUCR)

write.csv(res_TUCR, file = paste(outputdir,"/results_TUCR.csv",sep=""))

res_TUCR_summary <- 
  res_TUCR %>%
  mutate(dereg = 
           ifelse(res_TUCR$log2FoldChange >=1, "upregulated",
                    ifelse(res_TUCR$log2FoldChange <=-1,"downregulated","unchanged"))) %>%
  filter(dereg == "upregulated" | dereg == "downregulated") %>%
  group_by(dereg) %>%
  summarise(n = n())

p <- ggplot(res_TUCR_summary, aes(x=dereg,y=n,fill=dereg)) + 
  geom_bar(stat="identity", width = 0.7) + 
  geom_text(aes(label=n), vjust=1.6, color="black", size=7) + 
  scale_fill_manual(values=c("green", "red")) + 
  ggtitle(paste("TUCR Deregulation in ",disease)) + 
  xlab("Direction of Deregulation") + 
  ylab("Number of Deregulated Genes") + 
  theme(
    plot.title = element_text(size=rel(1.5), face="bold",hjust = 0.5),
    axis.title = element_text(size = rel(1.25), face="bold"),
    legend.title = element_blank(),
    legend.position = "none",
    panel.grid.major = element_blank(), 
    panel.grid.minor = element_blank(),
    panel.background = element_blank(), 
    axis.line = element_line(colour = "black"))

ggsave(file=paste(outputdir,"/",disease,"_deregGraph.png",sep=""),
       plot = print(p),
       width = 5, 
       height = 5, 
       dpi = 300)

```

## Writing a script to generate a volcano plot

```{r volcanoplot}

### Volcano plot

volcanoplot <- function (res,
                         genes = "all",
                         title = paste("Deregulated Genes in ",disease,sep=""), 
                         output = "results_volcanoplot.png",
                         height = 5, 
                         width = 6, 
                         dpi = 300){
  #res <- res_TUCR
  #genes = c("uc.110","uc.62")
  #title = "Deregulated Genes"
  vres <- res
  vres <- vres %>% 
    mutate(filter1 = abs(log2FoldChange)>=1,filter2 = padj <=0.05,Legend="",gene="")
  
  for (i in 1:length(vres$row)){
    if(is.na(vres$padj[i])){vres$Legend[i] <- "Not deregulated"}else{
    if(vres$filter1[i] == TRUE && vres$filter2[i] == TRUE){
    vres$Legend[i] <- ">2-Fold & <0.05 FDR"
  }else{
      if(vres$filter1[i] == TRUE){
    vres$Legend[i] <- ">2-Fold"
      }else{vres$Legend[i] <- "Not deregulated"}
  }}
    ifelse(genes!="all",
           ifelse(!is.na(
             match(vres$row[i],genes)),vres$gene[i] <- vres$row[i], ""),
           vres$gene[i] <- vres$row[i])
    }
  
  ## plot
if(repel==TRUE){p <- ggplot(vres) 
+ geom_point(aes(x=log2FoldChange, y=-log10(padj),col=Legend)) 
+ scale_color_manual(values = c("blue","red","black")) 
+ ggtitle(title) 
+ xlab("log2 fold change") 
+ ylab("-log10 adjusted p-value") 
+ theme(plot.title = element_text(size = rel(1.5), hjust = 0.5, face="bold"),
              axis.title = element_text(size = rel(1.25), face="bold"),
panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black")) 
+ geom_text_repel(aes(x=log2FoldChange, y=-log10(padj),label = gene),force=50)}
else{p <- ggplot(vres) +
        geom_point(aes(x=log2FoldChange, y=-log10(padj),col=Legend)) +
        scale_color_manual(values = c("blue","red","black")) +
        ggtitle(title) +
        xlab("log2 fold change") + 
        ylab("-log10 adjusted p-value") +
        theme(plot.title = element_text(size = rel(1.5), hjust = 0.5, face="bold"),
              axis.title = element_text(size = rel(1.25), face="bold"),
panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black"))}
ggsave(output, width = width, height = height, dpi = dpi)
}

```

## Generating volcano plot to visualize DE genes

```{r DESeq2plots_TUCRs}

volcanoplot(res_TUCR,genes=c("uc.110","uc.210","uc.427","uc.62"),
            title = paste("Deregulated Genes in ",disease,sep=""), 
            output = paste(outputdir,"/myron_tucr_results_volcanoplot.png",sep=""))

repel_hold = repel

repel = FALSE

volcanoplot(res_TUCR,
            title = paste("Deregulated Genes in ",disease,sep=""), 
            output = paste(outputdir,"/tucr_results_volcanoplot.png",sep=""))

repel = repel_hold

```

## Volcano plot for intergenic TUCRs only

```{r DESeq2plots_intergenicTUCRs}

tucr_annot <- read.csv("tucr_annot.csv")

res_intergenic <- inner_join(res_TUCR,tucr_annot,by = "row") %>%
  filter(annot == "intergenic")

write.csv(res_intergenic, file = paste(outputdir,"/results_intergenicOnly.csv",sep=""))

volcanoplot(res_intergenic,genes=c("uc.62","uc.110"), 
            title = paste("Intergenic TUCR deregulation in ",disease,sep=""),
            output = paste(outputdir,"/intergenic_tucr_results_volcanoplot.png",sep=""))

```

## Completing survival analysis for TUCRs

```{r surcor_TUCRs}
if(!file.exists("GBM.clin.merged.txt")){unzip("GBM.clin.merged.zip")}

countdata <- read.table(countfilename,header = TRUE)
metadata <- read_csv(file = metadatafile)

posdata <- countdata[,1:4]
countdata <- countdata[,5:length(colnames(countdata))]

countdata <- as.matrix(countdata)

n_index <- c(seq(1,5))
t_index <- c(seq(6,ncol(countdata)))

vm <- function(x){
  cond <- factor(ifelse(seq(1,dim(x)[2],1) %in% t_index, 1,  0))
  d <- model.matrix(~1+cond)
  x <- t(apply(x,1,as.numeric))
  ex <- voom(x,d,plot=F)
  return(ex$E)
}

count_vm <- vm(countdata)
colnames(count_vm) <- metadata$barcode

scal <- function(x,y){
  mean_n <- rowMeans(y)  # mean of normal
  sd_n <- apply(y,1,sd)  # SD of normal
  # z score as (value - mean normal)/SD normal
  res <- matrix(nrow=nrow(x), ncol=ncol(x))
  colnames(res) <- colnames(x)
  rownames(res) <- rownames(x)
  for(i in 1:dim(x)[1]){
    for(j in 1:dim(x)[2]){
      res[i,j] <- (x[i,j]-mean_n[i])/sd_n[i]
    }
  }
  return(res)
}

z_rna <- scal(count_vm[,t_index],count_vm[,n_index])
rownames(z_rna) <- posdata[,4]

clinical <- read.table("GBM.clin.merged.txt",header = TRUE)
#clinical <- read.table("GBM.oncolnc.merged.txt",header = TRUE)

clinical$time <- as.numeric(clinical$time)

sum(clinical$patient %in% colnames(z_rna))

ind_tum <- which(unique(colnames(z_rna)) %in% clinical$patient)
ind_clin <- which(clinical$patient %in% colnames(z_rna))

out.tab <- c()
for(x in 1:nrow(count_vm)){
  ind_gene <- x
  s <- Surv(clinical$time[ind_clin],clinical$status[ind_clin])
  cx <- coxph(formula = s ~ count_vm[ind_gene,ind_tum])
  cx <- tidy(cx)
  out.tab <- rbind(out.tab,cx)
}

surv_TUCR <- cbind(posdata,out.tab)
surv_TUCR <- surv_TUCR %>% dplyr::select(-term)

write_csv(surv_TUCR,paste(outputdir,"/survival_TUCRs.csv",sep=""))

if(makekpmplots == TRUE){

if(!dir.exists(paste(outputdir,"/tucrKPMplots",sep=""))){
  dir.create(paste(outputdir,"/tucrKPMplots",sep=""))
}

km_countdata <- countdata
colnames(km_countdata) <- metadata$barcode
posdata$median <- rowMedians(countdata)

probs <- c(0.25,0.5,0.75)
q <- rowQuantiles(countdata, probs = probs)
posdata$n25 <- q[,1]
posdata$n75 <- q[,3]
km_TUCRs <- cbind(posdata,km_countdata)
km_TUCRs <- km_TUCRs[,4:ncol(km_TUCRs)]
km_TUCRs <- km_TUCRs %>%
  gather(key = "sample", value = "count",-id,-median,-n75,-n25) %>%
  mutate(group = ifelse(count>=n75,"high",ifelse(count<=n25,"low",NA))) %>%
  dplyr::select(id,median,"patient"=sample,group) %>%
  left_join(clinical,by="patient") %>%
  dplyr::filter(median != 0)

TUCRids <- posdata$id
i = 1

ptable <- data.frame(matrix(ncol=3,nrow=0, dimnames=list(NULL, c("id","pvalue","method"))))

for(i in 1:length(TUCRids)){
  print(i)
  skip_to_next <- FALSE
  TUCR <- TUCRids[i]
  #TUCR <- "uc.1"
  kmdata <- km_TUCRs %>%
    dplyr::filter(id == TUCR)
  fit <- survfit(Surv(time, status) ~ group, data = kmdata)
  p <- tryCatch(ggsurvplot(fit,data=kmdata,conf.int = TRUE,pval = TRUE,risk.table = TRUE),
                error = function(e) { skip_to_next <<- TRUE})
  if(skip_to_next) { 
    rbinder <- cbind(as.character(TUCR),NA,NA)
    ptable[i,] <- rbinder
    next } 
  ggsave(file=paste(outputdir,"/tucrKPMplots/",TUCR,"_kpmplot.png",sep = ""),
         plot = print(p))
    p2value <- surv_pvalue(fit,  data = kmdata,method = "survdiff")
  rbinder <- cbind(as.character(TUCR),p2value$pval,p2value$method)
  ptable[i,] <- rbinder
}}

write_csv(ptable,paste(outputdir,"/kpm_summary_TUCRs.csv",sep=""))

```

## Summarize Survival Data 

```{r summary surv}

surv_TUCR <- read.csv(paste(outputdir,"/survival_TUCRs.csv",sep=""),header=TRUE)

ptable <- read.csv(paste(outputdir,"/kpm_summary_TUCRs.csv",sep=""),header=TRUE)

res_surv <- inner_join(ptable,surv_TUCR,by="id")

res_surv_sum <- res_surv

res_surv_sum <- res_surv_sum %>% mutate(gene="",Survival="")

for (i in 1:length(res_surv_sum$id)){
  ifelse(is.na(res_surv_sum$pvalue[i]),res_surv_sum$pvalue[i] <- 1,res_surv_sum$pvalue[i] 
         <- res_surv_sum$pvalue[i])
    
  ifelse(genes!="all",ifelse(!is.na(match(res_surv_sum$id[i],genes)),
                             res_surv_sum$gene[i] <- as.character(res_surv_sum$id[i]), ""),
                                  res_surv_sum$gene[i] <- res_surv_sum$id[i])
  
  ifelse((res_surv_sum$pvalue[i] <0.05 & res_surv_sum$p.value[i] <0.05 & 
            res_surv_sum$estimate[i] < 0),
         res_surv_sum$Survival[i] <- "Significant (Both, Good Prognosis)",
  ifelse((res_surv_sum$pvalue[i] <0.05 & res_surv_sum$p.value[i] <0.05 &
            res_surv_sum$estimate[i] > 0),
         res_surv_sum$Survival[i] <- "Significant (Both, Poor Prognosis)",
  ifelse((res_surv_sum$pvalue[i] >0.05 & res_surv_sum$p.value[i] <0.05 &
            res_surv_sum$estimate[i] < 0),
         res_surv_sum$Survival[i] <- "Significant (CH, Good Prognosis)",
  ifelse((res_surv_sum$pvalue[i] >0.05 & res_surv_sum$p.value[i] <0.05 &
            res_surv_sum$estimate[i] > 0),
         res_surv_sum$Survival[i] <- "Significant (CH, Poor Prognosis)",
  ifelse((res_surv_sum$pvalue[i] <0.05 & res_surv_sum$p.value[i] >0.05),
         res_surv_sum$Survival[i] <- "Significant (KM)",
  ifelse(res_surv_sum$Survival[i] <- "Not significant"))))))}

res_surv_sum <- res_surv_sum %>%
  group_by(Survival) %>%
  summarise(count = n())

write.csv(res_surv_sum,file=paste(outputdir,"/tucr_surv_table.csv",sep=""))

p <- ggplot(res_surv_sum, aes(x=Survival,y=count,fill=Survival)) + geom_bar(stat="identity", width = 0.7) + scale_fill_manual(values = c("gray","#619cff","#f8766d","#00ba38","#c77cff","black")) + ylab("# of TUCRs") + theme(
plot.title = element_text(size=rel(1.5), face="bold",hjust = 0.5),
axis.title = element_text(size = rel(1.25), face="bold"),
panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black"),
axis.text.x = element_blank())

ggsave(file=paste(outputdir,"/tucr_surv_bar.png",sep=""),plot = print(p), width = 5, height = 5, dpi = 300) 

```


## Writing a script to generate a volcano plot for survival

```{r volcanosurv}

### Volcano plot

volcanosurv <- function (res,genes = "all",title = paste("TUCRs correlated with survival in ",disease,sep=""), output = "results__survival_volcanoplot.png",height = 5, width = 7.5, dpi = 300){
  #res <- res_surv
  #genes = c("uc.110","uc.62")
  #title = paste("TUCR correlation with patient survival in ",disease,sep="")
  #output = paste(outputdir,"/intergenic_tucr_results_volcanosurv.png",sep="")
  i = 9
  vres <- res
  vres <- vres %>% mutate(gene="",Survival="")
  for (i in 1:length(vres$id)){
  ifelse(is.na(vres$pvalue[i]),vres$pvalue[i] <- 1,vres$pvalue[i] <- vres$pvalue[i])
    
  ifelse(genes!="all",ifelse(!is.na(match(vres$id[i],genes)),vres$gene[i] <- as.character(vres$id[i]), ""),vres$gene[i] <- vres$id[i])
  
  ifelse((vres$pvalue[i] <0.05 & vres$p.value[i] <0.05 & vres$estimate[i] < 0),vres$Survival[i] <- "Significant (Both, Good Prognosis)",
  ifelse((vres$pvalue[i] <0.05 & vres$p.value[i] <0.05 & vres$estimate[i] > 0),vres$Survival[i] <- "Significant (Both, Poor Prognosis)",
  ifelse((vres$pvalue[i] >0.05 & vres$p.value[i] <0.05 & vres$estimate[i] < 0),vres$Survival[i] <- "Significant (CH, Good Prognosis)",
  ifelse((vres$pvalue[i] >0.05 & vres$p.value[i] <0.05 & vres$estimate[i] > 0),vres$Survival[i] <- "Significant (CH, Poor Prognosis)",
  ifelse((vres$pvalue[i] <0.05 & vres$p.value[i] >0.05),vres$Survival[i] <- "Significant (KM)",
  ifelse(vres$Survival[i] <- "Not significant"))))))}

  
  ## plot
if(repel==TRUE){p <- ggplot(vres) +
        geom_point(aes(x=estimate, y=-log10(as.numeric(p.value)),col=Survival)) +
        scale_color_manual(values = c("gray","#619cff","#f8766d","#00ba38","#c77cff","black")) +
        ggtitle(title) +
        xlab("Cox Estimated Proportional Hazard") + 
        ylab("-log10 P-Value (CH)") +
        theme(plot.title = element_text(size = rel(1.5), hjust = 0.5, face="bold"),
              axis.title = element_text(size = rel(1.25), face="bold"),
panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black")) +
        geom_text_repel(aes(x=estimate, y=-log10(as.numeric(p.value)),label = gene),force=50)
}else{
  p <- ggplot(vres) +
        geom_point(aes(x=estimate, y=-log10(as.numeric(pvalue)),col=Survival)) +
    scale_color_manual(values = c("gray","#619cff","#f8766d","#00ba38","#c77cff","black")) +
        + ggtitle(title) +
        xlab("Cox Estimated Proportional Hazard") + 
        ylab("-log10 P-Value (CH)") +
        theme(plot.title = element_text(size = rel(1.5), hjust = 0.5, face="bold"),
              axis.title = element_text(size = rel(1.25), face="bold"),
panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black"))}
ggsave(output, width = width, height = height, dpi = dpi)
}

```

## Draw Volcano plot for TUCRs

```{r volcanosurv TUCRs}

volcanosurv(res_surv,genes=c("uc.174","uc.325","uc.363"), title = paste("TUCR correlation with patient survival in ",disease,sep=""),output = paste(outputdir,"/intergenic_tucr_results_volcanosurv.png",sep=""))

```


## Generate RPKMs for filtering

```{r RPKM_TUCRs}

rpkmcounts <- tucrcounts[,6:ncol(tucrcounts)]
rpkmcounts.info <- tucrcounts[,1:5]
rownames(rpkmcounts) <- tucrcounts$id

genelength <- tucrcounts$length

seqdepth <- read.csv(file = "seqdepth_counts.csv",row.names=1)

seqdepth <- as.vector(seqdepth$counts)

rpkm <- function(counts,len,dep){
  x <- counts/len
  return(t(t(x)/dep))
}

rpkm.TUCRs <- rpkm(rpkmcounts,genelength,seqdepth)
rpkm.TUCRs2 <- cbind(rpkmcounts.info,rpkm.TUCRs)

rpkm.median <- rpkm.TUCRs2 %>% dplyr::summarise(median_RPKM = rowMedians(rpkm.TUCRs))
rpkm.median <- cbind(rpkmcounts.info,rpkm.median)
rpkm.median <- rpkm.median %>% dplyr::select(id,median_RPKM)

write.csv(rpkm.TUCRs2,paste(outputdir,"/rpkm.TUCRs.csv",sep=""))

write.csv(rpkm.median,paste(outputdir,"/medianrpkm.TUCRs.csv",sep=""))
```

## Generate RPKM Box Plots for each TUCR

```{r boxplot_TUCRs}
if(makeRPKMboxplots == TRUE){
  
if(!dir.exists(paste(outputdir,"/RPKMboxplots",sep=""))){
  dir.create(paste(outputdir,"/RPKMboxplots",sep=""))
}
rpkm.dotplot <- rpkm.TUCRs2[,6:ncol(rpkm.TUCRs2)]
TUCRids <- rpkm.TUCRs2[,4]
rpkm.dotplot <- cbind(TUCRids,rpkm.dotplot)

i = 1
for(i in 1:length(TUCRids)){
#TUCR <- "uc.110"
TUCR <- TUCRids[i]
print(i)
rpkm.dotplot2 <- rpkm.dotplot %>% 
  gather(key = "Sample", value = "RPKM",-TUCRids) %>%
  dplyr::filter(TUCRids == TUCR)

p<- ggplot(rpkm.dotplot2, aes(x=TUCRids, y=RPKM)) +
  geom_boxplot() +
  geom_dotplot(binaxis='y', stackdir='center',
               stackratio=0.75, dotsize=0.20) + 
  ggtitle(paste("RPKM expression of ",TUCR)) +
  theme(plot.title = element_text(hjust = 0.5),
panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  labs(y="Reads per kilobase million (RPKM)", x = "TUCR") +
  stat_summary(fun.y=mean, geom="point", shape=20, size=10, color="red", fill="red")

ggsave(p,file=paste(outputdir,"/RPKMboxplots/",TUCR,"_rpkm_boxplot.png",sep = ""))
}

}
```

## Generate RPKM heatmap for all TUCRs

```{r heatmap_TUCRs}
heatmap.TUCRs <- rpkm.TUCRs2[,6:ncol(rpkm.TUCRs2)]
TUCRids <- rpkm.TUCRs2[,4]
means <- apply(heatmap.TUCRs,1,mean)

heatmap.TUCRs2 <- as.data.frame(apply(heatmap.TUCRs,1:2,function(x) {ifelse(x>=10,10,x)}))
heatmap.TUCRs2 <- cbind(TUCRids,means,heatmap.TUCRs2)

heatmap.TUCRs2 <- heatmap.TUCRs2 %>% 
  gather(key = "Sample", value = "RPKM",-TUCRids,-means)

p <- ggplot(data = heatmap.TUCRs2,mapping = aes(x = Sample,y = reorder(TUCRids,means), fill = RPKM)) +
  geom_tile() +
  scale_fill_gradientn(colours = c("blue", "white", "red"), values = c(0,0.1,1)) + ggtitle(paste("Expression of TUCRs in ",disease,sep="")) + xlab(paste("TCGA RNA-Seq Samples (n = ",ncol(heatmap.TUCRs),")",sep="")) + ylab(paste("TUCRs ordered by expression (n = ",nrow(heatmap.TUCRs),")",sep="")) +
  theme(plot.title = element_text(hjust = 0.5),
        panel.grid = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank())

ggsave(p,file=paste(outputdir,"/TUCR_heatmap.png",sep=""))
```

## Identify DE TUCRs that are correlated with each other

```{r correlations}

if(!dir.exists(paste(outputdir,"/correlations",sep=""))){
  dir.create(paste(outputdir,"/correlations",sep=""))
}

if(useintermediate==FALSE){
TUCRids <- read.table(countfilename,header = TRUE)
TUCRids <- as.character(TUCRids[,4])
normcounts <- read.table("normal_mergedcounts.txt",header = TRUE)
expcounts <- read.table(mergedcounts,header = TRUE)
posdata <- normcounts[,1:4]
normcounts <- normcounts[4:ncol(normcounts)]
expcounts <- expcounts[4:ncol(expcounts)]
tucrcounts <- inner_join(posdata,normcounts,by="id")
tucrcounts <- inner_join(tucrcounts,expcounts,by="id")

metadata <- read_csv(file = metadatafile)



tucr.cor <- tucrcounts[,5:ncol(tucrcounts)]
rownames(tucr.cor) <- as.character(tucrcounts[,4])

posdata <- tucrcounts[,1:4]

countdata <- as.matrix(tucr.cor)

n_index <- c(seq(1,5))
t_index <- c(seq(6,ncol(countdata)))

vm <- function(x){
  cond <- factor(ifelse(seq(1,dim(x)[2],1) %in% t_index, 1,  0))
  d <- model.matrix(~1+cond)
  x <- t(apply(x,1,as.numeric))
  ex <- voom(x,d,plot=F)
  return(ex$E)
}

count_vm <- vm(countdata)
colnames(count_vm) <- metadata$barcode
posdata$median <- rowMedians(as.matrix(count_vm)) 

tucr.cor <- cbind(posdata[,4:5],count_vm)

tucr.cor.matrix <- data.frame(row.names = posdata[,4])


i=3

for(i in 3:ncol(tucr.cor)){
  print(i)
  tucr.cor.matrix[,i-2] <- tucr.cor[,i]/tucr.cor[,2]


colnames(tucr.cor.matrix) <- metadata$id
tucr.cor.matrix2 <- t(as.data.frame(apply(as.data.frame(tucr.cor.matrix),1:2,function(x) {ifelse(x>=1,x,-(1/x))})))
tucr.cor.matrix2 <- as.data.frame(tucr.cor.matrix2)
tucr.cor.matrix2 <- tucr.cor.matrix2[1:180,]
geneindex <- rownames(tucr.cor.matrix)}

save(tucr.cor.matrix2,geneindex,TUCRids,file="inter.rData")
rm(list=ls())

load(file="inter.rData")

tucr.cor.matrix3 <- cor(tucr.cor.matrix2,method="spearman", use = "complete.obs")

save(tucr.cor.matrix3,geneindex,TUCRids,file=intermediatefile)
rm(list=ls())
load(file=intermediatefile)}else{
  load("matrixintermediate.rData")
}

header <- head(tucr.cor.matrix3)

summarycor <- data.frame(searchTUCR=character(0),ngenes=integer(0),meancor=numeric(0),mediancor=numeric(0),mincor=numeric(0),maxcor=numeric(0))

i = 1

for(i in 1:length(TUCRids)){
  print(i)
  searchTUCR <- TUCRids[i]
  #searchTUCR <- "uc.110"
###uc.110 is at row 20972 in tucr.cor.matrix3 via manual confirmation.
tucr.cor.matrix.match <- tucr.cor.matrix3[match(searchTUCR,geneindex),]
tucr.cor.matrix.match <- as.data.frame(tucr.cor.matrix.match)
tucr.cor.matrix.match.plus <- tucr.cor.matrix.match %>% 
  mutate(partner = geneindex) %>%
  mutate(gene = searchTUCR) %>%
  dplyr::select(gene,partner,"cor"=tucr.cor.matrix.match) %>%
  dplyr::filter(cor>=0.3 & partner != searchTUCR) %>%
  dplyr::mutate(direction = "negative")

write.csv(tucr.cor.matrix.match.plus,
          paste(outputdir,"/correlations/",searchTUCR,"_plus_correlations.csv",sep=""))

tucr.cor.matrix.match.minus <- tucr.cor.matrix.match %>% 
  mutate(partner = geneindex) %>%
  mutate(gene = searchTUCR) %>%
  dplyr::select(gene,partner,"cor"=tucr.cor.matrix.match) %>%
  dplyr::filter(cor<=-0.3 & partner != searchTUCR) %>%
  dplyr::mutate(direction = "negative")

write.csv(tucr.cor.matrix.match.minus,
          paste(outputdir,"/correlations/",searchTUCR,"_minus_correlations.csv",sep=""))

meancor <- mean(tucr.cor.matrix.match$cor,na.rm=TRUE)
mediancor <- median(tucr.cor.matrix.match$cor)
ngenes <- length(tucr.cor.matrix.match$cor)
mincor <- min(tucr.cor.matrix.match$cor)
maxcor <- max(tucr.cor.matrix.match$cor)
newrow <- cbind(searchTUCR,ngenes,meancor,mediancor,mincor,maxcor)
summarycor <- rbind(summarycor,newrow)}

summarycor %>%
readr::write_excel_csv(here::here(paste(outputdir,"/correlations",sep=""),paste("cor_summary.csv",sep="")))

```

## Generate GO Terms for coregulated genes

```{r GO Terms}
 
filenames <- list.files(paste(outputdir,"/correlations",sep=""))

if(!dir.exists(paste(outputdir,"/GO_terms",sep=""))){
  dir.create(paste(outputdir,"/GO_terms",sep=""))
}

if(!dir.exists(paste(outputdir,"/GO_terms_bp",sep=""))){
  dir.create(paste(outputdir,"/GO_terms_bp",sep=""))
}

if(!dir.exists(paste(outputdir,"/GO_terms_mf",sep=""))){
  dir.create(paste(outputdir,"/GO_terms_mf",sep=""))
}

if(!dir.exists(paste(outputdir,"/GO_terms_cc",sep=""))){
  dir.create(paste(outputdir,"/GO_terms_cc",sep=""))
}

allgo <- data.frame(matrix(ncol=7,nrow=0, dimnames=list(NULL, c("GO Term","Ont","N","DE","P.DE","direction","gene"))))

allgo_bp <- data.frame(matrix(ncol=7,nrow=0, dimnames=list(NULL, c("GO Term","Ont","N","DE","P.DE","direction","gene"))))

allgo_mf <- data.frame(matrix(ncol=7,nrow=0, dimnames=list(NULL, c("GO Term","Ont","N","DE","P.DE","direction","gene"))))

allgo_cc <- data.frame(matrix(ncol=7,nrow=0, dimnames=list(NULL, c("GO Term","Ont","N","DE","P.DE","direction","gene"))))

i = 2

for(i in 2:length(filenames)){
  print(i)
  skip_to_next <- FALSE
  df <- read.csv(paste(outputdir,"/correlations/",filenames[i],sep=""))
  gene <- as.character(df$gene[1])
  
  if(length(df[,1] > 0)){
  
  df2 <- cSplit(df, 'partner', sep="___", type.convert=FALSE)
  
  df2 <- df2 %>% arrange(desc(cor))
  
  IDs <- df2$partner_1
  
  entrezIDs <- tryCatch(mapIds(org.Hs.eg.db, IDs, 'ENTREZID', 'SYMBOL'), error = function(e) { skip_to_next <<- TRUE})
  
  if(skip_to_next) { next } 

  df2$entrez <- entrezIDs
  if(str_detect(filenames[i],"minus")){direction <- -1}else{
    if(str_detect(filenames[i],"plus")){direction <- 1}else{direction <- 0}
  }
  
  Genes<-entrezIDs
  g <- goana(Genes)
  topGO <- topGO(g) 
  
  topGO_all <- topGO %>%
    mutate(direction = direction) %>%
    mutate(gene = gene)
  
  write.csv(topGO_all,paste(outputdir,"/GO_terms/",filenames[i],sep=""))
  topGO_bp <- topGO %>%
    filter(Ont == "BP") %>%
    mutate(direction = direction) %>%
    mutate(gene = gene)
  
  write.csv(topGO_bp,paste(outputdir,"/GO_terms_bp/",filenames[i],sep=""))
  topGO_mf <- topGO %>%
    filter(Ont == "MF") %>%
    mutate(direction = direction) %>%
    mutate(gene = gene)
  
  write.csv(topGO_mf,paste(outputdir,"/GO_terms_mf/",filenames[i],sep=""))
  topGO_cc <- topGO %>%
    filter(Ont == "CC") %>%
    mutate(direction = direction) %>%
    mutate(gene = gene)
  
  write.csv(topGO_cc,paste(outputdir,"/GO_terms_cc/",filenames[i],sep=""))
  }else{}
  allgo <- rbind(allgo,topGO)
  allgo_bp <- rbind(allgo_bp,topGO_bp)
  allgo_mf <- rbind(allgo_mf,topGO_mf)
  allgo_cc <- rbind(allgo_cc,topGO_cc)
}

write_csv(allgo,paste(outputdir,"/GOterm_all_summary_TUCRs.csv",sep=""))

write_csv(allgo_bp,paste(outputdir,"/GOterm_bp_summary_TUCRs.csv",sep=""))

write_csv(allgo_mf,paste(outputdir,"/GOterm_mf_summary_TUCRs.csv",sep=""))

write_csv(allgo_cc,paste(outputdir,"/GOterm_cc_summary_TUCRs.csv",sep=""))
```

## Summarize GO Term Data (new)

```{r summary GO Terms new}

TUCRids <- read.table(countfilename,header = TRUE)
TUCRids <- as.character(TUCRids[,4])

allgo <- read.csv(paste(outputdir,"/GOterm_all_summary_TUCRs.csv",sep=""),header=TRUE)
allgo <- allgo %>%
  dplyr::select(Term)

rbinder.allgo <- allgo
heatmap.allgo <- data.frame(matrix(ncol=3,nrow=0, dimnames=list(NULL, c("Term","gene","direction"))))

i = 1
for(i in 1:length(TUCRids)){
  print(i)
  TUCR <- TUCRids[i]
  skip_to_next_p <<- FALSE
  skip_to_next_n <<- FALSE
  positive <- tryCatch(read.csv(paste(outputdir,"/GO_Terms/",TUCR,"_plus_correlations.csv",sep=""),
                       header=TRUE), error = function(e) { skip_to_next_p <<- TRUE})

  negative <- tryCatch(read.csv(paste(outputdir,"/GO_Terms/",TUCR,"_minus_correlations.csv",sep=""),
                       header=TRUE), error = function(e) { skip_to_next_n <<- TRUE})

  
  if(skip_to_next_p & skip_to_next_n) { next } 
  
  pos_neg <- rbind(positive,negative)
  pos_neg <- pos_neg %>%
    filter(P.DE <= 0.05) %>%
    dplyr::select(Term,gene,direction)
  
  rbinder.allgo.TUCR <- dplyr::left_join(rbinder.allgo,pos_neg,by="Term")
  rbinder.allgo.TUCR$gene <- TUCR
  heatmap.allgo <- rbind(heatmap.allgo,rbinder.allgo.TUCR)
}

library(factoextra)

heatmap.allgo.wide <- heatmap.allgo %>%
  group_by(Term) %>%
  mutate(sum = sum(direction)) %>%
  ungroup() %>%
  arrange(desc(sum)) %>%
  mutate(Term2 = fct_reorder(Term,sum))

heatmap.allgo.summary <- heatmap.allgo.wide %>%
  group_by(Term,sum) %>%
  summarise() %>%
  arrange(desc(sum))

heatmap.allgo.top20 <- heatmap.allgo.summary[1:20,]
heatmap.allgo.bottom20 <- heatmap.allgo.summary[(length(heatmap.allgo.summary$Term)):(length(heatmap.allgo.summary$Term)-19),]
  

p <- ggplot(data = heatmap.allgo.wide,mapping = aes(x = gene,y = Term2, fill = direction)) +
  geom_tile() +
  scale_fill_gradient2(low = "blue", mid = "white", high = "red", midpoint = 0) + ggtitle(paste("Expression of TUCRs in ",disease,sep="")) +
  theme(plot.title = element_text(hjust = 0.5),
        panel.grid = element_blank(),
        axis.text = element_blank(),
        axis.ticks = element_blank())

ggsave(p,file=paste(outputdir,"/GOTerm_all_heatmap.png",sep=""))

```  

## Summarize GO Terms (old)

```{r summary GO Terms old}

allgo <- read.csv(paste(outputdir,"/GOterm_all_summary_TUCRs.csv",sep=""),header=TRUE)

allgo_bp <- read.csv(paste(outputdir,"/GOterm_bp_summary_TUCRs.csv",sep=""),header=TRUE)

allgo_mf <- read.csv(paste(outputdir,"/GOterm_mf_summary_TUCRs.csv",sep=""),header=TRUE)

allgo_cc <- read.csv(paste(outputdir,"/GOterm_cc_summary_TUCRs.csv",sep=""),header=TRUE)

allgo_sum <- allgo %>% 
  dplyr::group_by(Term) %>%
  dplyr::summarise(count = n()) %>%
  dplyr::arrange(desc(count)) %>%
  dplyr::mutate(Term=factor(Term, levels=Term))

allgo_sum <- allgo_sum[1:20,]

p <- ggplot(allgo_sum, aes(x=Term,y=count,fill=Term)) + geom_bar(stat="identity", width = 0.7) + geom_text(aes(label=round(count,2)), vjust=1.6, color="black", size=rel(2)) + ggtitle(paste("Most common TUCR GO Terms in ",disease,sep="")) +xlab("GO Term") + ylab("# of TUCRs") + 
  theme(
  plot.title = element_text(size=rel(1.5), face="bold",hjust = 0.5),
axis.title = element_text(size = rel(1.25), face="bold"),
legend.title = element_blank(),
legend.position = "none",
axis.text.x = element_text(size=rel(1.5)),
axis.text.y = element_text(size=rel(1.5)),
panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black")) + coord_flip()

ggsave(file=paste(outputdir,"/tucr_allgo.png",sep=""),plot = print(p), width = 15, height = 10, dpi = 300) 


allgo_bp_sum <- allgo_bp %>% 
  dplyr::group_by(Term) %>%
  dplyr::summarise(count = n()) %>%
  dplyr::arrange(desc(count)) %>%
  dplyr::mutate(Term=factor(Term, levels=Term))

allgo_bp_sum <- allgo_bp_sum[1:20,]

p1 <- ggplot(allgo_bp_sum, aes(x=Term,y=count,fill=Term)) + geom_bar(stat="identity", width = 0.7) + geom_text(aes(label=round(count,2)), vjust=1.6, color="black", size=rel(2)) + ggtitle(paste("Most common TUCR Biological Processes in ",disease,sep="")) + xlab("Biological Processes") + ylab("# of TUCRs") + theme(
  plot.title = element_text(size=rel(1.5), face="bold",hjust = 0.5),
axis.title = element_text(size = rel(1.25), face="bold"),
legend.title = element_blank(),
legend.position = "none",
axis.text.x = element_text(size=rel(1.5)),
axis.text.y = element_text(size=rel(1.5)),
panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black")) + coord_flip()

ggsave(file=paste(outputdir,"/tucr_allgo_bp.png",sep=""),plot = print(p1), width = 15, height = 10, dpi = 300) 

allgo_mf_sum <- allgo_mf %>% 
  dplyr::group_by(Term) %>%
  dplyr::summarise(count = n()) %>%
  dplyr::arrange(desc(count)) %>%
  dplyr::mutate(Term=factor(Term, levels=Term))

allgo_mf_sum <- allgo_mf_sum[1:20,]

p2 <- ggplot(allgo_mf_sum, aes(x=Term,y=count,fill=Term)) + geom_bar(stat="identity", width = 0.7) + geom_text(aes(label=round(count,2)), vjust=1.6, color="black", size=rel(2)) + ggtitle(paste("Most common TUCR Molecular Functions in ",disease,sep="")) + xlab("Molecular Function") + ylab("# of TUCRs") + theme(
  plot.title = element_text(size=rel(1.5), face="bold",hjust = 0.5),
axis.title = element_text(size = rel(1.25), face="bold"),
legend.title = element_blank(),
legend.position = "none",
axis.text.x = element_text(size=rel(1.5)),
axis.text.y = element_text(size=rel(1.5)),
panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black")) + coord_flip()

ggsave(file=paste(outputdir,"/tucr_allgo_mf.png",sep=""),plot = print(p2), width = 15, height = 10, dpi = 300) 

allgo_cc_sum <- allgo_cc %>% 
  dplyr::group_by(Term) %>%
  dplyr::summarise(count = n()) %>%
  dplyr::arrange(desc(count)) %>%
  dplyr::mutate(Term=factor(Term, levels=Term))

allgo_cc_sum <- allgo_cc_sum[1:20,]

p3 <- ggplot(allgo_cc_sum, aes(x=Term,y=count,fill=Term)) + geom_bar(stat="identity", width = 0.7) + geom_text(aes(label=round(count,2)), vjust=1.6, color="black", size=rel(2)) + ggtitle(paste("Most common TUCR Cellular Compartments in ",disease,sep="")) + xlab("Cellular Compartment") + ylab("# of TUCRs") + theme(
  plot.title = element_text(size=rel(1.5), face="bold",hjust = 0.5),
axis.title = element_text(size = rel(1.25), face="bold"),
legend.title = element_blank(),
legend.position = "none",
axis.text.x = element_text(size=rel(1.5)),
axis.text.y = element_text(size=rel(1.5)),
panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black")) + coord_flip()

ggsave(file=paste(outputdir,"/tucr_allgo_cc.png",sep=""),plot = print(p3), width = 15, height = 10, dpi = 300) 

allgo_bp_sum <- allgo_bp_sum %>%
  dplyr::mutate(Ontology = "Biological Process")

allgo_mf_sum <- allgo_mf_sum %>%
  dplyr::mutate(Ontology = "Molecular Function")

allgo_cc_sum <- allgo_cc_sum %>%
  dplyr::mutate(Ontology = "Cellular Compartment")

allgo_fig <- rbind(allgo_bp_sum,allgo_mf_sum)

p4 <- ggplot(allgo_fig, aes(x=Term,y=count,fill=Term)) + geom_bar(stat="identity",position="dodge", width = 0.7) + geom_text(aes(label=round(count,2)), vjust=1.6, color="black", size=rel(4)) + ggtitle(paste("Most common TUCR GO Terms in ",disease,sep="")) + xlab("GO Term") + ylab("# of TUCRs") + facet_grid(~ Ontology,scales="free",space="free") + theme(
  plot.title = element_text(size=rel(1.75), face="bold",hjust = 0.5),
axis.title = element_text(size = rel(1.25), face="bold"),
legend.title = element_blank(),
legend.position = "none",
axis.text.x = element_text(size=rel(1.5)),
axis.text.y = element_text(size=rel(1.5)),
strip.text = element_text(size=rel(1.5)),
panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black")) 

ggsave(file=paste(outputdir,"/tucr_allgo_fig.png",sep=""),plot = print(p4), width = 15, height = 15, dpi = 300)


```


## Generate figures for cell fractionation

```{r cellfrac_TUCR}

if(!dir.exists(paste(outputdir,"/cellfrac",sep=""))){
  dir.create(paste(outputdir,"/cellfrac",sep=""))
}

CellFrackin <- function(filename){
all_content = readLines(filename)
skip_second = all_content[-c(1:19)]
x <- read.csv(textConnection(skip_second), header = TRUE, stringsAsFactors = FALSE)
samplenames <- t(as.data.frame(strsplit(as.character(x$Sample), " ")))
colnames(samplenames) <- c("Line","Fraction")
x <- cbind(x,samplenames)
x<- x %>%
    dplyr::select(Well,Target,Sample,Line,Fraction,Cq) %>%
    dplyr::group_by(Target,Line,Fraction) %>%
    dplyr::summarize(meanCq = mean(Cq)) %>%
    dplyr::mutate(dct = meanCq - min(meanCq)) %>%
    dplyr::mutate(FoldChange = 2^(-dct)) %>%
    dplyr::mutate(ID = paste(Target,Line,Fraction,sep=" "))

p <- ggplot(x,aes(x = Target, y = FoldChange, fill = Fraction)) + 
  labs(title = "Cellular localization of TUCRs in GBM cell lines") +
  ylab("Fold Change") +
  geom_bar(stat = "identity",position="dodge") + 
  theme(axis.text.x = element_text(angle = -90, size = 10),
        axis.text.y = element_text(size = 10),
        axis.ticks.x = element_blank(),
        plot.title = element_text(size = 15, face = "bold",hjust = 0.5),
        axis.title.y = element_text("Fold Change",size = 10),
panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black")) +
  facet_wrap(~Line)

ggsave(p,file=paste(outputdir,"/",filename,".png",sep=""), width = 7.5, height = 5, dpi = 300)
x %>%
  select(-ID) %>%
  write_csv(paste(outputdir,"/",filename," (Output).csv",sep=""))
}

path <- list.files("cellfrac")
i = 1

for(i in 1:length(path)){
  print(i)
  CellFrackin(paste("cellfrac",path[i],sep="/"))
}

```

## Make figures for fold change

```{r overexp_TUCR}

overexp_TUCR_cp <- c("#C77CFF","#F8766D","#619CFF","#00BA38","#FFFF00")

filename <- "tucroverexpression_U87.csv"
x <- read.csv(filename)

p <- ggplot(x,aes(x = Target, y = FC, fill = Sample)) + 
  labs(title = "TUCR Overexpression in U87 cells") +
  ylab("Fold Change") +
  geom_bar(stat = "identity",position="dodge",colour="black") + 
  theme(axis.text.x = element_text(angle = -90, size = 10),
        axis.text.y = element_text(size = 10),
        axis.ticks.x = element_blank(),
        plot.title = element_text(size = 15, face = "bold",hjust = 0.5),
        axis.title.y = element_text("Fold Change",size = 10)) +
  facet_zoom(ylim = c(0, 2)) +
  scale_fill_manual(values=overexp_TUCR_cp)

ggsave(p,file=paste(outputdir,"/",filename,".png",sep=""), width = 7.5, height = 5, dpi = 300)


```

## Show that TUCRs are contained within novel genes

```{r novelgenes}
tucr_novelgenes <- read.csv("tucr_novelgenes.csv")
  
p <- ggplot(tucr_novelgenes, aes(x=row,y=value,fill=row)) + geom_bar(stat="identity", width = 0.7) +
  geom_text(aes(label=round(value,2)), vjust=1.6, color="black", size=rel(4)) + ggtitle(paste("Some intergenic TUCRs are \n contained within Novel Genes")) + xlab("Gene Annotation and Disease") + ylab("Number of Genes") + theme(
plot.title = element_text(size=rel(1.5), face="bold",hjust = 0.5),
axis.title = element_text(size = rel(1.25), face="bold"),
axis.text.x = element_text(angle = -90, size = 10),
legend.title = element_blank(),
legend.position = "none",
panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank(), axis.line = element_line(colour = "black"))

ggsave(file=paste(outputdir,"/tucr_novelgenes.png",sep=""),plot = print(p), width = 5, height = 5, dpi = 300) 
  

```
