---
title: 'T-type calcium channel expression in neuron connect versus neuron disconnected glioma'
author: "Myron Keith Gibert Jr"
date: "February 26, 2020"
output: pdf_document
toc: true
---

## Introduction

## Set parameters

```{r parameters}

countfilename <- "mousegene_counts.txt"

outputdir <- "Outputs"

```

## Install required programs

```{r setup, message = FALSE}
knitr::opts_chunk$set(echo = TRUE)

if (!require("tidyverse")) install.packages("tidyverse")
library("tidyverse")

if (!require("ggplot2")) install.packages("ggplot2")
library("ggplot2")

if (!require("ggforce")) install.packages("ggforce")
library("ggforce")

if (!require("ggrepel")) install.packages("ggrepel")
library("ggrepel")

if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
if (!requireNamespace("DESeq2", quietly = TRUE))BiocManager::install("DESeq2")
library("DESeq2")

```

```{r outputs}
if(!dir.exists(outputdir)){
  dir.create(outputdir)
}
```


```{r tinytex, include=FALSE}
## INSTALLING LATEX FOR RMARKDOWN

#RMarkdown requires LaTex to create pdf documents as an output. More information can be found [here](https://bookdown.org/yihui/rmarkdown/pdf-document.html). Alternatively, output can be set to "html_document" or "word_document". End users may install LaTex by setting the "wanttinytex" variable to FALSE and running the following script:

#Install tinytex to let RMarkdown create a pdf document? Default: wanttinytex <- FALSE
wanttinytex <- TRUE

if(tinytex:::is_tinytex()==FALSE && wanttinytex == TRUE){
if (!require("tinytex")) install.packages("tinytex")
tinytex::install_tinytex()
}
```

## Aquire the data using the SRA toolkit

```{bash setup, eval = FALSE}

#using SRA Toolkit version 2.10.5

#RNA-Seq connected Glioma 2

prefetch SRR444442

#RNA-Seq connected Glioma 1

prefetch SRR444478

```

## Generate fastq files

```{bash fastq, eval = FALSE}

for sra in SRR*
do
echo $sra
fastq-dump $sra
done

```

## Align mm10 genome

```{bash mm10, eval = FALSE}

wget https://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz
gunzip hg38.fa.gz

#Rivanna only.  Skip otherwise.
module load gcc/7.1.0
module load bowtie2/2.2.9

bowtie2-build hg38.fa hg38

```

## Align fastq files to reference genome

```{bash sam, eval = FALSE}

for fq in *.fastq
do
name=$(echo $fq | awk -F".fastq" '{print $1}')
echo $name
bowtie2 -x hg38 -U $fq -S $name.sam
done

```

## Convert sam to bam

```{bash bam, eval = FALSE}

for sam in *.sam
do
name=$(echo $sam | awk -F".sam" '{print $1}')
echo $name
samtools view -b $sam | samtools sort -o $name.bam
done
rm *sam

```

## Index bam files

```{bash index, eval = FALSE}

for bam in *bam
do
name=$(echo $bam | awk -F ".bam" '{print $1}')
echo $name
samtools index $bam
done

```

## Download mouse genes and convert to bed

I got this methodology from [here](https://github.com/stevekm/reference-annotations/blob/master/Makefile) since my other methods were not working.

```{bash refgenes, eval = FALSE}

macs2 callpeak -t SRR444442.bam -n SRR444442 -g hs
macs2 callpeak -t SRR444478.bam -n SRR444478 -g hs

wc -l *Peak

cat SRR444442_peaks.narrowPeak | cut -f 1-4 | sed -e "s/^/chr/g" | grep -v chrMT | grep -v chrGL | grep -v chrKI > h3k4me3.bed
cat SRR444478_peaks.narrowPeak | cut -f 1-4 | sed -e "s/^/chr/g" | grep -v chrMT | grep -v chrGL | grep -v chrKI > pol2.bed

```

## Generate counts for mouse genes

```{bash counts, eval = FALSE}

wget ftp://ftp.ensembl.org/pub/release-90/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz

gunzip Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz

samtools faidx Homo_sapiens.GRCh38.dna.primary_assembly.fa

awk -v OFS='\t' {'print $1,$2'} Homo_sapiens.GRCh38.dna.primary_assembly.fa.fai > hg38_genomeFile.txt

cat hg38_genomeFile.txt | grep -v KI | grep -v MT | grep -v GL | sed -e "s/^/chr/g" > temp

mv temp hg38_genomeFile.txt

```

## Identifying differentially expressed TUCRs with DESeq2

```{bash slop, results="hide", eval = FALSE}

bedtools slop -b 10000 -i hg38.ultraConserved.bed -g hg38_genomeFile.txt > tucrs_pm10kb.bed

bedtools slop -b 10000 -i TUCR_intergenic.bed -g hg38_genomeFile.txt > intergenic_tucrs_pm10kb.bed

bedtools slop -b 10000 -i CHESScoding.bed -g hg38_genomeFile.txt > coding_pm10kb.bed

bedtools slop -b 10000 -i CHESSlncRNA.bed -g hg38_genomeFile.txt > lncRNAs_pm10kb.bed

bedtools slop -b 10000 -i CHESSantisense.bed -g hg38_genomeFile.txt > antisense_pm10kb.bed

bedtools slop -b 10000 -i CHESSmisc.bed -g hg38_genomeFile.txt > misc_pm10kb.bed

```

## Fisher

```{bash fisher, results="hide", eval = FALSE}

bedtools fisher -a tucrs_pm10kb.bed -b pol2.bed -g hg38_genomeFile.txt > pol2_tucrs_output.txt

bedtools fisher -a intergenic_tucrs_pm10kb.bed -b pol2.bed -g hg38_genomeFile.txt > pol2_intergenic_tucrs_output.txt

bedtools fisher -a coding_pm10kb.bed -b pol2.bed -g hg38_genomeFile.txt > pol2_coding_output.txt

bedtools fisher -a lncRNAs_pm10kb.bed -b pol2.bed -g hg38_genomeFile.txt > pol2_lncRNA_output.txt

bedtools fisher -a antisense_pm10kb.bed -b pol2.bed -g hg38_genomeFile.txt > pol2_antisense_output.txt

bedtools fisher -a misc_pm10kb.bed -b pol2.bed -g hg38_genomeFile.txt > pol2_misc_output.txt

bedtools fisher -a tucrs_pm10kb.bed -b h3k4me3.bed -g hg38_genomeFile.txt > h3k4me3_tucrs_output.txt

bedtools fisher -a intergenic_tucrs_pm10kb.bed -b h3k4me3.bed -g hg38_genomeFile.txt > h3k4me3_intergenic_tucrs_output.txt

bedtools fisher -a coding_pm10kb.bed -b h3k4me3.bed -g hg38_genomeFile.txt > h3k4me3_coding_output.txt

bedtools fisher -a lncRNAs_pm10kb.bed -b h3k4me3.bed -g hg38_genomeFile.txt > h3k4me3_lncRNA_output.txt

bedtools fisher -a antisense_pm10kb.bed -b h3k4me3.bed -g hg38_genomeFile.txt > h3k4me3_antisense_output.txt

bedtools fisher -a misc_pm10kb.bed -b h3k4me3.bed -g hg38_genomeFile.txt > h3k4me3_misc_output.txt

```