---
title: 'Uncovering a role for transcribed ultra-conserved regions as long non-coding RNAs in glioblastoma'
author: "Myron Keith Gibert Jr"
date: "January 27, 2020"
output: pdf_document
toc: true
---

## Introduction
For this first programming assignment I wrote three functions that are meant to interact with dataset that accompanies this assignment. The dataset is contained in a zip file specdata.zip that is included in the GitHub repository.

## Set parameters

```{r parameters}

makerpkmboxplots <- FALSE

```

## Install required programs

```{r setup, message = FALSE}
knitr::opts_chunk$set(echo = TRUE)

if (!require("tidyverse")) install.packages("tidyverse")
library("tidyverse")

if (!require("ggplot2")) install.packages("ggplot2")
library("ggplot2")

if (!require("ggrepel")) install.packages("ggrepel")
library("ggrepel")

if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
if (!requireNamespace("DESeq2", quietly = TRUE))BiocManager::install("DESeq2")
library("DESeq2")

```

```{r tinytex, include=FALSE}
## INSTALLING LATEX FOR RMARKDOWN

#RMarkdown requires LaTex to create pdf documents as an output. More information can be found [here](https://bookdown.org/yihui/rmarkdown/pdf-document.html). Alternatively, output can be set to "html_document" or "word_document". End users may install LaTex by setting the "wanttinytex" variable to FALSE and running the following script:

#Install tinytex to let RMarkdown create a pdf document? Default: wanttinytex <- FALSE
wanttinytex <- TRUE

if(tinytex:::is_tinytex()==FALSE && wanttinytex == TRUE){
if (!require("tinytex")) install.packages("tinytex")
tinytex::install_tinytex()
}
```

## Stringtie

```{bash stringtie, eval = FALSE}
wget ftp://ftp.ensembl.org/pub/release-99/gtf/homo_sapiens/Homo_sapiens.GRCh38.99.gtf.gz
mv Homo_sapiens.GRCh38.99.gtf.gz stringtie
gunzip stringtie/Homo_sapiens.GRCh38.99.gtf.gz

module load gcc/7.1.0
module load stringtie/2.0.6

for bam in $(find GBM-RNASEQ-RAW/* -name '*.bam')
do
name=$(echo $bam | awk -F ".bam" '{print $1}')
name2=$(basename "$name")
if [ ! -f stringtie/$name2.gtf ] 
then
    echo "$name2.gtf does not exist"
    stringtie $bam -o $name.gtf
    mv $name.gtf ./stringtie
else
    echo "$name2.gtf exists... skipping"
fi

if [ ! -f $name.gtf ]; then
  stringtie $bam -o $name.gtf
fi
mv $name.gtf ./stringtie
done

cd stringtie

stringtie --merge *rehead.gtf -G Homo_sapiens.GRCh38.99.gtf -o stringtie_merged.gtf

intersectBed -a hg38.ultraConserved.bed -b CHESSgenes.bed -wa -wb > intersectchess_TUCRs.bed

intersectBed -a hg38.ultraConserved.bed -b stringtie_merged.gtf -wa -wb > intersectstringtie_TUCRs.bed

cp stringtie_merged.gtf stringtie_merged.bed
```

## Setup for TCGA RNA-Seq analysis

```{bash tcga, eval = FALSE}
## Create directory for all outputs from TCGA analysis data
mkdir tcga_analysis/
```

## Extract headers from BAM files

```{bash tcgaheader, eval = FALSE}
## Extract header information from BAM files
# Create directory for extracted headerfiles
mkdir tcga_analysis/head/

# Create extracted header files
for bam in $(find GBM-RNASEQ-RAW/* -name '*.bam')
do
name=$(echo $bam | awk -F ".bam" '{print $1}')
echo $name
samtools view -H $bam > $name.headerids.txt
mv $name.headerids.txt ./tcga_analysis/head/
done
```

## Extract TCGA patient barcodes from BAM headers

```{bash tcgaids, eval = FALSE}
## Get new file names for BAM files using TCGA IDs from extracted header
mkdir tcga_analysis/ids/

# Initialize summary file with TCGA IDs
echo "RAWID,TCGAID" > tcga_analysis/ids/tcgaIDs.csv 

# Extract IDs from header files
for header in $(find tcga_analysis/head/* -name '*headerids.txt')
do
name=$(echo $header | awk -F ".headerids.txt" '{print $1}')
name2=$(echo $name | cut -c20-)
ids1=$(awk '$1=="@RG"{print $3}' $header)
ids2=$(echo $ids1 | cut -c4-)
echo "$name2.bam,$ids2.bam"
echo "$name2.bam,$ids2.bam" > $name.tcgaid.csv
echo "$name2.bam,$ids2.bam" >> tcga_analysis/ids/tcgaIDs.csv 
mv $name.tcgaid.csv ./tcga_analysis/ids/
done
```

## Reindex TCGA BAM files under new names

```{bash tcgaindex, eval = FALSE}

##Rename BAM files using TCGA ids
# Create folder for new files
mkdir tcga_analysis/reindex/
# Copy BAM files into new directory with updated names

while IFS=, read orig target; do
orig2=$(find GBM-RNASEQ-RAW/* -name $orig)
echo $orig2
cp $orig2 $target
mv $target tcga_analysis/reindex/
done < tcga_analysis/ids/tcgaIDs.csv

# Reindex the renamed files using samtools.
for bam in $(find tcga_analysis/reindex/ -name '*.bam')
do
name=$(echo $bam | awk -F ".bam" '{print $1}')
echo $name
samtools index $bam
done
```

## Generating count tables for survival and differential expression.

```{bash tcgacounts, eval = FALSE}
mkdir tcga_analysis/foldchange

#Write header line
echo -e chrom"\t"start"\t"end"\t"id tcga_analysis/reindex/*bam | sed -e "s/ /\t/g" > tcga_analysis/foldchange/TUCR_counts.txt

#Counts
multiBamCov -bams tcga_analysis/reindex/*bam -bed hg38.ultraConserved.bed -q 10 >> tcga_analysis/foldchange/TUCR_counts.txt

#Write header line
echo -e chrom"\t"start"\t"end"\t"id *bam | sed -e "s/ /\t/g" > gene_counts.txt

#Counts
multiBamCov -bams *bam -bed CHESSgenes.bed -q 10 >> gene_counts.txt
```

## Acquiring total reads from RNA-Seq BAM files

```{bash, eval = FALSE}

mkdir tcga_analysis/readcounts/

module load samtools/1.9

for bam in $(find tcga_analysis/reindex/*.bam -name '*.bam')
do
name=$(echo $bam | awk -F ".bam" '{print $1}')
echo $name
samtools view -c -F 260 $bam > $name.readcounts.txt
mv $name.readcounts.txt ./tcga_analysis/readcounts/
done

echo "id,counts" > tcga_analysis/readcounts/mergedcounts.csv 

for txt in $(find tcga_analysis/readcounts/* -name '*.txt')
do
name=$(echo $txt | awk -F ".readcounts.txt" '{print $1}')
echo $name
reads=$(head $txt)
echo "$name,$reads" >> tcga_analysis/readcounts/mergedcounts.csv 
done

```

## Acquiring clinical survival data

```{bash, eval = FALSE}

wget http://gdac.broadinstitute.org/runs/stddata__2016_01_28/data/GBM/20160128/gdac.broadinstitute.org_GBM.Merge_Clinical.Level_1.2016012800.0.0.tar.gz

tar -xvzf gdac.broadinstitute.org_GBM.Merge_Clinical.Level_1.2016012800.0.0.tar.gz

mv gdac.broadinstitute.org_GBM.Merge_Clinical.Level_1.2016012800.0.0 GBM.Merge_Clinical

```

## Acquiring SNPs

```{bash, eval = FALSE}
mkdir SNPs

cd SNPs

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_1.bed.gz
gunzip bed_chr_1.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_10.bed.gz
gunzip bed_chr_10.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_11.bed.gz
gunzip bed_chr_11.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_13.bed.gz
gunzip bed_chr_13.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_12.bed.gz
gunzip bed_chr_12.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_14.bed.gz
gunzip bed_chr_14.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_15.bed.gz
gunzip bed_chr_15.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_16.bed.gz
gunzip bed_chr_16.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_17.bed.gz
gunzip bed_chr_17.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_18.bed.gz
gunzip bed_chr_18.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_19.bed.gz
gunzip bed_chr_19.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_2.bed.gz
gunzip bed_chr_2.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_20.bed.gz
gunzip bed_chr_20.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_21.bed.gz
gunzip bed_chr_21.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_22.bed.gz
gunzip bed_chr_22.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_3.bed.gz
gunzip bed_chr_3.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_4.bed.gz
gunzip bed_chr_4.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_5.bed.gz
gunzip bed_chr_5.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_6.bed.gz
gunzip bed_chr_6.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_7.bed.gz
gunzip bed_chr_7.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_8.bed.gz
gunzip bed_chr_8.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_9.bed.gz
gunzip bed_chr_9.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_X.bed.gz
gunzip bed_chr_X.bed.gz

wget ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606_b151_GRCh38p7/BED/bed_chr_Y.bed.gz
gunzip bed_chr_Y.bed.gz

module load gcc/7.1.0
module load bedtools/2.26.0

for bed in *.bed
do
echo $bed
intersectBed -a hg38.ultraConserved.bed -b $bed -wa -wb > $bed.txt
done

cat *.txt > merged.txt
```

## Generating PsychoTUCRs

```{bash, eval = FALSE}

bedtools random -g hg38.genome -n 481 -l 243 -seed 71346

```


## Completing survival analysis

```{r surcor, eval = FALSE, message = FALSE}
if(!file.exists("GBM.clin.merged.txt")){unzip("GBM.clin.merged.zip")}

tucrcounts <- read.table("TUCR_counts.txt",header = TRUE)

scal.exp <- function(x){
  res <- matrix(nrow=nrow(x),ncol=ncol(x))
  colnames(res) <- colnames(x)
  rownames(res) <- rownames(x)
  for (i in 1:dim(x)[1]){
    median_n <- median(x[i,])
    for (j in 1:dim(x)[2]){
      if(x[i,j]>median_n){
        res[i,j] <- 1
      }else{
        res[i,j] <- 0
      }
    }
  }
  return(res)
}

scaledata <- tucrcounts[,5:length(colnames(tucrcounts))]
rowdata <- tucrcounts[,1:4]
scaledata <- as.matrix(scaledata)
#countdata.exp <- scal.exp(scaledata)
#countdata.exp2 <- cbind(rowdata,countdata.exp)

write.csv(countdata.exp2,"countdataexp.csv")

survdata <- read.table("GBM.clin.merged.txt",header = TRUE)
survdata$survival <- as.numeric(survdata$survival)

```

## Identifying differentially expressed TUCRs with DESeq2

```{r DESeq2, results="hide", message = FALSE}

tucrcounts <- read.table("TUCR_counts.txt",header = TRUE)
metadata <- read_csv(file = "TUCR_counts_metadata.csv")

tucrcounts2 <- tucrcounts[,6:ncol(tucrcounts)]
rownames(tucrcounts2) <- tucrcounts[,5]


dds_TUCR <- DESeqDataSetFromMatrix(countData = tucrcounts2,
                              colData = metadata,
                              design = ~dex)
dds_TUCR <- DESeq(dds_TUCR)
res_TUCR <- results(dds_TUCR, tidy=TRUE)
res_TUCR <- as_tibble(res_TUCR)

write.csv(res_TUCR, file = "results_TUCR.csv")

```

## Generating volcano plot to visualize DE genes

```{r DESeq2plots, message = FALSE}

### Volcano plot

volcanoplot <- function (res,genes = "all",title = "Deregulated Genes", output = "results_volcanoplot.png"){
  #res <- res_TUCR
  #genes = c("uc.110","uc.62")
  #title = "Deregulated Genes"
  vres <- res
  vres <- vres %>% mutate(filter1 = abs(log2FoldChange)>=1,filter2 = padj <=0.05,filter3="",gene="")
  for (i in 1:length(vres$row)){
    if(is.na(vres$padj[i])){vres$filter3[i] <- "black"}else{
    if(vres$filter1[i] == TRUE && vres$filter2[i] == TRUE){
    vres$filter3[i] <- "red"
  }else{
      if(vres$filter1[i] == TRUE){
    vres$filter3[i] <- "blue"
      }else{vres$filter3[i] <- "black"}
  }}
    ifelse(genes!="all",ifelse(!is.na(match(vres$row[i],genes)),vres$gene[i] <- vres$row[i], ""),vres$gene[i] <- vres$row[i])
    }
  
  
  
  ## plot
  p <- ggplot(vres) +
        geom_point(aes(x=log2FoldChange, y=-log10(padj),col=filter3)) +
        scale_color_manual(values = c("black","blue","red")) +
        ggtitle(title) +
        xlab("log2 fold change") + 
        ylab("-log10 adjusted p-value") +
        theme(legend.position = "none",
              plot.title = element_text(size = rel(1.5), hjust = 0.5),
              axis.title = element_text(size = rel(1.25))) +
        geom_text_repel(aes(x=log2FoldChange, y=-log10(padj),label = gene),force=50)
  
  ggsave(output)
}

volcanoplot(res_TUCR,genes=c("uc.62","uc.110","uc.193","uc.210","uc.427"),title = "TUCR deregulation in GBM", output = "tucr_results_volcanoplot.png")

```

## Generate RPKMs for filtering

```{r RPKM, results="hide"}

rpkmcounts <- tucrcounts[,6:ncol(tucrcounts)]
rpkmcounts.info <- tucrcounts[,1:5]
rownames(rpkmcounts) <- tucrcounts$id

genelength <- tucrcounts$length

seqdepth <- read.csv(file = "mergedcounts.csv",row.names=1)

seqdepth <- as.vector(seqdepth$counts)

rpkm <- function(counts,len,dep){
  x <- counts/len
  return(t(t(x)/dep))
}

rpkm.TUCRs <- rpkm(rpkmcounts,genelength,seqdepth)
rpkm.TUCRs2 <- cbind(rpkmcounts.info,rpkm.TUCRs)

write.csv(rpkm.TUCRs2,"rpkm.TUCRs.csv")
```

## Generate RPKM Box Plots for each TUCR

```{r boxplot, results = "hide"}
if(makerpkmboxplots == TRUE){
  
if(!dir.exists("RPKMboxplots")){
  dir.create("RPKMboxplots")
}
rpkm.dotplot <- rpkm.TUCRs2[,5:ncol(rpkm.TUCRs2)]
TUCRids <- rpkm.dotplot[,1]

i = 1
for(i in 1:length(TUCRids)){
TUCR <- "uc.110"
TUCR <- TUCRids[i]
print(i)
rpkm.dotplot2 <- rpkm.dotplot %>% 
  gather(key = "Sample", value = "RPKM",-id) %>%
  filter(id == TUCR)

p<- ggplot(rpkm.dotplot2, aes(x=id, y=RPKM)) +
  geom_boxplot() +
  geom_dotplot(binaxis='y', stackdir='center',
               stackratio=0.75, dotsize=0.5) + 
  ggtitle(paste("RPKM expression of ",TUCR)) +
  theme(plot.title = element_text(hjust = 0.5)) +
  labs(y="Reads per kilobase million (RPKM)", x = "TUCR") +
  stat_summary(fun.y=mean, geom="point", shape=20, size=10, color="red", fill="red")

ggsave(p,file=paste("RPKMboxplots/",TUCR,"_rpkm_boxplot.png",sep = ""))
}

}

```

